---
layout: default
title: Example Workflow
description: How to run the MADRID Workflow
show_in_nav: true
nav_bar_name: Example Workflow
weight: 2
permalink: /example-workflow
---
{% include workflow_header.html %}

<div>
  <p id="#example-header">
    This example will identify drug targets for <a href="https://en.wikipedia.org/wiki/Rheumatoid_arthritis">Rheumatoid arthritis</a> using GSMN of na√Øve <a href="https://en.wikipedia.org/wiki/T_helper_cell">CD4+ T-cell subtypes</a>.
  </p>
  <p>
    Follow the steps found at <a href="{{ "/#start-container" | relative_url }}">Starting the Container</a> for help on getting started.
  </p>
  <br>
</div>

<div id="example-step-1">
  <h2>
    <a href="#example-step-1" class="body-link">
      Step 1: Setting up Jupyter Notebook
    </a>
  </h2>
  <p>
    The outputs of <a href="https://github.com/alexdobin/STAR">STAR Aligner</a> with the <code>--GeneCounts</code> argument can be directly interfaced with MADRID. Under the folder <code>/data/STAR_output/</code> folder, we provide a template structure with Naive B control cells from Bulk RNA-sequencing experiments found in <a href="https://www.ncbi.nlm.nih.gov/geo/">NCBI's Gene Expression Omnibus</a>. We can run the file <code>/py/rnaseq_preprocess.py</code> with the <code>create_count_matrix</code> argument set to <code>True</code> to merge the counts from each replicate in each study into a single matrix. This follows <b>Step 1</b> in the MADRID container.
  </p>
  <p>
    Alternatively, you can provide a count matrix directly and run with <code>create_count_matrix</code> set to <code>False</code> to just fetch the necessary information for normalization.
  </p>
</div>

<div id="example-step-2">
  <h2>
    <a href="#example-step-2" class="body-link">
      STEP 2 GOES HERE
    </a>
  </h2>
  <p>
    In the <code>/data/config_sheets/</code> folder, we included GEO accession numbers of microarray data in <code>micrarray_data_inputs.xlsx</code>, and sample names of protemics data in <code>proteomics_data_inputs.xlsx</code>. The protein abundance data is provided under <code>/data/config_sheets/ProteomicsDataMatrix_Naive.csv</code>. Sample names for Bulk RNA-sequencing data is given in <code>/data/config_sheets/bulk_data_inputs.xlsx</code>. Running <code>/py/rnaseq_preprocess.py</code> will create <code>data/results/Gene_Info_[CELL_TYPE].csv</code> and <code>/data/data_matrices/BulkRNAseqMatrix_[CELL_TYPE].csv</code> if <code>create_counts_matrix</code> is set to <code>True</code>.
  </p>
  <p>
    <code>/data/config_sheets/micrarray_data_inputs.xlsx</code> includes microarray samples of naive CD4+ T cells from GSE22886, GSE43005, GSE22045, and GSE24634. The file <code>/data/config_sheets/proteomics_data_inputs.xlsx</code> contains sample names of naive CD4+ T cells, with its results found in the file <code>/data/data_matrices/Naive/ProteomicsDataMatrix_Naive.csv</code>.
  </p>
  <p>
    Using <code>/py/merge_xomics.py</code>, you can specify any number of available data sources (microarray, bulk RNA-seq, and proteomics) as inputs. You can also set the <code>expression_requirement</code> parameter, which defines the minimum number of data sources that must have gene expression above the threshold limit for the said gene to be considered active.
    <br>
    Note: If a gene is not supported by a data source/platform, the <code>expression_requirement</code> value will decrease by one for each input data source that does not support the gene.
  </p>
  <p>
    Running <b>Step 1</b> in the MADRID container will generate "gene activity" files based on transcriptomics and proteomics data, as described by <a href="https://doi.org/10.1038/s41540-020-00165-3"><cite>Puniya et al., 2020</cite></a>.
  </p>
  <p>
    This will save final output in <code>GeneExpression_Naive_Merged.csv</code> and its path in <code>step1_results_files.json</code>.
  </p>
</div>

<div id="example-step-3">
  <h2>
    <a href="#example-step-3" class="body-link">
      STEP 3 GOES HERE
    </a>
  </h2>
  <p>
    Our pipeline includes a modified version of the <a href="https://doi.org/10.1038/nbt.4072"><cite>Recon3D</cite></a> model to use as a reference for model contextualization. The modified version of Recon3D is available at <code>/data/GeneralModel.mat</code>.
  </p>
  <p>
    <b>Step 4</b> in MADRID will use <code>GeneExpression_Naive_Merged.csv</code> (from Step 1, above) in combination with the <code>GeneralModel.mat</code> to construct a cell-type specific model of Naive CD4+ cells.
  </p>
  <p>
    This step will save the output file at <code>/data/Naive/Naive_SpecificModel.json</code>.
  </p>
  <p>
    Running MADRID <b>Step 4</b> with gene activity and Recon3D will generate a model for naive CD4+ T cells. We can use this model in the next steps. However, we advise users to properly investigate, manually curate, and reupload the refined version in <code>/data/results/</code> to use in <b>Step 4</b>. We provided pre-curated versions of the Naive CD4+ T cell model at <code>/data/Naive/NaiveModel.mat</code>.
  </p>
</div>

<div id="example-step-4">
  <h2>
    <a href="#example-step-4" class="body-link">
      STEP 4 GOES HERE
    </a>
  </h2>
  <p>
    We used a dataset (GSE56649) of Rheumatoid arthritis to identify differentially expressed genes (disease genes). We defined accession IDs of this dataset in the input file <code>/data/config_sheets/disease/disease_data_inputs_Naive.xlsx</code>.
  </p>
  <p>
    This step will generate files <code>/data/Disease_UP_GSE56649.txt</code> and <code>/data/Disease_DOWN_GSE56649.txt</code>, and their paths, at <code>step2_results_files.json</code>. Finally, this step will create a <code>disease_files</code> variable that will include paths of files for up- and down-regulated genes.
  </p>
</div>

<div id="example-step-5">
  <h2>
    <a href="#example-step-5" class="body-link">
      STEP 5 GOES HERE
    </a>
  </h2>
  <p>
    This step will use the model (constructed in <b>Step 2</b>/uploaded curated version) and perform knock-out simulations of genes overlapping with the drug-target data file obtained from the <a href="https://www.broadinstitute.org/connectivity-map-cmap">ConnectivityMap</a> database. We refined the drug target-data file and included it at <code>/data/RepurposingHub.txt</code>.
  </p>
  <p>
    This step will use the following files:
  </p>
  <ul>
    <li><code>Naive_SpecificModel.json</code> (or a pre-curated version uploaded as <code>NaiveModel.mat</code>)</li>
    <li><code>Disease_UP_GSE56649.txt</code></li>
    <li><code>Disease_DOWN_GSE56649.txt</code></li>
    <li><code>RepurposingHub.txt</code></li>
  </ul>
  <p>
    The final output files will include drug targets ranked based on Perburbation Effect Score (PES) as described by <a href="https://doi.org/10.1038/s41540-020-00165-3"><cite>Puniya et al., 2020</cite></a>.
  </p>
  <p>
    The output file <code>/output/d_score.csv</code> will contain Entrez IDs of ranked genes and their corresponding PES. The file <code>drug_score.csv</code> will contain PES ranked drug targets (Entrez IDs and Gene Symbols) with mapped repurposed drugs.
  </p>
</div>
