---
layout: default
title: User-Defined Data
description: Using your own data in MADRID
show_in_nav: false
nav_bar_name: User-Defined Workflow
weight: 1
permalink: /user-defined-workflow
---
{% include workflow_header.html %}

<div id="user-step-1">
  <h2>
    <a href="#user-step-1" class="body-link">
      Step 1
    </a>
  </h2>
  <p>
    See <code></code>
  </p>
</div>


<div id="user-step-2">
  <h2>
    <a href="#user-step-2" class="body-link">
      Step 2
    </a>
  </h2>
  <p>
    If proteomics data is available, upload a data file with protein abundances (see <code>/data/data_matrices/Naive/ProteomicsDataMatrix_Naive.csv</code>) and define sample names in the configuration file for proteomics data (see <code>/data/config_sheets/proteomics_data_inputs.xlsx</code>). For multiple cell types, use the same names for the same subtype in transcriptomics (sheet names) and proteomics (headers) config files.
  </p>
</div>


<div id="user-step-3">
  <h2>
    <a href="#user-step-3" class="body-link">
      Step 3
    </a>
  </h2>
  <p>
    We provide a generic human metabolic model in <code>/data/GeneralModel.mat</code> for the pipeline. If you wish to provide a new version of the human metabolic model, it should contain Entrez Gene IDs in the gene rules.
  </p>
</div>


<div id="user-step-4">
  <h2>
    <a href="#user-step-4" class="body-link">
      Step 4
    </a>
  </h2>
  <p>
    See <code>/data/data_matrices</code>
  </p>
</div>


<div id="user-step-5">
  <h2>
    <a href="#user-step-5" class="body-link">
      Step 5
    </a>
  </h2>
  <p>
    For using the automatically created model(s), no change is required in the previous step. If you want to use refined models, upload them in <code>.mat</code> format, uncomment the relevant lines, and change the file names in <b>Step 4.3</b> of the MADRID container.
  </p>
</div>

<div id="user-step-6">
  <h2>
    <a href="#user-step-6" class="body-link">
      Step 6
    </a>
  </h2>
  <p>
    See final outputs in <code>/output</code>.
  </p>
</div>
