{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **MADRID - MetAbolic Drug Repurposing and IDentification** \n",
    "\n",
    "## **Who should use this pipeline?**\n",
    "\n",
    "This Jupyterlab pipeline has everything necessary to build a context-specific constraint-based metabolic model (CBMM) from any single source or combination of sources from the following '-omics' data.\n",
    "- Bulk RNA-seq\n",
    "- Single-cell RNA-seq\n",
    "- Proteomics (specific name???)\n",
    "- Microarray"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It also serves as a platform to use these models to identify drug targets and potentially repurposable drugs for metabolism-impacting diseases.\n",
    "\n",
    "MADRID does not require any amount of programming experience to create models. However every step of the pipeline is packaged in its own .py file to promote accessible modification, addition, or replacement of analysis steps. The Jupyterlab container comes pre-loaded with the most popular R and Python libraries, however, if you would like to use a library and cannot install it for any reason, please request it on our Github page!<br>\n",
    "https://github.com/HelikarLab/MADRID\n",
    "\n",
    "In this pipeline, **the term \"context\" refers to a specific state that can be subset from a genome-scale metabolic model using experimental data.** This context can be a specific type of cell, or tissue in a specific experimental state such as a control or a disease. For drug perturbation scoring of a specific cell-type or tissue, **it is only necessary to build a CBMM (Steps 1-5) for the healthy control state.** Differential gene expression (Step 6) will be used to for disease analysis so that multiple diseaes can be analyzed using the same CBMM.\n",
    "\n",
    "***Warning! If you terminate your session after running the Docker, any changes you make WILL NOT BE SAVED! Please mount a local directory to the Docker image as instructed on the Github and Dockerhub README's to prevent data loss.**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "\n",
    "## **Before You Start**\n",
    "\n",
    "### **Before running, you must provide the proper files depending on the types of data you will use for model creation.**\n",
    "- For RNA-seq: Either a correctly formatted folder named \"MADRID_input\" in the the data directory. Proper inputs can be generated using our Snakemake pipeline specifically designed for MADRID: https://github.com/HelikarLab/FastqToGeneCounts.  RNA-seq data can either be single-cell or bulk, but our Snakemake pipeline provided is for bulk only at the time. If processing RNA-seq data with an alternative procedure, or importing a pre-made gene count matrix, follow the instructions in Step 1.\n",
    "\n",
    "- For proteomics. A matrix of measurement ****NEED TO DISCUSS WITH BHANWAR WHAT THESE VAULES MEAN***** where rows are proteins in Entrez format and columns are arbitrary sample names.\n",
    "\n",
    "- For microarray, results must be uploaded to Gene Expression Omnibus, the only thing to provide in MADRID is a configuration file with GSE, GSM, and GPL codes. microarray_data_inputs.xlsx shows a template to use. Note that microarray has become obsolete and it is recomended to use RNA-seq if possible.\n",
    "\n",
    "### **Six steps to identifying drug targets, stop after step 3 if building a context-specific model for other purposes**\n",
    "        \n",
    "1. Preprocess Bulk RNAseq data by converting STAR outputed gene counts files into a unified matrix, fetch necessary info about each gene required for normalization, and generate a configuration sheet.\n",
    "\n",
    "2. Analyze any combination of microarray, RNAseq (total, polyA, single-cell), and proteomics data, output a list of active genes for each strategy.\n",
    "\n",
    "3. Check for consensus amongst strategies according to desired rigor and merge into a singular set of active genes\n",
    "\n",
    "4. Create tissue specific models based on the list of active genes. If required the user can manually refine these models and supply them in Step 4. \n",
    "\n",
    "5. Identify differential gene expressions from disease datasets using either microarray or bulk RNAseq transcriptomics information.\n",
    "\n",
    "6. Identify drug targets and repruposable drugs. This step consists of four substeps. \n",
    "    - mapping drugs on automatically created or user-supplied models\n",
    "    - knock-out simulation\n",
    "    - compare simulation results of perturbed and unperturbed models\n",
    "    - integrate with disease genes and score drug targets.\n",
    "\n",
    "### **Configuration sheet information**\n",
    "The user should upload config .xlsx files to `/work/data/config_sheets`. The sheet names in these config files should correspond to the context (tissue name, cell name, control, etc.) where each sheet contains the sample names to include in that context-specifc model. These sample names should correspond to the sample (column) names in the source data matrix which should be uploaded (or outputed) in `/work/data/data_matrices/<model name>/`\n",
    "    \n",
    "In the original Docker image, some exemplary input files are included to build metabolic models of naive, Th1, Th2, and Th17 subtypes and identify drug targets for rheumatoid arthritis. User can follow the documentation and the format of the exemplary input files and and use the provided template files to create your own input files."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Step 1: Initialize and Preprocess RNA-seq data**\n",
    "**(Skip if not using RNA-seq)**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "RNA-seq data is read by MADRID as count matrix where each column is a different sample/replicate named 'exampleTissueName_SXRYrZ' where:\n",
    "- X is the study (or batch) number\n",
    "- Y is the replicate number\n",
    "- Z is the run number. If the replicate does not contain multiple runs for a single replicate, then \"rZ\" should be neglected.\n",
    "- exampleTissueName is the name of the model that will be built from this data. It should be consistant with other data sources if they are to be integrated. **Note that this identifier should not have any special characters including \"\\_\" since it may interfere with parsing.**\n",
    "<br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Replicates should come from the same study/batch group and different study/batch numbers can come from different published studies (or batches) as long as the tissue/cell was under similar enough conditions for your modeling purpose. Run numbers in the same replicate will be summed together.\n",
    "<br>\n",
    "<br>\n",
    "#### **Example:**\n",
    "Say S1 represents a study or batch, and S2 represents a different study or batch of RNA-seq data from m0 macrophages whose model we will name m0Macro. The studies were conducted in a different lab, by different researchers at different times using a different library preparation kit. In each study there are two replicates (R1, and R2) obtained for each one. m0Macro_S1R1 and m0Macro_S1R2 will be checked for consensus to generate a list of active genes in both replicates. These active genes will then be checked for consensus with the consensus of m0Macro_S2R1 and m0Macro_S2R2 to output a list of active genes in both studies.\n",
    "<br><br>\n",
    "The reason this system is used is not only to help keep you and MADRID organized. Most types of normalized gene counts are not good for direct comparisons across replicates, and are especially not suitable for comparisons across different experiments. Therefore, MADRID will convert normalized gene counts into a boolean list of active genes. These lists will be compared at the the level of replicates in a batch (or study) and then again at the level of all provided batches (or study). Finally, the active genes will be merged with the outputs of proteomics, microarray and differnt RNA-seq stategies if provided. The rigor used at each level can be easily modified by the user.\n",
    "<br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Two ways to initialize RNA-seq data**\n",
    "\n",
    "1. Import a properly formatted `MADRID_inputs` folder in the `data` directory. \n",
    "<br>\n",
    "\n",
    "> **It is recommended that you use our Snakemake pipeline to align and create a properly formated MADRID_inputs folder.** The pipeline also runs a series of important quality control methods to help determine if any of the provided samples are not suitable for model creation. This pipeline can be found at https://github.com/HelikarLab/FastqToGeneCounts.\n",
    "\n",
    "> Or, **if using your own alignment protocol**, follow this guide to create a `MADRID_inputs` folder.\n",
    "\n",
    "> - The top level of the directory has seperate tissues/cells to create seperate models from. The next level must have a folder called `geneCounts` and optionally a `strandedness` folder. If using zFPKM normalization, there should also be two more folders called `layouts` and `fragmentSizes`. Inside each of these folders should be folders named SX (wherer X is an arbitrary user-defined number that replicates are associated with. \n",
    "<br>    \n",
    "\n",
    "> - Inside these study number folders of `geneCounts` should be outputs of STAR aligner with using --quantMode GeneCounts. To help MADRID (and you!) stay organized, these outputs should be renamed `exampleTissueName_SXRYrZ.tab` where X is the study (or batch) number, Y is the replicate number, and Z is the run number. If the replicate does not contain multiple runs the rZ should be neglected. Replicates should come from the same study/sample group and different samples can come from different published studies (or batches) as long as the tissue/cell was under similar enough conditions for your modeling purpose.\n",
    "<br>\n",
    "\n",
    "> - Inside the study number folders of `strandedness` should be files named `exampleTissue_SXRYrZ_strandedness.txt`. These files must tell the strandedness of the the RNA-seq method used, and must contain one of the following text (and nothing else):\n",
    ">> * NONE\n",
    ">> * FIRST_READ_TRANSCRIPTION_STRAND\n",
    ">> * SECOND_READ_TRANSCRIPTION_STRAND\n",
    "<br>\n",
    "\n",
    "> - Inside the study number folders of `layouts` should be files named `exampleTissue_SXRYrZ_layout.txt` where each file tells the layout of the library used, and must contain one of the following text (and nothing else):\n",
    ">> * paired-end\n",
    ">> * single-end\n",
    "<br>\n",
    "\n",
    "> - Inside the study number folders of `fragmentSizes` should be files named `exampleTissue_SXRYrZ_fragment_sizes.txt` and contain the output of RSeQC's RNA_fragment_size.py function.\n",
    "\n",
    "> - Inside the study number folders of `prepMethods` should be files named `exampleTissue_SXRYrZ_prep_method.txt` where each file tells the library preparation strategy, which must be one of the following.\n",
    ">> * total\n",
    ">> * mRNA\n",
    "<br><br>\n",
    "where 'total' refers to Total RNA and mRNA refers to polyA enriched RNA. Note that these strategies serve only to differentiate the methods in the event that both are used to build a model. If a different library strategy is desired, you can use either one as a placeholder, or will very little Python knowledge, it is easy add a new strategy to `merge_xomics.py`.  \n",
    "\n",
    "2. Import a properly formatted counts matrix in `/work/data/data_matrices/exampleTissue/gene_counts_matrix_exampleTissue.csv` where the rows are name exampleTissue_SXRY (note the lack of run number since runs should be summed into a single set of counts). **If providing the count matrix this way, instead of generating one using method 1, you will also have to create a configuration file** that has each sample's name, study/batch number, and if using zFPKM, layout and mean fragment length. **Use the provided template** to help. Once provided, run rnaseq_preprocess.py with the '--provide-matrix' argument. <br><br> ***This method is best if you are downloading a premade count matrix, or using single-cell data that has already been batch corrected, clustered, and sorted into only the cell type of interest!**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### **Structure of MADRD_input folder**\n",
    "├── STAR_output <br />\n",
    "&emsp;&emsp;&emsp;└── exampleTissue <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── geneCounts <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── S1 <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S1R1.tab <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S1R2.tab <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S1R3.tab <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;└── exampleTissue_S1R4.tab <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── S2 <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S2R1.tab <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S2R2.tab <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S2R3.tab <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;└── exampleTissue_S2R4.tab <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;└── exampleTissue_S2R4.tab <br />\n",
    "&nbsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;|&emsp;&emsp;&emsp;&ensp;└── S3 <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; ├── exampleTissue_S3R1.tab <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; ├── exampleTissue_S3R2.tab <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; ├── exampleTissue_S3R3r1.tab <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; ├── exampleTissue_S3R3r2.tab <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; ├── exampleTissue_S3R3r3.tab <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; ├── exampleTissue_S3R4r1.tab <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; ├── exampleTissue_S3R4r2.tab <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; ├── exampleTissue_S3R4r3.tab <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; ├── exampleTissue_S3R5r1.tab <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; ├── exampleTissue_S3R5r2.tab <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; ├── exampleTissue_S3R5r3.tab <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; ├── exampleTissue_S3R6r1.tab <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; ├── exampleTissue_S3R6r2.tab <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; └── exampleTissue_S3R6r3.tab <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── strandedness <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── S1 <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├──  exampleTissue_S1R1_strandedness.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S1R2_strandedness.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S1R3_strandedness.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;└── exampleTissue_S1R4_strandedness.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── S2 <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S2R1_strandedness.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S2R2_strandedness.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S2R3_strandedness.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;└── exampleTissue_S2R4_strandedness.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;└── S3 <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R1_strandedness.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R2_strandedness.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R3r1_strandedness.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R3r2_strandedness.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R3r3_strandedness.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R4r1_strandedness.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R4r2_strandedness.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R4r3_strandedness.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R5r1_strandedness.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R5r2_strandedness.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R5r3_strandedness.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R6r1_strandedness.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R6r2_strandedness.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;└── exampleTissue_S3R6r3_strandedness.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── layouts <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── S1 <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S1R1_layout.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S1R2_layout.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S1R3_layout.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;└── exampleTissue_S1R4_layout.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── S2 <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S2R1_layout.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S2R2_layout.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S2R3_layout.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;└── exampleTissue_S2R4_layout.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;└── S3 <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R1_layout.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R2_layout.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R3r1_layout.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R3r2_layout.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R3r3_layout.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R4r1_layout.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R4r2_layout.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R4r3_layout.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R5r1_layout.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R5r2_layout.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R5r3_layout.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R6r1_layout.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R6r2_layout.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;└── exampleTissue_S3R6r3_layout.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── prepMethods (Optional, unless using multiple RNA-seq types) <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── S1 <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S1R1_prep_method.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S1R2_prep_method <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S1R3_prep_method.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;└── exampleTissue_S1R4_prep_method.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── S2 <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S2R1_prep_method.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S2R2_prep_method.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S2R3_prep_method.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;└── exampleTissue_S2R4_prep_method.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;└── S3 <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R1_prep_method.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R2_prep_method.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R3r1_prep_method.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R3r2_prep_method.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R3r3_prep_method.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R4r1_prep_method.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R4r2_prep_method.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R4r3_prep_method.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R5r1_prep_method.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R5r2_prep_method.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R5r3_prep_method.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R6r1_prep_method.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R6r2_prep_method.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;└── exampleTissue_S3R6r3_prep_method.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;└── fragementSizes (optional, for zFPKM only) <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; &emsp;&emsp;&emsp;├── S1 <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; &emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S1R1_fragment_size.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; &emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S1R2_fragment_size.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; &emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S1R3_fragment_size.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; &emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;└── exampleTissue_S1R4_fragment_size.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; &emsp;&emsp;&emsp;├── S2 <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; &emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S2R1_fragment_size.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; &emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S2R2_fragment_size.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; &emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;├── exampleTissue_S2R3_fragment_size.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; &emsp;&emsp;&emsp;│&emsp;&emsp;&emsp;└── exampleTissue_S2R4_fragment_size.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; &emsp;&emsp;&emsp;└── S3  <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R2_fragment_size.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R3r1_fragment_size.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R3r2_fragment_size.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R3r3_fragment_size.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R4r1_fragment_size.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R4r2_fragment_size.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R4r3_fragment_size.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R5r1_fragment_size.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R5r2_fragment_size.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R5r3_fragment_size.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R6r1_fragment_size.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;├── exampleTissue_S3R6r2_fragment_size.txt <br />\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;└── exampleTissue_S3R6r3_fragment_size.txt <br />"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Currently, MADRID can filter raw RNA-seq counts using three normalization techniques.\n",
    "\n",
    "> - TPM Quantile, where each replicate is normalized with Transcripts-per-million and an upper quantile is taken to create a boolean list of active genes for the replicate. Replicates are compared for consensus within the study/batch number according to user-defined ratios and then study/batch numbers are checked for consensus according to different user defined ratios. **Recomended if user wants more control over the size of the model, like a smaller model that allows for only the most expressed reactions, or a larger more encompassing one that contains less essential reactions.\n",
    "\n",
    "> - zFPKM method outlined in: https://pubmed.ncbi.nlm.nih.gov/24215113/ can be used. Counts will be normalized using zFPKM and genes > -3 will be considered expressed per thier recommendation. Expressed genes will be checked for consensus at the replicate and study/batch levels the same as TPM Quantile. **Recommended if user wants to give least input over gene essentially determination and use the most standardized method of active gene determination. \n",
    "\n",
    "> - flat cutoff of CPM (counts per million) normalized values, check for consensus the same as other methods. **Not recommended**\n",
    "\n",
    "Regardless of normalization technique used, or provided files used for RNA-seq, preprocessing is required to fetch relevent gene information needed for harmonization and normalization such as Entrez ID, and the start and end postions.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import necessary python packages\n",
    "import sys\n",
    "import os\n",
    "import pandas\n",
    "import numpy\n",
    "import json\n",
    "import re\n",
    "from subprocess import call\n",
    "from project import configs\n",
    "import bioservices\n",
    "import pprint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001B[0;1;31mSystem has not been booted with systemd as init system (PID 1). Can't operate.\u001B[0m\n",
      "\u001B[0;1;31mFailed to create bus connection: Host is down\u001B[0m\n",
      "['rnaseq_preprocess.py', '--context-names', \"['m0Macro', 'm1Macro', 'm2Macro', 'naiveB', 'nsmB', 'smB', 'plasmaB']\", '--gene-format', 'Ensembl', '--taxon-id', 'human', '--create-matrix']\n",
      "Preprocessing m0Macro\n",
      "Gene info output directory is \"/home/jupyteruser/work/data/results/m0Macro\"\n",
      "Looking for STAR gene count tables in '/home/jupyteruser/work/data/MADRID_input/m0Macro'\n",
      "Creating Counts Matrix for 'm0Macro'\n",
      "[1] \"Organizing Files\"\n",
      "[1] \"Creating counts matrix\"\n",
      "Count Matrix written at  /home/jupyteruser/work/data/data_matrices/m0Macro/gene_counts_matrix_full_m0Macro.csv \n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m0Macro/geneCounts/S1/m0Macro_S1R1r3.tab\n",
      "label:  S1R1r3\n",
      "['r3']\n",
      "rep #:  R1\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m0Macro/geneCounts/S1/m0Macro_S1R3r2.tab\n",
      "label:  S1R3r2\n",
      "['r2']\n",
      "rep #:  R3\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m0Macro/geneCounts/S1/m0Macro_S1R2r1.tab\n",
      "label:  S1R2r1\n",
      "['r1']\n",
      "rep #:  R2\n",
      "m0Macro_S1R2\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m0Macro/geneCounts/S1/m0Macro_S1R2r3.tab\n",
      "label:  S1R2r3\n",
      "['r3']\n",
      "rep #:  R2\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m0Macro/geneCounts/S1/m0Macro_S1R1r2.tab\n",
      "label:  S1R1r2\n",
      "['r2']\n",
      "rep #:  R1\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m0Macro/geneCounts/S1/m0Macro_S1R1r1.tab\n",
      "label:  S1R1r1\n",
      "['r1']\n",
      "rep #:  R1\n",
      "m0Macro_S1R1\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m0Macro/geneCounts/S1/m0Macro_S1R3r1.tab\n",
      "label:  S1R3r1\n",
      "['r1']\n",
      "rep #:  R3\n",
      "m0Macro_S1R3\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m0Macro/geneCounts/S1/m0Macro_S1R2r2.tab\n",
      "label:  S1R2r2\n",
      "['r2']\n",
      "rep #:  R2\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m0Macro/geneCounts/S1/m0Macro_S1R3r4.tab\n",
      "label:  S1R3r4\n",
      "['r4']\n",
      "rep #:  R3\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m0Macro/geneCounts/S1/m0Macro_S1R2r4.tab\n",
      "label:  S1R2r4\n",
      "['r4']\n",
      "rep #:  R2\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m0Macro/geneCounts/S1/m0Macro_S1R3r3.tab\n",
      "label:  S1R3r3\n",
      "['r3']\n",
      "rep #:  R3\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m0Macro/geneCounts/S1/m0Macro_S1R1r4.tab\n",
      "label:  S1R1r4\n",
      "['r4']\n",
      "rep #:  R1\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m0Macro/geneCounts/S2/m0Macro_S2R3.tab\n",
      "label:  S2R3\n",
      "[]\n",
      "rep #:  R3\n",
      "m0Macro_S2R3\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m0Macro/geneCounts/S2/m0Macro_S2R2.tab\n",
      "label:  S2R2\n",
      "[]\n",
      "rep #:  R2\n",
      "m0Macro_S2R2\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m0Macro/geneCounts/S2/m0Macro_S2R1.tab\n",
      "label:  S2R1\n",
      "[]\n",
      "rep #:  R1\n",
      "m0Macro_S2R1\n",
      "Preprocessing m1Macro\n",
      "Gene info output directory is \"/home/jupyteruser/work/data/results/m1Macro\"\n",
      "Looking for STAR gene count tables in '/home/jupyteruser/work/data/MADRID_input/m1Macro'\n",
      "Creating Counts Matrix for 'm1Macro'\n",
      "[1] \"Organizing Files\"\n",
      "[1] \"Creating counts matrix\"\n",
      "Count Matrix written at  /home/jupyteruser/work/data/data_matrices/m1Macro/gene_counts_matrix_full_m1Macro.csv \n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m1Macro/geneCounts/S1/m1Macro_S1R3r1.tab\n",
      "label:  S1R3r1\n",
      "['r1']\n",
      "rep #:  R3\n",
      "m1Macro_S1R3\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m1Macro/geneCounts/S1/m1Macro_S1R2r2.tab\n",
      "label:  S1R2r2\n",
      "['r2']\n",
      "rep #:  R2\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m1Macro/geneCounts/S1/m1Macro_S1R2r1.tab\n",
      "label:  S1R2r1\n",
      "['r1']\n",
      "rep #:  R2\n",
      "m1Macro_S1R2\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m1Macro/geneCounts/S1/m1Macro_S1R1r1.tab\n",
      "label:  S1R1r1\n",
      "['r1']\n",
      "rep #:  R1\n",
      "m1Macro_S1R1\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m1Macro/geneCounts/S1/m1Macro_S1R1r2.tab\n",
      "label:  S1R1r2\n",
      "['r2']\n",
      "rep #:  R1\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m1Macro/geneCounts/S1/m1Macro_S1R1r4.tab\n",
      "label:  S1R1r4\n",
      "['r4']\n",
      "rep #:  R1\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m1Macro/geneCounts/S1/m1Macro_S1R3r3.tab\n",
      "label:  S1R3r3\n",
      "['r3']\n",
      "rep #:  R3\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m1Macro/geneCounts/S1/m1Macro_S1R1r3.tab\n",
      "label:  S1R1r3\n",
      "['r3']\n",
      "rep #:  R1\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m1Macro/geneCounts/S1/m1Macro_S1R3r4.tab\n",
      "label:  S1R3r4\n",
      "['r4']\n",
      "rep #:  R3\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m1Macro/geneCounts/S1/m1Macro_S1R2r3.tab\n",
      "label:  S1R2r3\n",
      "['r3']\n",
      "rep #:  R2\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m1Macro/geneCounts/S1/m1Macro_S1R3r2.tab\n",
      "label:  S1R3r2\n",
      "['r2']\n",
      "rep #:  R3\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m1Macro/geneCounts/S1/m1Macro_S1R2r4.tab\n",
      "label:  S1R2r4\n",
      "['r4']\n",
      "rep #:  R2\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m1Macro/geneCounts/S2/m1Macro_S2R2.tab\n",
      "label:  S2R2\n",
      "[]\n",
      "rep #:  R2\n",
      "m1Macro_S2R2\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m1Macro/geneCounts/S2/m1Macro_S2R1.tab\n",
      "label:  S2R1\n",
      "[]\n",
      "rep #:  R1\n",
      "m1Macro_S2R1\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m1Macro/geneCounts/S3/m1Macro_S3R4.tab\n",
      "label:  S3R4\n",
      "[]\n",
      "rep #:  R4\n",
      "m1Macro_S3R4\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m1Macro/geneCounts/S3/m1Macro_S3R1.tab\n",
      "label:  S3R1\n",
      "[]\n",
      "rep #:  R1\n",
      "m1Macro_S3R1\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m1Macro/geneCounts/S3/m1Macro_S3R2.tab\n",
      "label:  S3R2\n",
      "[]\n",
      "rep #:  R2\n",
      "m1Macro_S3R2\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m1Macro/geneCounts/S3/m1Macro_S3R3.tab\n",
      "label:  S3R3\n",
      "[]\n",
      "rep #:  R3\n",
      "m1Macro_S3R3\n",
      "Preprocessing m2Macro\n",
      "Gene info output directory is \"/home/jupyteruser/work/data/results/m2Macro\"\n",
      "Looking for STAR gene count tables in '/home/jupyteruser/work/data/MADRID_input/m2Macro'\n",
      "Creating Counts Matrix for 'm2Macro'\n",
      "[1] \"Organizing Files\"\n",
      "[1] \"Creating counts matrix\"\n",
      "Count Matrix written at  /home/jupyteruser/work/data/data_matrices/m2Macro/gene_counts_matrix_full_m2Macro.csv \n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m2Macro/geneCounts/S1/m2Macro_S1R2r3.tab\n",
      "label:  S1R2r3\n",
      "['r3']\n",
      "rep #:  R2\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m2Macro/geneCounts/S1/m2Macro_S1R3r4.tab\n",
      "label:  S1R3r4\n",
      "['r4']\n",
      "rep #:  R3\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m2Macro/geneCounts/S1/m2Macro_S1R1r4.tab\n",
      "label:  S1R1r4\n",
      "['r4']\n",
      "rep #:  R1\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m2Macro/geneCounts/S1/m2Macro_S1R3r1.tab\n",
      "label:  S1R3r1\n",
      "['r1']\n",
      "rep #:  R3\n",
      "m2Macro_S1R3\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m2Macro/geneCounts/S1/m2Macro_S1R1r2.tab\n",
      "label:  S1R1r2\n",
      "['r2']\n",
      "rep #:  R1\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m2Macro/geneCounts/S1/m2Macro_S1R1r1.tab\n",
      "label:  S1R1r1\n",
      "['r1']\n",
      "rep #:  R1\n",
      "m2Macro_S1R1\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m2Macro/geneCounts/S1/m2Macro_S1R2r2.tab\n",
      "label:  S1R2r2\n",
      "['r2']\n",
      "rep #:  R2\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m2Macro/geneCounts/S1/m2Macro_S1R2r4.tab\n",
      "label:  S1R2r4\n",
      "['r4']\n",
      "rep #:  R2\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m2Macro/geneCounts/S1/m2Macro_S1R1r3.tab\n",
      "label:  S1R1r3\n",
      "['r3']\n",
      "rep #:  R1\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m2Macro/geneCounts/S1/m2Macro_S1R3r2.tab\n",
      "label:  S1R3r2\n",
      "['r2']\n",
      "rep #:  R3\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m2Macro/geneCounts/S1/m2Macro_S1R2r1.tab\n",
      "label:  S1R2r1\n",
      "['r1']\n",
      "rep #:  R2\n",
      "m2Macro_S1R2\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m2Macro/geneCounts/S1/m2Macro_S1R3r3.tab\n",
      "label:  S1R3r3\n",
      "['r3']\n",
      "rep #:  R3\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m2Macro/geneCounts/S2/m2Macro_S2R1.tab\n",
      "label:  S2R1\n",
      "[]\n",
      "rep #:  R1\n",
      "m2Macro_S2R1\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m2Macro/geneCounts/S2/m2Macro_S2R3.tab\n",
      "label:  S2R3\n",
      "[]\n",
      "rep #:  R3\n",
      "m2Macro_S2R3\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m2Macro/geneCounts/S2/m2Macro_S2R4.tab\n",
      "label:  S2R4\n",
      "[]\n",
      "rep #:  R4\n",
      "m2Macro_S2R4\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/m2Macro/geneCounts/S2/m2Macro_S2R2.tab\n",
      "label:  S2R2\n",
      "[]\n",
      "rep #:  R2\n",
      "m2Macro_S2R2\n",
      "Preprocessing naiveB\n",
      "Gene info output directory is \"/home/jupyteruser/work/data/results/naiveB\"\n",
      "Looking for STAR gene count tables in '/home/jupyteruser/work/data/MADRID_input/naiveB'\n",
      "Creating Counts Matrix for 'naiveB'\n",
      "[1] \"Organizing Files\"\n",
      "[1] \"Creating counts matrix\"\n",
      "Count Matrix written at  /home/jupyteruser/work/data/data_matrices/naiveB/gene_counts_matrix_full_naiveB.csv \n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/naiveB/geneCounts/S1/naiveB_S1R4.tab\n",
      "label:  S1R4\n",
      "[]\n",
      "rep #:  R4\n",
      "naiveB_S1R4\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/naiveB/geneCounts/S1/naiveB_S1R1.tab\n",
      "label:  S1R1\n",
      "[]\n",
      "rep #:  R1\n",
      "naiveB_S1R1\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/naiveB/geneCounts/S1/naiveB_S1R2.tab\n",
      "label:  S1R2\n",
      "[]\n",
      "rep #:  R2\n",
      "naiveB_S1R2\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/naiveB/geneCounts/S1/naiveB_S1R3.tab\n",
      "label:  S1R3\n",
      "[]\n",
      "rep #:  R3\n",
      "naiveB_S1R3\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/naiveB/geneCounts/S2/naiveB_S2R4.tab\n",
      "label:  S2R4\n",
      "[]\n",
      "rep #:  R4\n",
      "naiveB_S2R4\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/naiveB/geneCounts/S2/naiveB_S2R1.tab\n",
      "label:  S2R1\n",
      "[]\n",
      "rep #:  R1\n",
      "naiveB_S2R1\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/naiveB/geneCounts/S2/naiveB_S2R2.tab\n",
      "label:  S2R2\n",
      "[]\n",
      "rep #:  R2\n",
      "naiveB_S2R2\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/naiveB/geneCounts/S2/naiveB_S2R3.tab\n",
      "label:  S2R3\n",
      "[]\n",
      "rep #:  R3\n",
      "naiveB_S2R3\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/naiveB/geneCounts/S3/naiveB_S3R3.tab\n",
      "label:  S3R3\n",
      "[]\n",
      "rep #:  R3\n",
      "naiveB_S3R3\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/naiveB/geneCounts/S3/naiveB_S3R1.tab\n",
      "label:  S3R1\n",
      "[]\n",
      "rep #:  R1\n",
      "naiveB_S3R1\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/naiveB/geneCounts/S3/naiveB_S3R2.tab\n",
      "label:  S3R2\n",
      "[]\n",
      "rep #:  R2\n",
      "naiveB_S3R2\n",
      "Preprocessing nsmB\n",
      "Gene info output directory is \"/home/jupyteruser/work/data/results/nsmB\"\n",
      "Looking for STAR gene count tables in '/home/jupyteruser/work/data/MADRID_input/nsmB'\n",
      "Creating Counts Matrix for 'nsmB'\n",
      "[1] \"Organizing Files\"\n",
      "[1] \"Creating counts matrix\"\n",
      "Count Matrix written at  /home/jupyteruser/work/data/data_matrices/nsmB/gene_counts_matrix_full_nsmB.csv \n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/nsmB/geneCounts/S2/nsmB_S2R4.tab\n",
      "label:  S2R4\n",
      "[]\n",
      "rep #:  R4\n",
      "nsmB_S2R4\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/nsmB/geneCounts/S2/nsmB_S2R2.tab\n",
      "label:  S2R2\n",
      "[]\n",
      "rep #:  R2\n",
      "nsmB_S2R2\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/nsmB/geneCounts/S2/nsmB_S2R1.tab\n",
      "label:  S2R1\n",
      "[]\n",
      "rep #:  R1\n",
      "nsmB_S2R1\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/nsmB/geneCounts/S2/nsmB_S2R3.tab\n",
      "label:  S2R3\n",
      "[]\n",
      "rep #:  R3\n",
      "nsmB_S2R3\n",
      "Preprocessing smB\n",
      "Gene info output directory is \"/home/jupyteruser/work/data/results/smB\"\n",
      "Looking for STAR gene count tables in '/home/jupyteruser/work/data/MADRID_input/smB'\n",
      "Creating Counts Matrix for 'smB'\n",
      "[1] \"Organizing Files\"\n",
      "[1] \"Creating counts matrix\"\n",
      "Count Matrix written at  /home/jupyteruser/work/data/data_matrices/smB/gene_counts_matrix_full_smB.csv \n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/smB/geneCounts/S1/smB_S1R2.tab\n",
      "label:  S1R2\n",
      "[]\n",
      "rep #:  R2\n",
      "smB_S1R2\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/smB/geneCounts/S1/smB_S1R4.tab\n",
      "label:  S1R4\n",
      "[]\n",
      "rep #:  R4\n",
      "smB_S1R4\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/smB/geneCounts/S1/smB_S1R1.tab\n",
      "label:  S1R1\n",
      "[]\n",
      "rep #:  R1\n",
      "smB_S1R1\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/smB/geneCounts/S1/smB_S1R3.tab\n",
      "label:  S1R3\n",
      "[]\n",
      "rep #:  R3\n",
      "smB_S1R3\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/smB/geneCounts/S2/smB_S2R1.tab\n",
      "label:  S2R1\n",
      "[]\n",
      "rep #:  R1\n",
      "smB_S2R1\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/smB/geneCounts/S2/smB_S2R3.tab\n",
      "label:  S2R3\n",
      "[]\n",
      "rep #:  R3\n",
      "smB_S2R3\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/smB/geneCounts/S2/smB_S2R2.tab\n",
      "label:  S2R2\n",
      "[]\n",
      "rep #:  R2\n",
      "smB_S2R2\n",
      "Preprocessing plasmaB\n",
      "Gene info output directory is \"/home/jupyteruser/work/data/results/plasmaB\"\n",
      "Looking for STAR gene count tables in '/home/jupyteruser/work/data/MADRID_input/plasmaB'\n",
      "Creating Counts Matrix for 'plasmaB'\n",
      "[1] \"Organizing Files\"\n",
      "[1] \"Creating counts matrix\"\n",
      "Count Matrix written at  /home/jupyteruser/work/data/data_matrices/plasmaB/gene_counts_matrix_full_plasmaB.csv \n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/plasmaB/geneCounts/S1/plasmaB_S1R16.tab\n",
      "label:  S1R16\n",
      "[]\n",
      "rep #:  R16\n",
      "plasmaB_S1R16\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/plasmaB/geneCounts/S1/plasmaB_S1R5.tab\n",
      "label:  S1R5\n",
      "[]\n",
      "rep #:  R5\n",
      "plasmaB_S1R5\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/plasmaB/geneCounts/S1/plasmaB_S1R6.tab\n",
      "label:  S1R6\n",
      "[]\n",
      "rep #:  R6\n",
      "plasmaB_S1R6\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/plasmaB/geneCounts/S1/plasmaB_S1R14.tab\n",
      "label:  S1R14\n",
      "[]\n",
      "rep #:  R14\n",
      "plasmaB_S1R14\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/plasmaB/geneCounts/S1/plasmaB_S1R10.tab\n",
      "label:  S1R10\n",
      "[]\n",
      "rep #:  R10\n",
      "plasmaB_S1R10\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/plasmaB/geneCounts/S1/plasmaB_S1R17.tab\n",
      "label:  S1R17\n",
      "[]\n",
      "rep #:  R17\n",
      "plasmaB_S1R17\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/plasmaB/geneCounts/S1/plasmaB_S1R9.tab\n",
      "label:  S1R9\n",
      "[]\n",
      "rep #:  R9\n",
      "plasmaB_S1R9\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/plasmaB/geneCounts/S1/plasmaB_S1R15.tab\n",
      "label:  S1R15\n",
      "[]\n",
      "rep #:  R15\n",
      "plasmaB_S1R15\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/plasmaB/geneCounts/S1/plasmaB_S1R2.tab\n",
      "label:  S1R2\n",
      "[]\n",
      "rep #:  R2\n",
      "plasmaB_S1R2\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/plasmaB/geneCounts/S1/plasmaB_S1R1.tab\n",
      "label:  S1R1\n",
      "[]\n",
      "rep #:  R1\n",
      "plasmaB_S1R1\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/plasmaB/geneCounts/S1/plasmaB_S1R7.tab\n",
      "label:  S1R7\n",
      "[]\n",
      "rep #:  R7\n",
      "plasmaB_S1R7\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/plasmaB/geneCounts/S1/plasmaB_S1R11.tab\n",
      "label:  S1R11\n",
      "[]\n",
      "rep #:  R11\n",
      "plasmaB_S1R11\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/plasmaB/geneCounts/S1/plasmaB_S1R13.tab\n",
      "label:  S1R13\n",
      "[]\n",
      "rep #:  R13\n",
      "plasmaB_S1R13\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/plasmaB/geneCounts/S1/plasmaB_S1R3.tab\n",
      "label:  S1R3\n",
      "[]\n",
      "rep #:  R3\n",
      "plasmaB_S1R3\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/plasmaB/geneCounts/S1/plasmaB_S1R12.tab\n",
      "label:  S1R12\n",
      "[]\n",
      "rep #:  R12\n",
      "plasmaB_S1R12\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/plasmaB/geneCounts/S1/plasmaB_S1R8.tab\n",
      "label:  S1R8\n",
      "[]\n",
      "rep #:  R8\n",
      "plasmaB_S1R8\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/plasmaB/geneCounts/S1/plasmaB_S1R4.tab\n",
      "label:  S1R4\n",
      "[]\n",
      "rep #:  R4\n",
      "plasmaB_S1R4\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/plasmaB/geneCounts/S2/plasmaB_S2R3.tab\n",
      "label:  S2R3\n",
      "[]\n",
      "rep #:  R3\n",
      "plasmaB_S2R3\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/plasmaB/geneCounts/S2/plasmaB_S2R1.tab\n",
      "label:  S2R1\n",
      "[]\n",
      "rep #:  R1\n",
      "plasmaB_S2R1\n",
      "next\n",
      "/home/jupyteruser/work/data/MADRID_input/plasmaB/geneCounts/S2/plasmaB_S2R2.tab\n",
      "label:  S2R2\n",
      "[]\n",
      "rep #:  R2\n",
      "plasmaB_S2R2\n",
      "Fetching gene info\n",
      "Total Genes to Retrieve: 61541\n",
      "retrieve 0:300\n",
      "retrieve 300:600\n",
      "retrieve 600:900\n",
      "retrieve 900:1200\n",
      "retrieve 1200:1500\n",
      "retrieve 1500:1800\n",
      "retrieve 1800:2100\n",
      "retrieve 2100:2400\n",
      "retrieve 2400:2700\n",
      "retrieve 2700:3000\n",
      "retrieve 3000:3300\n",
      "retrieve 3300:3600\n",
      "retrieve 3600:3900\n",
      "retrieve 3900:4200\n",
      "retrieve 4200:4500\n",
      "retrieve 4500:4800\n",
      "retrieve 4800:5100\n",
      "retrieve 5100:5400\n",
      "retrieve 5400:5700\n",
      "retrieve 5700:6000\n",
      "retrieve 6000:6300\n",
      "retrieve 6300:6600\n",
      "retrieve 6600:6900\n",
      "retrieve 6900:7200\n",
      "retrieve 7200:7500\n",
      "retrieve 7500:7800\n",
      "retrieve 7800:8100\n",
      "retrieve 8100:8400\n",
      "retrieve 8400:8700\n",
      "retrieve 8700:9000\n",
      "retrieve 9000:9300\n",
      "retrieve 9300:9600\n",
      "retrieve 9600:9900\n",
      "retrieve 9900:10200\n",
      "retrieve 10200:10500\n",
      "retrieve 10500:10800\n",
      "retrieve 10800:11100\n",
      "retrieve 11100:11400\n",
      "retrieve 11400:11700\n",
      "retrieve 11700:12000\n",
      "retrieve 12000:12300\n",
      "retrieve 12300:12600\n",
      "retrieve 12600:12900\n",
      "retrieve 12900:13200\n",
      "retrieve 13200:13500\n",
      "retrieve 13500:13800\n",
      "retrieve 13800:14100\n",
      "retrieve 14100:14400\n",
      "retrieve 14400:14700\n",
      "retrieve 14700:15000\n",
      "retrieve 15000:15300\n",
      "retrieve 15300:15600\n",
      "retrieve 15600:15900\n",
      "retrieve 15900:16200\n",
      "retrieve 16200:16500\n",
      "retrieve 16500:16800\n",
      "retrieve 16800:17100\n",
      "retrieve 17100:17400\n",
      "retrieve 17400:17700\n",
      "retrieve 17700:18000\n",
      "retrieve 18000:18300\n",
      "retrieve 18300:18600\n",
      "retrieve 18600:18900\n",
      "retrieve 18900:19200\n",
      "retrieve 19200:19500\n",
      "retrieve 19500:19800\n",
      "retrieve 19800:20100\n",
      "retrieve 20100:20400\n",
      "retrieve 20400:20700\n",
      "retrieve 20700:21000\n",
      "retrieve 21000:21300\n",
      "retrieve 21300:21600\n",
      "retrieve 21600:21900\n",
      "retrieve 21900:22200\n",
      "retrieve 22200:22500\n",
      "retrieve 22500:22800\n",
      "retrieve 22800:23100\n",
      "retrieve 23100:23400\n",
      "retrieve 23400:23700\n",
      "retrieve 23700:24000\n",
      "retrieve 24000:24300\n",
      "retrieve 24300:24600\n",
      "retrieve 24600:24900\n",
      "retrieve 24900:25200\n",
      "retrieve 25200:25500\n",
      "retrieve 25500:25800\n",
      "retrieve 25800:26100\n",
      "retrieve 26100:26400\n",
      "retrieve 26400:26700\n",
      "retrieve 26700:27000\n",
      "retrieve 27000:27300\n",
      "retrieve 27300:27600\n",
      "retrieve 27600:27900\n",
      "retrieve 27900:28200\n",
      "retrieve 28200:28500\n",
      "retrieve 28500:28800\n",
      "retrieve 28800:29100\n",
      "retrieve 29100:29400\n",
      "retrieve 29400:29700\n",
      "retrieve 29700:30000\n",
      "retrieve 30000:30300\n",
      "retrieve 30300:30600\n",
      "retrieve 30600:30900\n",
      "retrieve 30900:31200\n",
      "retrieve 31200:31500\n",
      "retrieve 31500:31800\n",
      "retrieve 31800:32100\n",
      "retrieve 32100:32400\n",
      "retrieve 32400:32700\n",
      "retrieve 32700:33000\n",
      "retrieve 33000:33300\n",
      "retrieve 33300:33600\n",
      "retrieve 33600:33900\n",
      "retrieve 33900:34200\n",
      "retrieve 34200:34500\n",
      "retrieve 34500:34800\n",
      "retrieve 34800:35100\n",
      "retrieve 35100:35400\n",
      "retrieve 35400:35700\n",
      "retrieve 35700:36000\n",
      "retrieve 36000:36300\n",
      "retrieve 36300:36600\n",
      "retrieve 36600:36900\n",
      "retrieve 36900:37200\n",
      "retrieve 37200:37500\n",
      "retrieve 37500:37800\n",
      "retrieve 37800:38100\n",
      "retrieve 38100:38400\n",
      "retrieve 38400:38700\n",
      "retrieve 38700:39000\n",
      "retrieve 39000:39300\n",
      "retrieve 39300:39600\n",
      "retrieve 39600:39900\n",
      "retrieve 39900:40200\n",
      "retrieve 40200:40500\n",
      "retrieve 40500:40800\n",
      "retrieve 40800:41100\n",
      "retrieve 41100:41400\n",
      "retrieve 41400:41700\n",
      "retrieve 41700:42000\n",
      "retrieve 42000:42300\n",
      "retrieve 42300:42600\n",
      "retrieve 42600:42900\n",
      "retrieve 42900:43200\n",
      "retrieve 43200:43500\n",
      "retrieve 43500:43800\n",
      "retrieve 43800:44100\n",
      "retrieve 44100:44400\n",
      "retrieve 44400:44700\n",
      "retrieve 44700:45000\n",
      "retrieve 45000:45300\n",
      "retrieve 45300:45600\n",
      "retrieve 45600:45900\n",
      "retrieve 45900:46200\n",
      "retrieve 46200:46500\n",
      "retrieve 46500:46800\n",
      "retrieve 46800:47100\n",
      "retrieve 47100:47400\n",
      "retrieve 47400:47700\n",
      "retrieve 47700:48000\n",
      "retrieve 48000:48300\n",
      "retrieve 48300:48600\n",
      "retrieve 48600:48900\n",
      "retrieve 48900:49200\n",
      "retrieve 49200:49500\n",
      "retrieve 49500:49800\n",
      "retrieve 49800:50100\n",
      "retrieve 50100:50400\n",
      "retrieve 50400:50700\n",
      "retrieve 50700:51000\n",
      "retrieve 51000:51300\n",
      "retrieve 51300:51600\n",
      "retrieve 51600:51900\n",
      "retrieve 51900:52200\n",
      "retrieve 52200:52500\n",
      "retrieve 52500:52800\n",
      "retrieve 52800:53100\n",
      "retrieve 53100:53400\n",
      "retrieve 53400:53700\n",
      "retrieve 53700:54000\n",
      "retrieve 54000:54300\n",
      "retrieve 54300:54600\n",
      "retrieve 54600:54900\n",
      "retrieve 54900:55200\n",
      "retrieve 55200:55500\n",
      "retrieve 55500:55800\n",
      "retrieve 55800:56100\n",
      "retrieve 56100:56400\n",
      "retrieve 56400:56700\n",
      "retrieve 56700:57000\n",
      "retrieve 57000:57300\n",
      "retrieve 57300:57600\n",
      "retrieve 57600:57900\n",
      "retrieve 57900:58200\n",
      "retrieve 58200:58500\n",
      "retrieve 58500:58800\n",
      "retrieve 58800:59100\n",
      "retrieve 59100:59400\n",
      "retrieve 59400:59700\n",
      "retrieve 59700:60000\n",
      "retrieve 60000:60300\n",
      "retrieve 60300:60600\n",
      "retrieve 60600:60900\n",
      "retrieve 60900:61200\n",
      "retrieve 61200:61500\n",
      "retrieve 61500:61541\n",
      "Gene Info file written at '/home/jupyteruser/work/data/gene_info.csv'\n"
     ]
    }
   ],
   "source": [
    "# Step 1: Preprocess RNAseq data by generating a counts matrix, config sheet, and gene file from MADRID_inputs.\n",
    "\n",
    "# tissue name or list of tissue names within a string\n",
    "context_names = \"['m0Macro', 'm1Macro', 'm2Macro', 'naiveB', 'nsmB', 'smB', 'plasmaB']\"\n",
    "create_counts_matrix = True # set to false if using a pregenerated matrix file\n",
    "gene_format = \"Ensembl\"     # accepts 'Entrez', 'Ensembl', and 'Symbol'\n",
    "taxon_id = \"human\" # accepts integer (bioDBnet taxon id) or 'human' or 'mouse'\n",
    "preprocess_mode = \"create-matrix\" # \"create-matrix\" or \"provide-matrix\"\n",
    "    \n",
    "cmd = ' '.join(['python3', 'rnaseq_preprocess.py',\n",
    "                '--context-names', '\"{}\"'.format(context_names),\n",
    "                '--gene-format', '\"{}\"'.format(gene_format),\n",
    "                '--taxon-id', '\"{}\"'.format(taxon_id),\n",
    "                '--{}'.format(preprocess_mode)])\n",
    "!{cmd}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2: Identifying Gene Activity in Transcriptomics and Proteomics Datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Identify gene activity in the following data sources \n",
    " - RNA-seq (bulk or single-cell)\n",
    " - Proteomics\n",
    " - Microarray\n",
    "\n",
    " Only one sources is required for model generation, multiple can be helpful for additional validation if of high-quality. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Microarrays"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From wikipedia: A microarray is a multiplex lab-on-a-chip. Its purpose is to simultaneously detect the expression of thousands of genes from a sample (e.g. from a tissue). It is a two-dimensional array on a solid substrate—usually a glass slide or silicon thin-film cell—that assays (tests) large amounts of biological material using high-throughput screening miniaturized, multiplexed and parallel processing and detection methods.\n",
    "\n",
    "MADRID can directly download and analyze microarray data from GEO for Agilent and Affymetrix platoforms. Follow the template and example in `/work/data/config_sheets/` to use microarray data in your analysis.\n",
    "\n",
    "Microarray technology is becoming increasingly obsolete, if possible it is recommended that you use RNA-seq instead. Although different strategies exist for microarrays, MADRID does not distinguish between them nor are  there any plans to in the future due to it's obsoelecense."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Input file is  microarray_data_inputs.xlsx\n",
      "Expression Proportion for Gene Expression is  0.5\n",
      "Top proportion for high-confidence genes is  0.9\n",
      "---\n",
      "Start Collecting Data for:\n",
      "['GSE22886' 'GSE43005' 'GSE22045' 'GSE24634']\n",
      "['GSM565273' 'GSM565274' 'GSM565275' 'GSM565290' 'GSM565291' 'GSM565292'\n",
      " 'GSM1054773' 'GSM1054779' 'GSM1054781' 'GSM1054789' 'GSM548000'\n",
      " 'GSM548001' 'GSM607510' 'GSM607511' 'GSM607512']\n",
      "---\n",
      "\n",
      "Initialize project (GSE22886):\n",
      "Root: /home/jupyteruser/work\n",
      "Raw data: /home/jupyteruser/work/data/GSE22886_RAW\n",
      "Retrieve Sample: /home/jupyteruser/work/data/GSE22886_RAW/GSM565273.tar\n",
      "Extract to: /home/jupyteruser/work/data/GSE22886_RAW/GPL96\n",
      "Retrieve Sample: /home/jupyteruser/work/data/GSE22886_RAW/GSM565274.tar\n",
      "Extract to: /home/jupyteruser/work/data/GSE22886_RAW/GPL96\n",
      "Retrieve Sample: /home/jupyteruser/work/data/GSE22886_RAW/GSM565275.tar\n",
      "Extract to: /home/jupyteruser/work/data/GSE22886_RAW/GPL96\n",
      "Retrieve Sample: /home/jupyteruser/work/data/GSE22886_RAW/GSM565290.tar\n",
      "Extract to: /home/jupyteruser/work/data/GSE22886_RAW/GPL96\n",
      "Retrieve Sample: /home/jupyteruser/work/data/GSE22886_RAW/GSM565291.tar\n",
      "Extract to: /home/jupyteruser/work/data/GSE22886_RAW/GPL96\n",
      "Retrieve Sample: /home/jupyteruser/work/data/GSE22886_RAW/GSM565292.tar\n",
      "Extract to: /home/jupyteruser/work/data/GSE22886_RAW/GPL96\n",
      "Retrieve Samples Completed.\n",
      "Create new table: /home/jupyteruser/work/data/GSE22886_RAW/GSE22886_sc500_full_table.csv\n",
      "03-Mar-2022 18:26:33 DEBUG utils - Directory ./ already exists. Skipping.\n",
      "03-Mar-2022 18:26:33 INFO GEOparse - Downloading ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE22nnn/GSE22886/soft/GSE22886_family.soft.gz to ./GSE22886_family.soft.gz\n",
      "100%|██████████████████████████████████████| 41.1M/41.1M [00:00<00:00, 43.3MB/s]\n",
      "03-Mar-2022 18:26:35 DEBUG downloader - Size validation passed\n",
      "03-Mar-2022 18:26:35 DEBUG downloader - Moving /tmp/tmp80nq4fa6 to /home/jupyteruser/work/py/GSE22886_family.soft.gz\n",
      "03-Mar-2022 18:26:35 DEBUG downloader - Successfully downloaded ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE22nnn/GSE22886/soft/GSE22886_family.soft.gz\n",
      "03-Mar-2022 18:26:35 INFO GEOparse - Parsing ./GSE22886_family.soft.gz: \n",
      "03-Mar-2022 18:26:35 DEBUG GEOparse - DATABASE: GeoMiame\n",
      "03-Mar-2022 18:26:35 DEBUG GEOparse - SERIES: GSE22886\n",
      "03-Mar-2022 18:26:35 DEBUG GEOparse - PLATFORM: GPL96\n",
      "03-Mar-2022 18:26:36 DEBUG GEOparse - PLATFORM: GPL97\n",
      "03-Mar-2022 18:26:36 DEBUG GEOparse - SAMPLE: GSM565269\n",
      "03-Mar-2022 18:26:36 DEBUG GEOparse - SAMPLE: GSM565270\n",
      "03-Mar-2022 18:26:36 DEBUG GEOparse - SAMPLE: GSM565271\n",
      "03-Mar-2022 18:26:37 DEBUG GEOparse - SAMPLE: GSM565272\n",
      "03-Mar-2022 18:26:37 DEBUG GEOparse - SAMPLE: GSM565273\n",
      "03-Mar-2022 18:26:37 DEBUG GEOparse - SAMPLE: GSM565274\n",
      "03-Mar-2022 18:26:37 DEBUG GEOparse - SAMPLE: GSM565275\n",
      "03-Mar-2022 18:26:37 DEBUG GEOparse - SAMPLE: GSM565276\n",
      "03-Mar-2022 18:26:37 DEBUG GEOparse - SAMPLE: GSM565277\n",
      "03-Mar-2022 18:26:37 DEBUG GEOparse - SAMPLE: GSM565278\n",
      "03-Mar-2022 18:26:37 DEBUG GEOparse - SAMPLE: GSM565279\n",
      "03-Mar-2022 18:26:37 DEBUG GEOparse - SAMPLE: GSM565280\n",
      "03-Mar-2022 18:26:37 DEBUG GEOparse - SAMPLE: GSM565281\n",
      "03-Mar-2022 18:26:37 DEBUG GEOparse - SAMPLE: GSM565282\n",
      "03-Mar-2022 18:26:37 DEBUG GEOparse - SAMPLE: GSM565283\n",
      "03-Mar-2022 18:26:37 DEBUG GEOparse - SAMPLE: GSM565284\n",
      "03-Mar-2022 18:26:37 DEBUG GEOparse - SAMPLE: GSM565285\n",
      "03-Mar-2022 18:26:37 DEBUG GEOparse - SAMPLE: GSM565286\n",
      "03-Mar-2022 18:26:37 DEBUG GEOparse - SAMPLE: GSM565287\n",
      "03-Mar-2022 18:26:38 DEBUG GEOparse - SAMPLE: GSM565288\n",
      "03-Mar-2022 18:26:38 DEBUG GEOparse - SAMPLE: GSM565289\n",
      "03-Mar-2022 18:26:38 DEBUG GEOparse - SAMPLE: GSM565290\n",
      "03-Mar-2022 18:26:38 DEBUG GEOparse - SAMPLE: GSM565291\n",
      "03-Mar-2022 18:26:38 DEBUG GEOparse - SAMPLE: GSM565292\n",
      "03-Mar-2022 18:26:38 DEBUG GEOparse - SAMPLE: GSM565293\n",
      "03-Mar-2022 18:26:38 DEBUG GEOparse - SAMPLE: GSM565294\n",
      "03-Mar-2022 18:26:38 DEBUG GEOparse - SAMPLE: GSM565295\n",
      "03-Mar-2022 18:26:38 DEBUG GEOparse - SAMPLE: GSM565296\n",
      "03-Mar-2022 18:26:38 DEBUG GEOparse - SAMPLE: GSM565297\n",
      "03-Mar-2022 18:26:38 DEBUG GEOparse - SAMPLE: GSM565298\n",
      "03-Mar-2022 18:26:38 DEBUG GEOparse - SAMPLE: GSM565299\n",
      "03-Mar-2022 18:26:38 DEBUG GEOparse - SAMPLE: GSM565300\n",
      "03-Mar-2022 18:26:38 DEBUG GEOparse - SAMPLE: GSM565301\n",
      "03-Mar-2022 18:26:38 DEBUG GEOparse - SAMPLE: GSM565302\n",
      "03-Mar-2022 18:26:38 DEBUG GEOparse - SAMPLE: GSM565303\n",
      "03-Mar-2022 18:26:38 DEBUG GEOparse - SAMPLE: GSM565304\n",
      "03-Mar-2022 18:26:38 DEBUG GEOparse - SAMPLE: GSM565305\n",
      "03-Mar-2022 18:26:38 DEBUG GEOparse - SAMPLE: GSM565306\n",
      "03-Mar-2022 18:26:39 DEBUG GEOparse - SAMPLE: GSM565307\n",
      "03-Mar-2022 18:26:39 DEBUG GEOparse - SAMPLE: GSM565308\n",
      "03-Mar-2022 18:26:39 DEBUG GEOparse - SAMPLE: GSM565309\n",
      "03-Mar-2022 18:26:39 DEBUG GEOparse - SAMPLE: GSM565310\n",
      "03-Mar-2022 18:26:39 DEBUG GEOparse - SAMPLE: GSM565311\n",
      "03-Mar-2022 18:26:39 DEBUG GEOparse - SAMPLE: GSM565312\n",
      "03-Mar-2022 18:26:39 DEBUG GEOparse - SAMPLE: GSM565313\n",
      "03-Mar-2022 18:26:39 DEBUG GEOparse - SAMPLE: GSM565314\n",
      "03-Mar-2022 18:26:39 DEBUG GEOparse - SAMPLE: GSM565315\n",
      "03-Mar-2022 18:26:39 DEBUG GEOparse - SAMPLE: GSM565316\n",
      "03-Mar-2022 18:26:39 DEBUG GEOparse - SAMPLE: GSM565317\n",
      "03-Mar-2022 18:26:39 DEBUG GEOparse - SAMPLE: GSM565318\n",
      "03-Mar-2022 18:26:39 DEBUG GEOparse - SAMPLE: GSM565319\n",
      "03-Mar-2022 18:26:39 DEBUG GEOparse - SAMPLE: GSM565320\n",
      "03-Mar-2022 18:26:39 DEBUG GEOparse - SAMPLE: GSM565321\n",
      "03-Mar-2022 18:26:39 DEBUG GEOparse - SAMPLE: GSM565322\n",
      "03-Mar-2022 18:26:39 DEBUG GEOparse - SAMPLE: GSM565323\n",
      "03-Mar-2022 18:26:39 DEBUG GEOparse - SAMPLE: GSM565324\n",
      "03-Mar-2022 18:26:40 DEBUG GEOparse - SAMPLE: GSM565325\n",
      "03-Mar-2022 18:26:40 DEBUG GEOparse - SAMPLE: GSM565326\n",
      "03-Mar-2022 18:26:40 DEBUG GEOparse - SAMPLE: GSM565327\n",
      "03-Mar-2022 18:26:40 DEBUG GEOparse - SAMPLE: GSM565328\n",
      "03-Mar-2022 18:26:40 DEBUG GEOparse - SAMPLE: GSM565329\n",
      "03-Mar-2022 18:26:40 DEBUG GEOparse - SAMPLE: GSM565330\n",
      "03-Mar-2022 18:26:40 DEBUG GEOparse - SAMPLE: GSM565331\n",
      "03-Mar-2022 18:26:40 DEBUG GEOparse - SAMPLE: GSM565332\n",
      "03-Mar-2022 18:26:40 DEBUG GEOparse - SAMPLE: GSM565333\n",
      "03-Mar-2022 18:26:40 DEBUG GEOparse - SAMPLE: GSM565334\n",
      "03-Mar-2022 18:26:40 DEBUG GEOparse - SAMPLE: GSM565335\n",
      "03-Mar-2022 18:26:40 DEBUG GEOparse - SAMPLE: GSM565336\n",
      "03-Mar-2022 18:26:40 DEBUG GEOparse - SAMPLE: GSM565337\n",
      "03-Mar-2022 18:26:40 DEBUG GEOparse - SAMPLE: GSM565338\n",
      "03-Mar-2022 18:26:40 DEBUG GEOparse - SAMPLE: GSM565339\n",
      "03-Mar-2022 18:26:40 DEBUG GEOparse - SAMPLE: GSM565340\n",
      "03-Mar-2022 18:26:40 DEBUG GEOparse - SAMPLE: GSM565341\n",
      "03-Mar-2022 18:26:40 DEBUG GEOparse - SAMPLE: GSM565342\n",
      "03-Mar-2022 18:26:40 DEBUG GEOparse - SAMPLE: GSM565343\n",
      "03-Mar-2022 18:26:41 DEBUG GEOparse - SAMPLE: GSM565344\n",
      "03-Mar-2022 18:26:41 DEBUG GEOparse - SAMPLE: GSM565345\n",
      "03-Mar-2022 18:26:41 DEBUG GEOparse - SAMPLE: GSM565346\n",
      "03-Mar-2022 18:26:41 DEBUG GEOparse - SAMPLE: GSM565347\n",
      "03-Mar-2022 18:26:41 DEBUG GEOparse - SAMPLE: GSM565348\n",
      "03-Mar-2022 18:26:41 DEBUG GEOparse - SAMPLE: GSM565349\n",
      "03-Mar-2022 18:26:41 DEBUG GEOparse - SAMPLE: GSM565350\n",
      "03-Mar-2022 18:26:41 DEBUG GEOparse - SAMPLE: GSM565351\n",
      "03-Mar-2022 18:26:41 DEBUG GEOparse - SAMPLE: GSM565352\n",
      "03-Mar-2022 18:26:41 DEBUG GEOparse - SAMPLE: GSM565353\n",
      "03-Mar-2022 18:26:41 DEBUG GEOparse - SAMPLE: GSM565354\n",
      "03-Mar-2022 18:26:41 DEBUG GEOparse - SAMPLE: GSM565355\n",
      "03-Mar-2022 18:26:41 DEBUG GEOparse - SAMPLE: GSM565356\n",
      "03-Mar-2022 18:26:41 DEBUG GEOparse - SAMPLE: GSM565357\n",
      "03-Mar-2022 18:26:41 DEBUG GEOparse - SAMPLE: GSM565358\n",
      "03-Mar-2022 18:26:41 DEBUG GEOparse - SAMPLE: GSM565359\n",
      "03-Mar-2022 18:26:41 DEBUG GEOparse - SAMPLE: GSM565360\n",
      "03-Mar-2022 18:26:41 DEBUG GEOparse - SAMPLE: GSM565361\n",
      "03-Mar-2022 18:26:41 DEBUG GEOparse - SAMPLE: GSM565362\n",
      "03-Mar-2022 18:26:42 DEBUG GEOparse - SAMPLE: GSM565363\n",
      "03-Mar-2022 18:26:42 DEBUG GEOparse - SAMPLE: GSM565364\n",
      "03-Mar-2022 18:26:42 DEBUG GEOparse - SAMPLE: GSM565365\n",
      "03-Mar-2022 18:26:42 DEBUG GEOparse - SAMPLE: GSM565366\n",
      "03-Mar-2022 18:26:42 DEBUG GEOparse - SAMPLE: GSM565367\n",
      "03-Mar-2022 18:26:42 DEBUG GEOparse - SAMPLE: GSM565368\n",
      "03-Mar-2022 18:26:42 DEBUG GEOparse - SAMPLE: GSM565369\n",
      "03-Mar-2022 18:26:42 DEBUG GEOparse - SAMPLE: GSM565370\n",
      "03-Mar-2022 18:26:42 DEBUG GEOparse - SAMPLE: GSM565371\n",
      "03-Mar-2022 18:26:42 DEBUG GEOparse - SAMPLE: GSM565372\n",
      "03-Mar-2022 18:26:42 DEBUG GEOparse - SAMPLE: GSM565373\n",
      "03-Mar-2022 18:26:42 DEBUG GEOparse - SAMPLE: GSM565374\n",
      "03-Mar-2022 18:26:42 DEBUG GEOparse - SAMPLE: GSM565375\n",
      "03-Mar-2022 18:26:42 DEBUG GEOparse - SAMPLE: GSM565376\n",
      "03-Mar-2022 18:26:42 DEBUG GEOparse - SAMPLE: GSM565377\n",
      "03-Mar-2022 18:26:42 DEBUG GEOparse - SAMPLE: GSM565378\n",
      "03-Mar-2022 18:26:42 DEBUG GEOparse - SAMPLE: GSM565379\n",
      "03-Mar-2022 18:26:42 DEBUG GEOparse - SAMPLE: GSM565380\n",
      "03-Mar-2022 18:26:43 DEBUG GEOparse - SAMPLE: GSM565381\n",
      "03-Mar-2022 18:26:43 DEBUG GEOparse - SAMPLE: GSM565382\n",
      "03-Mar-2022 18:26:43 DEBUG GEOparse - SAMPLE: GSM565986\n",
      "03-Mar-2022 18:26:43 DEBUG GEOparse - SAMPLE: GSM565987\n",
      "03-Mar-2022 18:26:43 DEBUG GEOparse - SAMPLE: GSM565988\n",
      "03-Mar-2022 18:26:43 DEBUG GEOparse - SAMPLE: GSM565989\n",
      "03-Mar-2022 18:26:43 DEBUG GEOparse - SAMPLE: GSM565990\n",
      "03-Mar-2022 18:26:43 DEBUG GEOparse - SAMPLE: GSM565991\n",
      "03-Mar-2022 18:26:43 DEBUG GEOparse - SAMPLE: GSM565992\n",
      "03-Mar-2022 18:26:43 DEBUG GEOparse - SAMPLE: GSM565993\n",
      "03-Mar-2022 18:26:43 DEBUG GEOparse - SAMPLE: GSM565994\n",
      "03-Mar-2022 18:26:43 DEBUG GEOparse - SAMPLE: GSM565995\n",
      "03-Mar-2022 18:26:43 DEBUG GEOparse - SAMPLE: GSM565996\n",
      "03-Mar-2022 18:26:43 DEBUG GEOparse - SAMPLE: GSM565997\n",
      "03-Mar-2022 18:26:43 DEBUG GEOparse - SAMPLE: GSM565998\n",
      "03-Mar-2022 18:26:43 DEBUG GEOparse - SAMPLE: GSM565999\n",
      "03-Mar-2022 18:26:43 DEBUG GEOparse - SAMPLE: GSM566000\n",
      "03-Mar-2022 18:26:43 DEBUG GEOparse - SAMPLE: GSM566001\n",
      "03-Mar-2022 18:26:43 DEBUG GEOparse - SAMPLE: GSM566002\n",
      "03-Mar-2022 18:26:44 DEBUG GEOparse - SAMPLE: GSM566003\n",
      "03-Mar-2022 18:26:44 DEBUG GEOparse - SAMPLE: GSM566004\n",
      "03-Mar-2022 18:26:44 DEBUG GEOparse - SAMPLE: GSM566005\n",
      "03-Mar-2022 18:26:44 DEBUG GEOparse - SAMPLE: GSM566006\n",
      "03-Mar-2022 18:26:44 DEBUG GEOparse - SAMPLE: GSM566007\n",
      "03-Mar-2022 18:26:44 DEBUG GEOparse - SAMPLE: GSM566008\n",
      "03-Mar-2022 18:26:44 DEBUG GEOparse - SAMPLE: GSM566009\n",
      "03-Mar-2022 18:26:44 DEBUG GEOparse - SAMPLE: GSM566010\n",
      "03-Mar-2022 18:26:44 DEBUG GEOparse - SAMPLE: GSM566011\n",
      "03-Mar-2022 18:26:44 DEBUG GEOparse - SAMPLE: GSM566012\n",
      "03-Mar-2022 18:26:44 DEBUG GEOparse - SAMPLE: GSM566013\n",
      "03-Mar-2022 18:26:44 DEBUG GEOparse - SAMPLE: GSM566014\n",
      "03-Mar-2022 18:26:44 DEBUG GEOparse - SAMPLE: GSM566015\n",
      "03-Mar-2022 18:26:44 DEBUG GEOparse - SAMPLE: GSM566016\n",
      "03-Mar-2022 18:26:44 DEBUG GEOparse - SAMPLE: GSM566017\n",
      "03-Mar-2022 18:26:44 DEBUG GEOparse - SAMPLE: GSM566018\n",
      "03-Mar-2022 18:26:44 DEBUG GEOparse - SAMPLE: GSM566019\n",
      "03-Mar-2022 18:26:44 DEBUG GEOparse - SAMPLE: GSM566020\n",
      "03-Mar-2022 18:26:45 DEBUG GEOparse - SAMPLE: GSM566021\n",
      "03-Mar-2022 18:26:45 DEBUG GEOparse - SAMPLE: GSM566022\n",
      "03-Mar-2022 18:26:45 DEBUG GEOparse - SAMPLE: GSM566023\n",
      "03-Mar-2022 18:26:45 DEBUG GEOparse - SAMPLE: GSM566024\n",
      "03-Mar-2022 18:26:45 DEBUG GEOparse - SAMPLE: GSM566025\n",
      "03-Mar-2022 18:26:45 DEBUG GEOparse - SAMPLE: GSM566026\n",
      "03-Mar-2022 18:26:45 DEBUG GEOparse - SAMPLE: GSM566027\n",
      "03-Mar-2022 18:26:45 DEBUG GEOparse - SAMPLE: GSM566028\n",
      "03-Mar-2022 18:26:45 DEBUG GEOparse - SAMPLE: GSM566029\n",
      "03-Mar-2022 18:26:45 DEBUG GEOparse - SAMPLE: GSM566030\n",
      "03-Mar-2022 18:26:45 DEBUG GEOparse - SAMPLE: GSM566031\n",
      "03-Mar-2022 18:26:45 DEBUG GEOparse - SAMPLE: GSM566032\n",
      "03-Mar-2022 18:26:45 DEBUG GEOparse - SAMPLE: GSM566033\n",
      "03-Mar-2022 18:26:45 DEBUG GEOparse - SAMPLE: GSM566034\n",
      "03-Mar-2022 18:26:45 DEBUG GEOparse - SAMPLE: GSM566035\n",
      "03-Mar-2022 18:26:45 DEBUG GEOparse - SAMPLE: GSM566036\n",
      "03-Mar-2022 18:26:45 DEBUG GEOparse - SAMPLE: GSM566037\n",
      "03-Mar-2022 18:26:45 DEBUG GEOparse - SAMPLE: GSM566038\n",
      "03-Mar-2022 18:26:46 DEBUG GEOparse - SAMPLE: GSM566039\n",
      "03-Mar-2022 18:26:46 DEBUG GEOparse - SAMPLE: GSM566040\n",
      "03-Mar-2022 18:26:46 DEBUG GEOparse - SAMPLE: GSM566041\n",
      "03-Mar-2022 18:26:46 DEBUG GEOparse - SAMPLE: GSM566042\n",
      "03-Mar-2022 18:26:46 DEBUG GEOparse - SAMPLE: GSM566043\n",
      "03-Mar-2022 18:26:46 DEBUG GEOparse - SAMPLE: GSM566044\n",
      "03-Mar-2022 18:26:46 DEBUG GEOparse - SAMPLE: GSM566045\n",
      "03-Mar-2022 18:26:46 DEBUG GEOparse - SAMPLE: GSM566046\n",
      "03-Mar-2022 18:26:46 DEBUG GEOparse - SAMPLE: GSM566047\n",
      "03-Mar-2022 18:26:46 DEBUG GEOparse - SAMPLE: GSM566048\n",
      "03-Mar-2022 18:26:46 DEBUG GEOparse - SAMPLE: GSM566049\n",
      "03-Mar-2022 18:26:46 DEBUG GEOparse - SAMPLE: GSM566050\n",
      "03-Mar-2022 18:26:46 DEBUG GEOparse - SAMPLE: GSM566051\n",
      "03-Mar-2022 18:26:46 DEBUG GEOparse - SAMPLE: GSM566052\n",
      "03-Mar-2022 18:26:46 DEBUG GEOparse - SAMPLE: GSM566053\n",
      "03-Mar-2022 18:26:46 DEBUG GEOparse - SAMPLE: GSM566054\n",
      "03-Mar-2022 18:26:46 DEBUG GEOparse - SAMPLE: GSM566055\n",
      "03-Mar-2022 18:26:46 DEBUG GEOparse - SAMPLE: GSM566056\n",
      "03-Mar-2022 18:26:47 DEBUG GEOparse - SAMPLE: GSM566057\n",
      "03-Mar-2022 18:26:47 DEBUG GEOparse - SAMPLE: GSM566058\n",
      "03-Mar-2022 18:26:47 DEBUG GEOparse - SAMPLE: GSM566059\n",
      "03-Mar-2022 18:26:47 DEBUG GEOparse - SAMPLE: GSM566060\n",
      "03-Mar-2022 18:26:47 DEBUG GEOparse - SAMPLE: GSM566061\n",
      "03-Mar-2022 18:26:47 DEBUG GEOparse - SAMPLE: GSM566062\n",
      "03-Mar-2022 18:26:47 DEBUG GEOparse - SAMPLE: GSM566063\n",
      "03-Mar-2022 18:26:47 DEBUG GEOparse - SAMPLE: GSM566064\n",
      "03-Mar-2022 18:26:47 DEBUG GEOparse - SAMPLE: GSM566065\n",
      "03-Mar-2022 18:26:47 DEBUG GEOparse - SAMPLE: GSM566066\n",
      "03-Mar-2022 18:26:47 DEBUG GEOparse - SAMPLE: GSM566067\n",
      "03-Mar-2022 18:26:47 DEBUG GEOparse - SAMPLE: GSM566068\n",
      "03-Mar-2022 18:26:47 DEBUG GEOparse - SAMPLE: GSM566069\n",
      "03-Mar-2022 18:26:47 DEBUG GEOparse - SAMPLE: GSM566070\n",
      "03-Mar-2022 18:26:47 DEBUG GEOparse - SAMPLE: GSM566071\n",
      "03-Mar-2022 18:26:47 DEBUG GEOparse - SAMPLE: GSM566072\n",
      "03-Mar-2022 18:26:47 DEBUG GEOparse - SAMPLE: GSM566073\n",
      "03-Mar-2022 18:26:47 DEBUG GEOparse - SAMPLE: GSM566074\n",
      "03-Mar-2022 18:26:48 DEBUG GEOparse - SAMPLE: GSM566075\n",
      "03-Mar-2022 18:26:48 DEBUG GEOparse - SAMPLE: GSM566076\n",
      "03-Mar-2022 18:26:48 DEBUG GEOparse - SAMPLE: GSM566077\n",
      "03-Mar-2022 18:26:48 DEBUG GEOparse - SAMPLE: GSM566078\n",
      "03-Mar-2022 18:26:48 DEBUG GEOparse - SAMPLE: GSM566079\n",
      "03-Mar-2022 18:26:48 DEBUG GEOparse - SAMPLE: GSM566080\n",
      "03-Mar-2022 18:26:48 DEBUG GEOparse - SAMPLE: GSM566081\n",
      "03-Mar-2022 18:26:48 DEBUG GEOparse - SAMPLE: GSM566082\n",
      "03-Mar-2022 18:26:48 DEBUG GEOparse - SAMPLE: GSM566083\n",
      "03-Mar-2022 18:26:48 DEBUG GEOparse - SAMPLE: GSM566084\n",
      "03-Mar-2022 18:26:48 DEBUG GEOparse - SAMPLE: GSM566085\n",
      "03-Mar-2022 18:26:48 DEBUG GEOparse - SAMPLE: GSM566086\n",
      "03-Mar-2022 18:26:48 DEBUG GEOparse - SAMPLE: GSM566087\n",
      "03-Mar-2022 18:26:48 DEBUG GEOparse - SAMPLE: GSM566088\n",
      "03-Mar-2022 18:26:48 DEBUG GEOparse - SAMPLE: GSM566089\n",
      "03-Mar-2022 18:26:48 DEBUG GEOparse - SAMPLE: GSM566090\n",
      "03-Mar-2022 18:26:48 DEBUG GEOparse - SAMPLE: GSM566091\n",
      "03-Mar-2022 18:26:48 DEBUG GEOparse - SAMPLE: GSM566092\n",
      "03-Mar-2022 18:26:49 DEBUG GEOparse - SAMPLE: GSM566093\n",
      "03-Mar-2022 18:26:49 DEBUG GEOparse - SAMPLE: GSM566094\n",
      "03-Mar-2022 18:26:49 DEBUG GEOparse - SAMPLE: GSM566095\n",
      "03-Mar-2022 18:26:49 DEBUG GEOparse - SAMPLE: GSM566096\n",
      "03-Mar-2022 18:26:49 DEBUG GEOparse - SAMPLE: GSM566097\n",
      "03-Mar-2022 18:26:49 DEBUG GEOparse - SAMPLE: GSM566098\n",
      "03-Mar-2022 18:26:49 DEBUG GEOparse - SAMPLE: GSM566099\n",
      "Skip Unsupported Platform: GPL96, affy\n",
      "affy Read Path: /home/jupyteruser/work/data/GSE22886_RAW/GPL96\n",
      "background correction: mas \n",
      "PM/MM correction : mas \n",
      "expression values: mas \n",
      "background correcting...done.\n",
      "22283 ids to be processed\n",
      "|                    |\n",
      "|####################|\n",
      "Getting probe level data...\n",
      "Computing p-values\n",
      "Making P/M/A Calls\n",
      "gsm table drop:                     GSM565273.CEL.gz  ... GSM565292.CEL.gz.2\n",
      "ENTREZ_GENE_ID                       ...                   \n",
      "780 /// 100616237        328.955241  ...           0.043968\n",
      "5982                     287.603857  ...           0.000959\n",
      "3310                     343.371802  ...           0.218983\n",
      "7849                    1628.834605  ...           0.007543\n",
      "2978                      35.385856  ...           0.418069\n",
      "...                             ...  ...                ...\n",
      "2597                    9055.858104  ...           0.000044\n",
      "6772                    2774.904630  ...           0.000044\n",
      "6772                     201.494952  ...           0.000052\n",
      "6772                     887.867803  ...           0.000060\n",
      "6772                     535.751158  ...           0.000044\n",
      "\n",
      "[20973 rows x 18 columns]\n",
      "GPL96: (20973, 18)\n",
      "Full: (20973, 18)\n",
      "Full table saved to:\n",
      "/home/jupyteruser/work/data/GSE22886_RAW/GSE22886_sc500_full_table.csv\n",
      "Initialize project (GSE43005):\n",
      "Root: /home/jupyteruser/work\n",
      "Raw data: /home/jupyteruser/work/data/GSE43005_RAW\n",
      "Retrieve Sample: /home/jupyteruser/work/data/GSE43005_RAW/GSM1054773.tar\n",
      "Extract to: /home/jupyteruser/work/data/GSE43005_RAW/GPL14550\n",
      "Retrieve Sample: /home/jupyteruser/work/data/GSE43005_RAW/GSM1054779.tar\n",
      "Extract to: /home/jupyteruser/work/data/GSE43005_RAW/GPL14550\n",
      "Retrieve Sample: /home/jupyteruser/work/data/GSE43005_RAW/GSM1054781.tar\n",
      "Extract to: /home/jupyteruser/work/data/GSE43005_RAW/GPL14550\n",
      "Retrieve Sample: /home/jupyteruser/work/data/GSE43005_RAW/GSM1054789.tar\n",
      "Extract to: /home/jupyteruser/work/data/GSE43005_RAW/GPL14550\n",
      "Retrieve Samples Completed.\n",
      "Create new table: /home/jupyteruser/work/data/GSE43005_RAW/GSE43005_sc500_full_table.csv\n",
      "03-Mar-2022 18:27:32 DEBUG utils - Directory ./ already exists. Skipping.\n",
      "03-Mar-2022 18:27:32 INFO GEOparse - Downloading ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE43nnn/GSE43005/soft/GSE43005_family.soft.gz to ./GSE43005_family.soft.gz\n",
      "100%|██████████████████████████████████████| 15.6M/15.6M [00:00<00:00, 24.3MB/s]\n",
      "03-Mar-2022 18:27:34 DEBUG downloader - Size validation passed\n",
      "03-Mar-2022 18:27:34 DEBUG downloader - Moving /tmp/tmp6j6g0w4s to /home/jupyteruser/work/py/GSE43005_family.soft.gz\n",
      "03-Mar-2022 18:27:34 DEBUG downloader - Successfully downloaded ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE43nnn/GSE43005/soft/GSE43005_family.soft.gz\n",
      "03-Mar-2022 18:27:34 INFO GEOparse - Parsing ./GSE43005_family.soft.gz: \n",
      "03-Mar-2022 18:27:34 DEBUG GEOparse - DATABASE: GeoMiame\n",
      "03-Mar-2022 18:27:34 DEBUG GEOparse - SERIES: GSE43005\n",
      "03-Mar-2022 18:27:34 DEBUG GEOparse - PLATFORM: GPL14550\n",
      "03-Mar-2022 18:27:35 DEBUG GEOparse - SAMPLE: GSM1054773\n",
      "03-Mar-2022 18:27:35 DEBUG GEOparse - SAMPLE: GSM1054774\n",
      "03-Mar-2022 18:27:35 DEBUG GEOparse - SAMPLE: GSM1054775\n",
      "03-Mar-2022 18:27:35 DEBUG GEOparse - SAMPLE: GSM1054776\n",
      "03-Mar-2022 18:27:35 DEBUG GEOparse - SAMPLE: GSM1054777\n",
      "03-Mar-2022 18:27:35 DEBUG GEOparse - SAMPLE: GSM1054778\n",
      "03-Mar-2022 18:27:35 DEBUG GEOparse - SAMPLE: GSM1054779\n",
      "03-Mar-2022 18:27:35 DEBUG GEOparse - SAMPLE: GSM1054780\n",
      "03-Mar-2022 18:27:35 DEBUG GEOparse - SAMPLE: GSM1054781\n",
      "03-Mar-2022 18:27:36 DEBUG GEOparse - SAMPLE: GSM1054782\n",
      "03-Mar-2022 18:27:36 DEBUG GEOparse - SAMPLE: GSM1054783\n",
      "03-Mar-2022 18:27:36 DEBUG GEOparse - SAMPLE: GSM1054784\n",
      "03-Mar-2022 18:27:36 DEBUG GEOparse - SAMPLE: GSM1054785\n",
      "03-Mar-2022 18:27:36 DEBUG GEOparse - SAMPLE: GSM1054786\n",
      "03-Mar-2022 18:27:36 DEBUG GEOparse - SAMPLE: GSM1054787\n",
      "03-Mar-2022 18:27:36 DEBUG GEOparse - SAMPLE: GSM1054788\n",
      "03-Mar-2022 18:27:36 DEBUG GEOparse - SAMPLE: GSM1054789\n",
      "03-Mar-2022 18:27:36 DEBUG GEOparse - SAMPLE: GSM1054790\n",
      "03-Mar-2022 18:27:36 DEBUG GEOparse - SAMPLE: GSM1054791\n",
      "03-Mar-2022 18:27:37 DEBUG GEOparse - SAMPLE: GSM1054792\n",
      "03-Mar-2022 18:27:37 DEBUG GEOparse - SAMPLE: GSM1054793\n",
      "03-Mar-2022 18:27:37 DEBUG GEOparse - SAMPLE: GSM1054794\n",
      "03-Mar-2022 18:27:37 DEBUG GEOparse - SAMPLE: GSM1054795\n",
      "03-Mar-2022 18:27:37 DEBUG GEOparse - SAMPLE: GSM1054796\n",
      "retrieve 0:500\n",
      "retrieve 500:1000\n",
      "retrieve 1000:1500\n",
      "retrieve 1500:2000\n",
      "retrieve 2000:2500\n",
      "retrieve 2500:3000\n",
      "retrieve 3000:3500\n",
      "retrieve 3500:4000\n",
      "retrieve 4000:4500\n",
      "retrieve 4500:5000\n",
      "retrieve 5000:5500\n",
      "retrieve 5500:6000\n",
      "retrieve 6000:6500\n",
      "retrieve 6500:7000\n",
      "retrieve 7000:7500\n",
      "retrieve 7500:8000\n",
      "retrieve 8000:8500\n",
      "retrieve 8500:9000\n",
      "retrieve 9000:9500\n",
      "retrieve 9500:10000\n",
      "retrieve 10000:10500\n",
      "retrieve 10500:11000\n",
      "retrieve 11000:11500\n",
      "retrieve 11500:12000\n",
      "retrieve 12000:12500\n",
      "retrieve 12500:13000\n",
      "retrieve 13000:13500\n",
      "retrieve 13500:14000\n",
      "retrieve 14000:14500\n",
      "retrieve 14500:15000\n",
      "retrieve 15000:15500\n",
      "retrieve 15500:16000\n",
      "retrieve 16000:16500\n",
      "retrieve 16500:17000\n",
      "retrieve 17000:17500\n",
      "retrieve 17500:18000\n",
      "retrieve 18000:18500\n",
      "retrieve 18500:19000\n",
      "retrieve 19000:19500\n",
      "retrieve 19500:20000\n",
      "retrieve 20000:20500\n",
      "retrieve 20500:21000\n",
      "retrieve 21000:21500\n",
      "retrieve 21500:22000\n",
      "retrieve 22000:22500\n",
      "retrieve 22500:23000\n",
      "retrieve 23000:23500\n",
      "retrieve 23500:24000\n",
      "retrieve 24000:24500\n",
      "retrieve 24500:25000\n",
      "retrieve 25000:25500\n",
      "retrieve 25500:26000\n",
      "retrieve 26000:26500\n",
      "retrieve 26500:27000\n",
      "retrieve 27000:27500\n",
      "retrieve 27500:28000\n",
      "retrieve 28000:28500\n",
      "retrieve 28500:29000\n",
      "retrieve 29000:29500\n",
      "retrieve 29500:30000\n",
      "retrieve 30000:30500\n",
      "retrieve 30500:31000\n",
      "retrieve 31000:31500\n",
      "retrieve 31500:32000\n",
      "retrieve 32000:32500\n",
      "retrieve 32500:33000\n",
      "retrieve 33000:33500\n",
      "retrieve 33500:34000\n",
      "retrieve 34000:34500\n",
      "retrieve 34500:35000\n",
      "retrieve 35000:35500\n",
      "retrieve 35500:36000\n",
      "retrieve 36000:36500\n",
      "retrieve 36500:37000\n",
      "retrieve 37000:37500\n",
      "retrieve 37500:38000\n",
      "retrieve 38000:38500\n",
      "retrieve 38500:39000\n",
      "retrieve 39000:39500\n",
      "retrieve 39500:40000\n",
      "retrieve 40000:40500\n",
      "retrieve 40500:41000\n",
      "retrieve 41000:41500\n",
      "retrieve 41500:42000\n",
      "retrieve 42000:42405\n",
      "Skip Unsupported Platform: GPL14550, agilent\n",
      "agilent Read Path: /home/jupyteruser/work/data/GSE43005_RAW/GPL14550\n",
      "Read GSM1054773_US91203659_252800415876_S01_GE1_107_Sep09_1_1.txt.gz \n",
      "Read GSM1054779_US91203659_252800415876_S01_GE1_107_Sep09_2_3.txt.gz \n",
      "Read GSM1054781_US91203659_252800415877_S01_GE1_107_Sep09_1_1.txt.gz \n",
      "Read GSM1054789_US91203659_252800415878_S01_GE1_107_Sep09_1_1.txt.gz \n",
      "Array 1 corrected\n",
      "Array 2 corrected\n",
      "Array 3 corrected\n",
      "Array 4 corrected\n",
      "retrieve 0:500\n",
      "retrieve 500:1000\n",
      "retrieve 1000:1500\n",
      "retrieve 1500:2000\n",
      "retrieve 2000:2500\n",
      "retrieve 2500:3000\n",
      "retrieve 3000:3500\n",
      "retrieve 3500:4000\n",
      "retrieve 4000:4500\n",
      "retrieve 4500:5000\n",
      "retrieve 5000:5500\n",
      "retrieve 5500:6000\n",
      "retrieve 6000:6500\n",
      "retrieve 6500:7000\n",
      "retrieve 7000:7500\n",
      "retrieve 7500:8000\n",
      "retrieve 8000:8500\n",
      "retrieve 8500:9000\n",
      "retrieve 9000:9500\n",
      "retrieve 9500:10000\n",
      "retrieve 10000:10500\n",
      "retrieve 10500:11000\n",
      "retrieve 11000:11500\n",
      "retrieve 11500:12000\n",
      "retrieve 12000:12500\n",
      "retrieve 12500:13000\n",
      "retrieve 13000:13500\n",
      "retrieve 13500:14000\n",
      "retrieve 14000:14500\n",
      "retrieve 14500:15000\n",
      "retrieve 15000:15500\n",
      "retrieve 15500:16000\n",
      "retrieve 16000:16500\n",
      "retrieve 16500:17000\n",
      "retrieve 17000:17500\n",
      "retrieve 17500:18000\n",
      "retrieve 18000:18500\n",
      "retrieve 18500:19000\n",
      "retrieve 19000:19500\n",
      "retrieve 19500:20000\n",
      "retrieve 20000:20500\n",
      "retrieve 20500:21000\n",
      "retrieve 21000:21500\n",
      "retrieve 21500:22000\n",
      "retrieve 22000:22500\n",
      "retrieve 22500:23000\n",
      "retrieve 23000:23500\n",
      "retrieve 23500:24000\n",
      "retrieve 24000:24500\n",
      "retrieve 24500:25000\n",
      "retrieve 25000:25500\n",
      "retrieve 25500:26000\n",
      "retrieve 26000:26500\n",
      "retrieve 26500:27000\n",
      "retrieve 27000:27500\n",
      "retrieve 27500:28000\n",
      "retrieve 28000:28500\n",
      "retrieve 28500:29000\n",
      "retrieve 29000:29500\n",
      "retrieve 29500:30000\n",
      "retrieve 30000:30500\n",
      "retrieve 30500:31000\n",
      "retrieve 31000:31500\n",
      "retrieve 31500:32000\n",
      "retrieve 32000:32500\n",
      "retrieve 32500:33000\n",
      "retrieve 33000:33500\n",
      "retrieve 33500:34000\n",
      "retrieve 34000:34500\n",
      "retrieve 34500:35000\n",
      "retrieve 35000:35500\n",
      "retrieve 35500:36000\n",
      "retrieve 36000:36500\n",
      "retrieve 36500:37000\n",
      "retrieve 37000:37500\n",
      "retrieve 37500:38000\n",
      "retrieve 38000:38500\n",
      "retrieve 38500:39000\n",
      "retrieve 39000:39500\n",
      "retrieve 39500:40000\n",
      "retrieve 40000:40500\n",
      "retrieve 40500:41000\n",
      "retrieve 41000:41500\n",
      "retrieve 41500:42000\n",
      "retrieve 42000:42500\n",
      "retrieve 42500:43000\n",
      "retrieve 43000:43500\n",
      "retrieve 43500:44000\n",
      "retrieve 44000:44500\n",
      "retrieve 44500:45000\n",
      "retrieve 45000:45500\n",
      "retrieve 45500:46000\n",
      "retrieve 46000:46500\n",
      "retrieve 46500:47000\n",
      "retrieve 47000:47500\n",
      "retrieve 47500:48000\n",
      "retrieve 48000:48500\n",
      "retrieve 48500:49000\n",
      "retrieve 49000:49500\n",
      "retrieve 49500:50000\n",
      "retrieve 50000:50500\n",
      "retrieve 50500:51000\n",
      "retrieve 51000:51500\n",
      "retrieve 51500:52000\n",
      "retrieve 52000:52500\n",
      "retrieve 52500:53000\n",
      "retrieve 53000:53500\n",
      "retrieve 53500:54000\n",
      "retrieve 54000:54500\n",
      "retrieve 54500:55000\n",
      "retrieve 55000:55500\n",
      "retrieve 55500:56000\n",
      "retrieve 56000:56500\n",
      "retrieve 56500:57000\n",
      "retrieve 57000:57500\n",
      "retrieve 57500:58000\n",
      "retrieve 58000:58500\n",
      "retrieve 58500:58717\n",
      "gsm table drop:                     ProbeName  ...  gsm1054789.cel.gz.2\n",
      "ENTREZ_GENE_ID                ...                     \n",
      "199746          A_23_P326296  ...                 0.05\n",
      "29893           A_24_P287941  ...                 0.05\n",
      "84186           A_24_P325046  ...                 0.05\n",
      "204             A_23_P200404  ...                 0.05\n",
      "84366            A_23_P15619  ...                 0.05\n",
      "...                      ...  ...                  ...\n",
      "56271            A_24_P40551  ...                 0.05\n",
      "161145          A_23_P336239  ...                 0.05\n",
      "10497           A_23_P253395  ...                 0.05\n",
      "55283            A_23_P12241  ...                 0.05\n",
      "146433          A_23_P100386  ...                 0.05\n",
      "\n",
      "[46018 rows x 14 columns]\n",
      "GPL14550: (46018, 14)\n",
      "Full: (46018, 14)\n",
      "Full table saved to:\n",
      "/home/jupyteruser/work/data/GSE43005_RAW/GSE43005_sc500_full_table.csv\n",
      "Initialize project (GSE22045):\n",
      "Root: /home/jupyteruser/work\n",
      "Raw data: /home/jupyteruser/work/data/GSE22045_RAW\n",
      "Retrieve Sample: /home/jupyteruser/work/data/GSE22045_RAW/GSM548000.tar\n",
      "Extract to: /home/jupyteruser/work/data/GSE22045_RAW/GPL570\n",
      "Retrieve Sample: /home/jupyteruser/work/data/GSE22045_RAW/GSM548001.tar\n",
      "Extract to: /home/jupyteruser/work/data/GSE22045_RAW/GPL570\n",
      "Retrieve Samples Completed.\n",
      "Create new table: /home/jupyteruser/work/data/GSE22045_RAW/GSE22045_sc500_full_table.csv\n",
      "03-Mar-2022 18:39:46 DEBUG utils - Directory ./ already exists. Skipping.\n",
      "03-Mar-2022 18:39:46 INFO GEOparse - Downloading ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE22nnn/GSE22045/soft/GSE22045_family.soft.gz to ./GSE22045_family.soft.gz\n",
      "100%|██████████████████████████████████████| 13.0M/13.0M [00:00<00:00, 19.7MB/s]\n",
      "03-Mar-2022 18:39:47 DEBUG downloader - Size validation passed\n",
      "03-Mar-2022 18:39:47 DEBUG downloader - Moving /tmp/tmp8omohoxt to /home/jupyteruser/work/py/GSE22045_family.soft.gz\n",
      "03-Mar-2022 18:39:47 DEBUG downloader - Successfully downloaded ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE22nnn/GSE22045/soft/GSE22045_family.soft.gz\n",
      "03-Mar-2022 18:39:47 INFO GEOparse - Parsing ./GSE22045_family.soft.gz: \n",
      "03-Mar-2022 18:39:47 DEBUG GEOparse - DATABASE: GeoMiame\n",
      "03-Mar-2022 18:39:47 DEBUG GEOparse - SERIES: GSE22045\n",
      "03-Mar-2022 18:39:47 DEBUG GEOparse - PLATFORM: GPL570\n",
      "/usr/local/lib/python3.8/dist-packages/GEOparse/GEOparse.py:401: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  return read_csv(StringIO(data), index_col=None, sep=\"\\t\")\n",
      "03-Mar-2022 18:39:49 DEBUG GEOparse - SAMPLE: GSM547998\n",
      "03-Mar-2022 18:39:49 DEBUG GEOparse - SAMPLE: GSM547999\n",
      "03-Mar-2022 18:39:49 DEBUG GEOparse - SAMPLE: GSM548000\n",
      "03-Mar-2022 18:39:49 DEBUG GEOparse - SAMPLE: GSM548001\n",
      "Skip Unsupported Platform: GPL570, affy\n",
      "affy Read Path: /home/jupyteruser/work/data/GSE22045_RAW/GPL570\n",
      "background correction: mas \n",
      "PM/MM correction : mas \n",
      "expression values: mas \n",
      "background correcting...* installing *source* package ‘hgu133plus2cdf’ ...\n",
      "** using staged installation\n",
      "** R\n",
      "** data\n",
      "** byte-compile and prepare package for lazy loading\n",
      "Warning: replacing previous import ‘AnnotationDbi::tail’ by ‘utils::tail’ when loading ‘hgu133plus2cdf’\n",
      "Warning: replacing previous import ‘AnnotationDbi::head’ by ‘utils::head’ when loading ‘hgu133plus2cdf’\n",
      "** help\n",
      "*** installing help indices\n",
      "** building package indices\n",
      "** testing if installed package can be loaded from temporary location\n",
      "Warning: replacing previous import ‘AnnotationDbi::tail’ by ‘utils::tail’ when loading ‘hgu133plus2cdf’\n",
      "Warning: replacing previous import ‘AnnotationDbi::head’ by ‘utils::head’ when loading ‘hgu133plus2cdf’\n",
      "** testing if installed package can be loaded from final location\n",
      "Warning: replacing previous import ‘AnnotationDbi::tail’ by ‘utils::tail’ when loading ‘hgu133plus2cdf’\n",
      "Warning: replacing previous import ‘AnnotationDbi::head’ by ‘utils::head’ when loading ‘hgu133plus2cdf’\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (hgu133plus2cdf)\n",
      "done.\n",
      "54675 ids to be processed\n",
      "|                    |\n",
      "|####################|\n",
      "Getting probe level data...\n",
      "Computing p-values\n",
      "Making P/M/A Calls\n",
      "gsm table drop:                     GSM548000.CEL.gz  ... GSM548001.CEL.gz.2\n",
      "ENTREZ_GENE_ID                       ...                   \n",
      "780 /// 100616237       1113.368608  ...           0.001141\n",
      "5982                      92.354407  ...           0.002228\n",
      "3310                      31.343923  ...           0.520620\n",
      "7849                     270.096850  ...           0.054470\n",
      "2978                      14.557996  ...           0.882839\n",
      "...                             ...  ...                ...\n",
      "2597                    3420.062742  ...           0.000095\n",
      "6772                    2816.308410  ...           0.000052\n",
      "6772                      48.066362  ...           0.239063\n",
      "6772                     264.173910  ...           0.008440\n",
      "6772                     184.587920  ...           0.000070\n",
      "\n",
      "[44134 rows x 6 columns]\n",
      "GPL570: (44134, 6)\n",
      "Full: (44134, 6)\n",
      "Full table saved to:\n",
      "/home/jupyteruser/work/data/GSE22045_RAW/GSE22045_sc500_full_table.csv\n",
      "Initialize project (GSE24634):\n",
      "Root: /home/jupyteruser/work\n",
      "Raw data: /home/jupyteruser/work/data/GSE24634_RAW\n",
      "Retrieve Sample: /home/jupyteruser/work/data/GSE24634_RAW/GSM607510.tar\n",
      "Extract to: /home/jupyteruser/work/data/GSE24634_RAW/GPL96\n",
      "Retrieve Sample: /home/jupyteruser/work/data/GSE24634_RAW/GSM607511.tar\n",
      "Extract to: /home/jupyteruser/work/data/GSE24634_RAW/GPL96\n",
      "Retrieve Sample: /home/jupyteruser/work/data/GSE24634_RAW/GSM607512.tar\n",
      "Extract to: /home/jupyteruser/work/data/GSE24634_RAW/GPL96\n",
      "Retrieve Samples Completed.\n",
      "Create new table: /home/jupyteruser/work/data/GSE24634_RAW/GSE24634_sc500_full_table.csv\n",
      "affy Read Path: /home/jupyteruser/work/data/GSE24634_RAW/GPL96\n",
      "background correction: mas \n",
      "PM/MM correction : mas \n",
      "expression values: mas \n",
      "background correcting...done.\n",
      "22283 ids to be processed\n",
      "|                    |\n",
      "|####################|\n",
      "Getting probe level data...\n",
      "Computing p-values\n",
      "Making P/M/A Calls\n",
      "gsm table drop:                     GSM607510_don1_0.CEL.gz  ... GSM607512_don3_0.CEL.gz.2\n",
      "ENTREZ_GENE_ID                              ...                          \n",
      "780 /// 100616237               504.039037  ...                  0.024711\n",
      "5982                            223.617136  ...                  0.027860\n",
      "3310                            121.771340  ...                  0.284747\n",
      "7849                            646.715758  ...                  0.039365\n",
      "2978                             62.204475  ...                  0.162935\n",
      "...                                    ...  ...                       ...\n",
      "2597                           7253.919345  ...                  0.000060\n",
      "6772                           2425.484969  ...                  0.000081\n",
      "6772                             66.899590  ...                  0.327080\n",
      "6772                            248.750552  ...                  0.185131\n",
      "6772                            162.019919  ...                  0.015183\n",
      "\n",
      "[20973 rows x 9 columns]\n",
      "GPL96: (20973, 9)\n",
      "Full: (20973, 9)\n",
      "Full table saved to:\n",
      "/home/jupyteruser/work/data/GSE24634_RAW/GSE24634_sc500_full_table.csv\n",
      "Traceback (most recent call last):\n",
      "  File \"microarray_gen.py\", line 341, in <module>\n",
      "    main(sys.argv[1:])\n",
      "  File \"microarray_gen.py\", line 333, in main\n",
      "    df_output = queryTest(df, expression_proportion, top_proportion)\n",
      "  File \"microarray_gen.py\", line 143, in queryTest\n",
      "    df_output = mergeLogicalTable(df_results)\n",
      "  File \"microarray_gen.py\", line 213, in mergeLogicalTable\n",
      "    df_results = pd.concat([df_results, pd.DataFrame(dup_rows)], ignore_index=True)\n",
      "  File \"/usr/local/lib/python3.8/dist-packages/pandas/core/frame.py\", line 737, in __init__\n",
      "    mgr = ndarray_to_mgr(\n",
      "  File \"/usr/local/lib/python3.8/dist-packages/pandas/core/internals/construction.py\", line 331, in ndarray_to_mgr\n",
      "    values = _prep_ndarray(values, copy=copy_on_sanitize)\n",
      "  File \"/usr/local/lib/python3.8/dist-packages/pandas/core/internals/construction.py\", line 591, in _prep_ndarray\n",
      "    raise ValueError(f\"Must pass 2-d input. shape={values.shape}\")\n",
      "ValueError: Must pass 2-d input. shape=(2, 1, 16)\n"
     ]
    }
   ],
   "source": [
    "# Step 2.1 Download and Analyze microarray data\n",
    "\n",
    "# config file for microarray\n",
    "microarray_config_file = 'microarray_data_inputs.xlsx'\n",
    "expression_proportion = 0.5\n",
    "top_proportion = 0.9\n",
    "\n",
    "# execute\n",
    "cmd = ' '.join(['python3', 'microarray_gen.py', \n",
    "      '-i', '\"{}\"'.format(microarray_config_file),\n",
    "      '-e', '\"{}\"'.format(expression_proportion),\n",
    "      '-t', '\"{}\"'.format(top_proportion)])\n",
    "!{cmd}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **RNA-seq Analysis**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "RNA-seq analysis has two primary types, bulk tissue, and single-cell. Bulk RNA-seq also has multiple strategies of library preparation. If using public data, the user may run into a situation where they wish to use a combination of bulk RNA-seq data produced using two very different library preparation strategies. \n",
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Bulk RNA-seq**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "MADRID currently supports the two most common strategies, mRNA (polyA) enriched RNA-seq, and total RNA-seq. \n",
    "\n",
    "Because of expected differences in distribution of transcripts, MADRID is written to handle each strategy seperately before the integration step. The recommended Snakemake alignment pipeline is designed to work with MADRID's preprocessing step (Step 1) to split RNA-seq data from GEO into seperate input matrices and config sheets.\n",
    "\n",
    "**To create a gene expression file for total RNA-seq data, use the \"total\" for the '--library-prep' argument**\n",
    "\n",
    "**To create a gene expression file for mRNA enriched / polA RNA-seq data, use the \"mRNA\" for the '--library-prep' argument.**\n",
    "\n",
    "The analysis of each strategy is identical so specifying the type only serves to ensure MADRID analyzes them seperately."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Single-cell RNA-seq**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "While the Snakemake pipeline does not yet support single-cell alignment and MADRID does not yet support automated configuration file and counts-matrix file creation for single-cell alignment output from STAR, it is possible to use single-cell data to create a model with MADRID. \n",
    "\n",
    "Since normalization strategies can be applied to single-cell the same way it is applied to bulk, rnaseq_gen.py can be used with a provided counts matrix and config sheet (see Step 1 to help create it). \n",
    "\n",
    "Just like 'total' and 'mRNA', rnaseq_gen.py can be run with \"SC\" as the '--library-prep' argument to help MADRID differentiate it from the bulk RNA-seq data if using multiple strategies.   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001B[0;1;31mSystem has not been booted with systemd as init system (PID 1). Can't operate.\u001B[0m\n",
      "\u001B[0;1;31mFailed to create bus connection: Host is down\u001B[0m\n",
      "Config file is \"trnaseq_data_inputs_auto.xlsx\"\n",
      "model:  m0Macro\n",
      "Input count matrix is at \"/home/jupyteruser/work/data/data_matrices/m0Macro/gene_counts_matrix_total_m0Macro.csv\"\n",
      "Gene info file is at \"/home/jupyteruser/work/data/gene_info.csv\"\n",
      "[1] \"Reading Counts Matrix\"\n",
      "[1] \"/home/jupyteruser/work/data/data_matrices/m0Macro/gene_counts_matrix_total_m0Macro.csv\"\n",
      "[1] \"m0Macro\"\n",
      "[1] \"Filtering Counts\"\n",
      "Test data saved to /home/jupyteruser/work/data/results/m0Macro/rnaseq_total_m0Macro.csv\n",
      "model:  m1Macro\n",
      "Input count matrix is at \"/home/jupyteruser/work/data/data_matrices/m1Macro/gene_counts_matrix_total_m1Macro.csv\"\n",
      "Gene info file is at \"/home/jupyteruser/work/data/gene_info.csv\"\n",
      "[1] \"Reading Counts Matrix\"\n",
      "[1] \"/home/jupyteruser/work/data/data_matrices/m1Macro/gene_counts_matrix_total_m1Macro.csv\"\n",
      "[1] \"m1Macro\"\n",
      "[1] \"Filtering Counts\"\n",
      "Test data saved to /home/jupyteruser/work/data/results/m1Macro/rnaseq_total_m1Macro.csv\n",
      "model:  m2Macro\n",
      "Input count matrix is at \"/home/jupyteruser/work/data/data_matrices/m2Macro/gene_counts_matrix_total_m2Macro.csv\"\n",
      "Gene info file is at \"/home/jupyteruser/work/data/gene_info.csv\"\n",
      "[1] \"Reading Counts Matrix\"\n",
      "[1] \"/home/jupyteruser/work/data/data_matrices/m2Macro/gene_counts_matrix_total_m2Macro.csv\"\n",
      "[1] \"m2Macro\"\n",
      "[1] \"Filtering Counts\"\n",
      "Test data saved to /home/jupyteruser/work/data/results/m2Macro/rnaseq_total_m2Macro.csv\n",
      "model:  naiveB\n",
      "Input count matrix is at \"/home/jupyteruser/work/data/data_matrices/naiveB/gene_counts_matrix_total_naiveB.csv\"\n",
      "Gene info file is at \"/home/jupyteruser/work/data/gene_info.csv\"\n",
      "[1] \"Reading Counts Matrix\"\n",
      "[1] \"/home/jupyteruser/work/data/data_matrices/naiveB/gene_counts_matrix_total_naiveB.csv\"\n",
      "[1] \"naiveB\"\n",
      "[1] \"Filtering Counts\"\n",
      "Test data saved to /home/jupyteruser/work/data/results/naiveB/rnaseq_total_naiveB.csv\n",
      "model:  nsmB\n",
      "Input count matrix is at \"/home/jupyteruser/work/data/data_matrices/nsmB/gene_counts_matrix_total_nsmB.csv\"\n",
      "Gene info file is at \"/home/jupyteruser/work/data/gene_info.csv\"\n",
      "[1] \"Reading Counts Matrix\"\n",
      "[1] \"/home/jupyteruser/work/data/data_matrices/nsmB/gene_counts_matrix_total_nsmB.csv\"\n",
      "[1] \"nsmB\"\n",
      "[1] \"Filtering Counts\"\n",
      "Test data saved to /home/jupyteruser/work/data/results/nsmB/rnaseq_total_nsmB.csv\n",
      "model:  smB\n",
      "Input count matrix is at \"/home/jupyteruser/work/data/data_matrices/smB/gene_counts_matrix_total_smB.csv\"\n",
      "Gene info file is at \"/home/jupyteruser/work/data/gene_info.csv\"\n",
      "[1] \"Reading Counts Matrix\"\n",
      "[1] \"/home/jupyteruser/work/data/data_matrices/smB/gene_counts_matrix_total_smB.csv\"\n",
      "[1] \"smB\"\n",
      "[1] \"Filtering Counts\"\n",
      "Test data saved to /home/jupyteruser/work/data/results/smB/rnaseq_total_smB.csv\n"
     ]
    }
   ],
   "source": [
    "# step 2.2 RNA-seq Analysis for Total RNA-seq library preparation\n",
    "\n",
    "# config for total rna-seq\n",
    "trnaseq_config_file = 'trnaseq_data_inputs_auto.xlsx'\n",
    "\n",
    "rep_ratio = 0.5          # proportion of replicates for a gene to be active in a sample\n",
    "group_ratio = 0.9        # proportion of samples with expression required for gene  \n",
    "rep_ratio_h = 0.5        # proportion of replicates with expression required for high-confidence\n",
    "group_ratio_h = 0.9      # proportion of replicates with expression required for high-confidence\n",
    "technique = \"zFPKM\"      # quantile, cpm, or zfpkm\n",
    "quantile = 50            # cutoff TPM quantile for quantile filtering\n",
    "min_count = \"default\"    # cutoff for CPM filtering\n",
    "prep_method = \"total\"    # library prepartion method ('total', mRNA', or 'SC')\n",
    "\n",
    "cmd = ' '.join(['python3', 'rnaseq_gen.py', \n",
    "      '--config-file', '\"{}\"'.format(trnaseq_config_file), \n",
    "      '--replicate-ratio', '\"{}\"'.format(rep_ratio),   \n",
    "      '--batch-ratio', '\"{}\"'.format(group_ratio),        \n",
    "      '--high-replicate-ratio', '\"{}\"'.format(rep_ratio_h),    \n",
    "      '--high-batch-ratio', '\"{}\"'.format(group_ratio_h),   \n",
    "      '--filt-technique', '\"{}\"'.format(technique),  \n",
    "      '--min-count', '\"{}\"'.format(min_count),\n",
    "      '--quantile', '\"{}\"'.format(quantile),\n",
    "      '--library-prep', '\"{}\"'.format(prep_method)])       \n",
    "                \n",
    "!{cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001B[0;1;31mSystem has not been booted with systemd as init system (PID 1). Can't operate.\u001B[0m\n",
      "\u001B[0;1;31mFailed to create bus connection: Host is down\u001B[0m\n",
      "Config file is \"mrnaseq_data_inputs_auto.xlsx\"\n",
      "model:  m0Macro\n",
      "Input count matrix is at \"/home/jupyteruser/work/data/data_matrices/m0Macro/gene_counts_matrix_mrna_m0Macro.csv\"\n",
      "Gene info file is at \"/home/jupyteruser/work/data/gene_info.csv\"\n",
      "[1] \"Reading Counts Matrix\"\n",
      "[1] \"/home/jupyteruser/work/data/data_matrices/m0Macro/gene_counts_matrix_mrna_m0Macro.csv\"\n",
      "[1] \"m0Macro\"\n",
      "[1] \"Filtering Counts\"\n",
      "Test data saved to /home/jupyteruser/work/data/results/m0Macro/rnaseq_mrna_m0Macro.csv\n",
      "model:  naiveB\n",
      "Input count matrix is at \"/home/jupyteruser/work/data/data_matrices/naiveB/gene_counts_matrix_mrna_naiveB.csv\"\n",
      "Gene info file is at \"/home/jupyteruser/work/data/gene_info.csv\"\n",
      "[1] \"Reading Counts Matrix\"\n",
      "[1] \"/home/jupyteruser/work/data/data_matrices/naiveB/gene_counts_matrix_mrna_naiveB.csv\"\n",
      "[1] \"naiveB\"\n",
      "[1] \"Filtering Counts\"\n",
      "Test data saved to /home/jupyteruser/work/data/results/naiveB/rnaseq_mrna_naiveB.csv\n",
      "model:  plasmaB\n",
      "Input count matrix is at \"/home/jupyteruser/work/data/data_matrices/plasmaB/gene_counts_matrix_mrna_plasmaB.csv\"\n",
      "Gene info file is at \"/home/jupyteruser/work/data/gene_info.csv\"\n",
      "[1] \"Reading Counts Matrix\"\n",
      "[1] \"/home/jupyteruser/work/data/data_matrices/plasmaB/gene_counts_matrix_mrna_plasmaB.csv\"\n",
      "[1] \"plasmaB\"\n",
      "[1] \"Filtering Counts\"\n",
      "Test data saved to /home/jupyteruser/work/data/results/plasmaB/rnaseq_mrna_plasmaB.csv\n"
     ]
    }
   ],
   "source": [
    " # step 2.3 mRNA capture (polyA) RNA-seq Analysis\n",
    "\n",
    "# config for mRNA (polyA) enriched RNA-seq \n",
    "mrnaseq_config_file = 'mrnaseq_data_inputs_auto.xlsx'\n",
    "\n",
    "rep_ratio = 0.5          # proportion of replicates for a gene to be active in a sample\n",
    "group_ratio = 0.9        # proportion of samples with expression required for gene  \n",
    "rep_ratio_h = 0.5        # proportion of replicates with expression required for high-confidence\n",
    "group_ratio_h = 0.9      # proportion of replicates with expression required for high-confidence\n",
    "technique = \"zfpkm\"      # quantile-tpm, cpm, or zfpkm\n",
    "min_count = \"default\"    # cutoff for CPM for filtering\n",
    "quantile = 50            # cutoff TPM quantile for filtering\n",
    "prep_method = \"mrna\"     # library preparation method\n",
    "\n",
    "cmd = ' '.join(['python3', 'rnaseq_gen.py', \n",
    "      '--config-file', '\"{}\"'.format(mrnaseq_config_file), \n",
    "      '--replicate-ratio', '\"{}\"'.format(rep_ratio),   \n",
    "      '--batch-ratio', '\"{}\"'.format(group_ratio),        \n",
    "      '--high-replicate-ratio', '\"{}\"'.format(rep_ratio_h),    \n",
    "      '--high-batch-ratio', '\"{}\"'.format(group_ratio_h),   \n",
    "      '--filt-technique', '\"{}\"'.format(technique),  \n",
    "      '--min-count', '\"{}\"'.format(min_count),\n",
    "      '--quantile', '\"{}\"'.format(quantile),\n",
    "      '--library-prep', '\"{}\"'.format(prep_method)])       \n",
    "                \n",
    "!{cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001B[0;1;31mSystem has not been booted with systemd as init system (PID 1). Can't operate.\u001B[0m\n",
      "\u001B[0;1;31mFailed to create bus connection: Host is down\u001B[0m\n",
      "[1] \"Making logical matrix\"\n",
      "[1] \"Clustering replicate level filtered data\"\n",
      "[1] \"FALSE\"\n",
      "[1] \"label\"\n",
      "[1] \"Clustering batch level filtered data\"\n",
      "[1] \"Clustering context level filtered data\"\n"
     ]
    }
   ],
   "source": [
    "# Step 2.4 Cluster RNA-seq data for outlier exploration\n",
    "\n",
    "clust_algo = \"umap\"\n",
    "label = \"False\"\n",
    "min_dist = 0.01\n",
    "rep_ratio = 0.99\n",
    "batch_ratio = 0.99\n",
    "n_neigh_rep = 30\n",
    "n_neigh_batch = 15\n",
    "n_neigh_cont = 4\n",
    "\n",
    "\n",
    "cmd = ' '.join(['python3', 'cluster_rnaseq.py', \n",
    "      '--context-names', '\"{}\"'.format(context_names),\n",
    "      '--filt-technique', '\"{}\"'.format(technique), \n",
    "      '--cluster-algorithm', '\"{}\"'.format(clust_algo),\n",
    "      '--label', '\"{}\"'.format(label),\n",
    "      '--min-dist', '\"{}\"'.format(min_dist),\n",
    "      '--replicate-ratio', '\"{}\"'.format(rep_ratio),   \n",
    "      '--batch-ratio', '\"{}\"'.format(batch_ratio), \n",
    "      '--n-neighbors-rep', '\"{}\"'.format(n_neigh_rep),\n",
    "      '--n-neighbors-batch', '\"{}\"'.format(n_neigh_batch),\n",
    "      '--n-neighbors-context', '\"{}\"'.format(n_neigh_cont),\n",
    "      '--min-count', '\"{}\"'.format(min_count),\n",
    "      '--quantile', '\"{}\"'.format(quantile)])       \n",
    "                \n",
    "!{cmd}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/jupyteruser/work/data'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "configs.datadir\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Config file is at \"/home/jupyteruser/work/data/config_sheets/proteomics_data_inputs.xlsx\"\n",
      "Data matrix is at \"/home/jupyteruser/work/data/data_matrices/Naive/ProteomicsDataMatrix_Naive.csv\"\n",
      "Test Data Saved to /home/jupyteruser/work/data/results/Naive/Proteomics_Naive.csv\n"
     ]
    }
   ],
   "source": [
    "# Step 2.5 Proteomics Analysis\n",
    "\n",
    "# config file for proteomics\n",
    "proteomics_config_file = 'proteomics_data_inputs.xlsx'\n",
    "\n",
    "# ratio of replicates required for a gene to be considered active in that sample\n",
    "active_ratio = 0.5\n",
    "\n",
    "# Genes can be considered high confidence if they are expressed in a high proportion of samples.\n",
    "# High confidence genes will be considered expressed regardless of agreement with other data sources\n",
    "high_ratio = 0.9\n",
    "\n",
    "quantile = 25\n",
    "\n",
    "cmd = ' '.join(['python3', 'proteomics_gen.py', \n",
    "      '-c', '\"{}\"'.format(proteomics_config_file),\n",
    "      '-e', '\"{}\"'.format(active_ratio),\n",
    "      '-t', '\"{}\"'.format(high_ratio),\n",
    "      '-p', '\"{}\"'.format(quantile)])\n",
    "!{cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "true\n",
      "true\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 3: Merge Expression from Different Data Sources"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So far, active genes have been determined for at least one data source. If using multiple sources of any combination of microarray, bulk RNA-seq of either total RNA or mRNA capture (polyA), proteomics, or single-cell RNA-seq. Now we can merge the active genes from each data source to make a list of active genes that is more comprehenisve or more strict than any individual list. \n",
    "\n",
    "`merge_xomics.py` takes each data source discussed so far as an argument, and it is easy to add new ones to the script if desired. The other arguments to consider are:\n",
    "- **--expression-requirement** which is the number of data souces with expression required for a gene to be considered active if not a high-confidence gene for any source (defaults to the total number of input data sources arguments provided)\n",
    "\n",
    "- **--requirement-adjust** is the method to adjust this requirement in the event that tissues have different numbers of provided data sources. (*Note that this does nothing if there is only one tissue type in the config files). \n",
    "\n",
    "> - \"progressive\" - expression requirement applies to tissue(s) with lowest number of data source types. Tissues with more will require 1 more source to have an active gene per additional source provided for the gene to be active in the model.\n",
    "        \n",
    "> - \"regressive\" - expression requirement applies to the tissue(s) with largest number of data source types. Tissues with less will require 1 less source to have an active gene per missing source for the gene to be active in the model\n",
    "                    \n",
    "> - \"flat\" - (default) expression requirement is used regardless of differences in number of data sources provided for different tissues\n",
    "\n",
    "> - \"custom\" - (Not yet implemented!) an .xlsx file where column one is the tissue type, and column two is the expression requirement. Requires additional argument '--requirements-file' whose value is the filename of this .xlsx file in `/work/data/`\n",
    "                          \n",
    "- **--no-hc** use this flag to prevent high-confidence gene from overiding expression_requirement\n",
    "\n",
    "- **--no-na-adjustment** use this flag to prevent genes that are not a present in one data source, but are present in others from subtracting one from the expression requirement.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If the '--no-hc' flag is not used, any gene that was determined to be high-confidence in any input data source will cause the gene to be active in the final model, regardless of agreement with other sources.**\n",
    "<br />\n",
    "\n",
    "If the '--no-na-adjustment' flag is not used, any time a gene is NA in a data source, meaning it was not tested for in the library of that data source, but was tested in the library of at least one other, it will subtract one from the expression requirement.\n",
    "\n",
    "***Adjusted expression requirement will never resolve to be < 1 or > the number of data sources that tissue is given**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  tissue req\n",
      "0  three   3\n",
      "1    two   2\n",
      "2    one   1\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df = pd.DataFrame({\"tissue\": [\"three\", \"two\", \"one\"], \"req\": [\"3\", \"2\", \"1\"]})\n",
    "print(df.head())\n",
    "\n",
    "int(df.loc[df[\"tissue\"] == \"three\", \"req\"].iloc[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001B[0;1;31mSystem has not been booted with systemd as init system (PID 1). Can't operate.\u001B[0m\n",
      "\u001B[0;1;31mFailed to create bus connection: Host is down\u001B[0m\n",
      "Counter({'m0Macro': 2, 'naiveB': 2, 'm1Macro': 1, 'm2Macro': 1, 'nsmB': 1, 'smB': 1, 'plasmaB': 1})\n",
      "Will merge data for: ['m0Macro', 'm1Macro', 'm2Macro', 'naiveB', 'nsmB', 'plasmaB', 'smB']\n",
      "Expression requirement of 1 adjusted to 2 using progressive adjustment method for m0Macro.\n",
      "Merging data for m0Macro\n",
      "Read from /home/jupyteruser/work/data/results/m0Macro/rnaseq_total_m0Macro.csv\n",
      "Read from /home/jupyteruser/work/data/results/m0Macro/rnaseq_mrna_m0Macro.csv\n",
      "m0Macro: save to /home/jupyteruser/work/data/results/m0Macro/ActiveGenes_m0Macro_Merged.csv\n",
      "\n",
      "Expression requirement of 1 adjusted to 1 using progressive adjustment method for m1Macro.\n",
      "Merging data for m1Macro\n",
      "Read from /home/jupyteruser/work/data/results/m1Macro/rnaseq_total_m1Macro.csv\n",
      "mrna gene expression file for m1Macro was not found at /home/jupyteruser/work/data/results/m1Macro/rnaseq_mrna_m1Macro.csv. This may be intentional. Contexts where mrna data can be found in /work/data/results/m1Macro/ will still be used if found for other contexts.\n",
      "m1Macro: save to /home/jupyteruser/work/data/results/m1Macro/ActiveGenes_m1Macro_Merged.csv\n",
      "\n",
      "Expression requirement of 1 adjusted to 1 using progressive adjustment method for m2Macro.\n",
      "Merging data for m2Macro\n",
      "Read from /home/jupyteruser/work/data/results/m2Macro/rnaseq_total_m2Macro.csv\n",
      "mrna gene expression file for m2Macro was not found at /home/jupyteruser/work/data/results/m2Macro/rnaseq_mrna_m2Macro.csv. This may be intentional. Contexts where mrna data can be found in /work/data/results/m2Macro/ will still be used if found for other contexts.\n",
      "m2Macro: save to /home/jupyteruser/work/data/results/m2Macro/ActiveGenes_m2Macro_Merged.csv\n",
      "\n",
      "Expression requirement of 1 adjusted to 2 using progressive adjustment method for naiveB.\n",
      "Merging data for naiveB\n",
      "Read from /home/jupyteruser/work/data/results/naiveB/rnaseq_total_naiveB.csv\n",
      "Read from /home/jupyteruser/work/data/results/naiveB/rnaseq_mrna_naiveB.csv\n",
      "naiveB: save to /home/jupyteruser/work/data/results/naiveB/ActiveGenes_naiveB_Merged.csv\n",
      "\n",
      "Expression requirement of 1 adjusted to 1 using progressive adjustment method for nsmB.\n",
      "Merging data for nsmB\n",
      "Read from /home/jupyteruser/work/data/results/nsmB/rnaseq_total_nsmB.csv\n",
      "mrna gene expression file for nsmB was not found at /home/jupyteruser/work/data/results/nsmB/rnaseq_mrna_nsmB.csv. This may be intentional. Contexts where mrna data can be found in /work/data/results/nsmB/ will still be used if found for other contexts.\n",
      "nsmB: save to /home/jupyteruser/work/data/results/nsmB/ActiveGenes_nsmB_Merged.csv\n",
      "\n",
      "Expression requirement of 1 adjusted to 1 using progressive adjustment method for plasmaB.\n",
      "Merging data for plasmaB\n",
      "total gene expression file for plasmaB was not found at /home/jupyteruser/work/data/results/plasmaB/rnaseq_total_plasmaB.csv. This may be intentional. Contexts where total data can be found in /work/data/results/plasmaB/ will still be used if found for other contexts.\n",
      "Read from /home/jupyteruser/work/data/results/plasmaB/rnaseq_mrna_plasmaB.csv\n",
      "plasmaB: save to /home/jupyteruser/work/data/results/plasmaB/ActiveGenes_plasmaB_Merged.csv\n",
      "\n",
      "Expression requirement of 1 adjusted to 1 using progressive adjustment method for smB.\n",
      "Merging data for smB\n",
      "Read from /home/jupyteruser/work/data/results/smB/rnaseq_total_smB.csv\n",
      "mrna gene expression file for smB was not found at /home/jupyteruser/work/data/results/smB/rnaseq_mrna_smB.csv. This may be intentional. Contexts where mrna data can be found in /work/data/results/smB/ will still be used if found for other contexts.\n",
      "smB: save to /home/jupyteruser/work/data/results/smB/ActiveGenes_smB_Merged.csv\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Step 2.4 Merge the gene lists of data sources, create a list of active gene IDs\n",
    "\n",
    "expression_requirement = 1\n",
    "        \n",
    "requirement_adjust = \"progressive\" \n",
    "\n",
    "\n",
    "cmd = ' '.join(['python3', 'merge_xomics.py', \n",
    "      #'--microarray-config-file', '\"{}\"'.format(microarray_config_file),\n",
    "      '--total-rnaseq-config-file', '\"{}\"'.format(trnaseq_config_file),\n",
    "      '--mrnaseq-config-file', '\"{}\"'.format(mrnaseq_config_file),\n",
    "      #'--proteomics-config-file', '\"{}\"'.format(proteomics_config_file),\n",
    "      '--expression-requirement', '\"{}\"'.format(expression_requirement),\n",
    "      '--requirement-adjust', '\"{}\"'.format(requirement_adjust)])\n",
    "!{cmd}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Step 4: Create Tissue/Cell-Type Specific Models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'m0Macro': '/home/jupyteruser/work/data/results/m0Macro/ActiveGenes_m0Macro_Merged.csv', 'm1Macro': '/home/jupyteruser/work/data/results/m1Macro/ActiveGenes_m1Macro_Merged.csv', 'm2Macro': '/home/jupyteruser/work/data/results/m2Macro/ActiveGenes_m2Macro_Merged.csv', 'naiveB': '/home/jupyteruser/work/data/results/naiveB/ActiveGenes_naiveB_Merged.csv', 'nsmB': '/home/jupyteruser/work/data/results/nsmB/ActiveGenes_nsmB_Merged.csv', 'plasmaB': '/home/jupyteruser/work/data/results/plasmaB/ActiveGenes_plasmaB_Merged.csv', 'smB': '/home/jupyteruser/work/data/results/smB/ActiveGenes_smB_Merged.csv'}\n"
     ]
    }
   ],
   "source": [
    "# Load the output of step 1, which is a dictionary that specifies the merged list of active Gene IDs for each tissue\n",
    "\n",
    "step1_results_file = os.path.join(configs.rootdir, 'data', 'results', 'step1_results_files.json')\n",
    "with open(step1_results_file) as json_file:\n",
    "    context_gene_exp = json.load(json_file)\n",
    "print(context_gene_exp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*** Specify input files for step 2 here ***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading /home/jupyteruser/work/data/local_files/VIS/boundary_rxns/naiveB_boundary_rxns.csv for boundary reactions\n",
      "Constructing model of m0Macro with FASTCORE reconstruction algorithm using GUROBI solver\n",
      "Set parameter WLSAccessID\n",
      "Set parameter WLSSecret\n",
      "Set parameter LicenseID\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "Read LP format model from file /tmp/tmpo_jhd4ea.lp\n",
      "Reading time = 0.06 seconds\n",
      ": 5837 rows, 21220 columns, 81074 nonzeros\n",
      "Under given boundary assumptions, there are \"1965\" infeasible reactions in the reference model.\n",
      "\n",
      "These reactions will not be considered active in context specific model construction. If any infeasible reactions are found to be active according to expression data, or, are found in the force reactions list, they can be found found in \"InfeasibleRxns.csv\n",
      "\n",
      "It is normal for this value to be quite large, however, if many of these reactions are active according to your expression data, it is likely that you are missing some critical exchange (media) reactions.\n",
      "\n",
      "/usr/local/lib/python3.8/dist-packages/cobamp/gpr/core.py:115: UserWarning: Will not normalize rules with more than 20 average tokens per gene\n",
      "  warnings.warn(\n",
      "Map gene expression to reactions, 0 errors.\n",
      "J size3701\n",
      "[    0     1     4 ... 10588 10606 10607]\n",
      "before LP7\n",
      "LP7\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "-0.37009999999997556\n",
      "done LP7\n",
      "LP9\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "306396.71664265037\n",
      "done LP9\n",
      "463 6051\n",
      "before LP7\n",
      "LP7\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "-0.016599999999342935\n",
      "done LP7\n",
      "LP9\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "1685.0000000070213\n",
      "done LP9\n",
      "463 6391\n",
      "before LP7\n",
      "LP7\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "0.0\n",
      "done LP7\n",
      "279 6391\n",
      "Flipped\n",
      "before LP7\n",
      "LP7\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "-0.02749999999999989\n",
      "done LP7\n",
      "LP9\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "5221.666666696468\n",
      "done LP9\n",
      "279 7190\n",
      "before LP7\n",
      "LP7\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "0.0\n",
      "done LP7\n",
      "4 7190\n",
      "Flipped\n",
      "before LP7\n",
      "LP7\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "0.0\n",
      "done LP7\n",
      "4 7190\n",
      "before LP7\n",
      "LP7\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "-0.0001\n",
      "done LP7\n",
      "LP9\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "20.0\n",
      "done LP9\n",
      "4 7194\n",
      "before LP7\n",
      "LP7\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "-0.0001\n",
      "done LP7\n",
      "LP9\n",
      "Academic license - for non-commercial use only - registered to bbessell2@unl.edu\n",
      "1.4901161193847656e-08\n",
      "done LP9\n",
      "2 7196\n",
      "0 7196\n",
      "Read LP format model from file /tmp/tmp0n8sarq9.lp\n",
      "Reading time = 0.05 seconds\n",
      ": 5837 rows, 21220 columns, 81074 nonzeros\n"
     ]
    }
   ],
   "source": [
    "# create tissue specific model, the names of output files are stored in dictionary tissue_spec_model\n",
    "\n",
    "general_model_file = \"GeneralModelUpdated.mat\"\n",
    "#exclude_rxns_file = os.path.join(configs.datadir, \"inconsistant_rxns.csv\") # flux inconsistant rxns to remove from core reactions in fastcore\n",
    "#force_rxns_file = os.path.join(configs.datadir, \"lit_core_rxns.csv\") \n",
    "#boundary_rxns_file = os.path.join(configs.datadir, \"exchange_rxns,csv\")\n",
    "recon_algorithm = 'FASTCORE' # troppo reconstruction algorithm to use\n",
    "solver = \"GUROBI\"\n",
    "objective = 'biomass_reaction_Mphage'\n",
    "#objective = 'BIOMASS_reaction'\n",
    "vis_folder = os.path.join(configs.datadir, \"local_files\", \"VIS\")\n",
    "force_rxns_filepath = os.path.join(vis_folder, \"force_rxns\", \"naiveB_force_rxns.csv\")\n",
    "boundary_rxns_filepath = os.path.join(vis_folder, \"boundary_rxns\", \"naiveB_boundary_rxns.csv\")\n",
    "output_filetypes = \"['xml', 'mat', 'json']\"\n",
    "\n",
    "for key,value in context_gene_exp.items():\n",
    "    #contextmodelfile = '{}_SpecificModel.mat'.format(key) # key is == tissue name\n",
    "    active_genes_file = re.split('/|\\\\\\\\', value)[-1]\n",
    "    active_genes_filepath = os.path.join(configs.datadir, \"results\", key, active_genes_file)\n",
    "    general_model_filepath = os.path.join(configs.datadir, general_model_file)\n",
    "    #tissue_gene_folder = os.path.join(configs.rootdir, 'data', key)\n",
    "    #os.makedirs(tissue_gene_folder, exist_ok=True)\n",
    "    cmd = ' '.join(['python3', 'create_context_specific_model.py', \n",
    "                      '--context-name', '\"{}\"'.format(key),\n",
    "                      '--reference-model-file', '\"{}\"'.format(general_model_filepath), \n",
    "                      '--active-genes-filepath', '\"{}\"'.format(active_genes_filepath),\n",
    "                      '--objective', '\"{}\"'.format(objective),\n",
    "                      '--boundary-reactions-filepath', '\"{}\"'.format(boundary_rxns_filepath),\n",
    "                      #'--exclude-reactions-filepath', '\"{}\"'.format(exclude_rxns_file),\n",
    "                      '--force-reactions-filepath', '\"{}\"'.format(force_rxns_filepath),\n",
    "                      '--algorithm', '\"{}\"'.format(recon_algorithm),\n",
    "                      '--solver', '\"{}\"'.format(solver),\n",
    "                      '--output-filetypes', '\"{}\"'.format(output_filetypes)])\n",
    "    !{cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate memote reports for each  model \n",
    "\n",
    "for key,value in context_gene_exp.items():\n",
    "    print(key)\n",
    "    report_file = os.path.join(configs.datadir, \"memote_report_{}.html\".format(key))\n",
    "    model_file = os.path.join(configs.datadir, \"results\", key, \"{}_SpecificModel.xml\".format(key))\n",
    "    log_file = os.path.join(configs.datadir, \"{}_memote.log\".format(key))\n",
    "    \n",
    "    cmd = ' '.join(['memote', 'report', 'snapshot', '--filename',\n",
    "                    '\"{}\"'.format(report_file), '\"{}\"'.format(model_file), \">\", '\"{}\"'.format(log_file)])\n",
    "    !{cmd}\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Step 5: Identifying disease related genes by analyzing transcriptomics data of patients\n",
    "Differential Expression Analysis\n",
    "\n",
    "In the config_sheets folder, there should be a folder called \"disease\". You can add a spreadsheet for each cell/tissue type called `disease_data_inputs_<tissue_name>`. Each sheet of this file should correspond to a seperate disease to analyze using DGE nfor that tissue. The source data can be either microarray or bulk RNA-seq and is formatted the same as if creating the base tissue model. The sheet names should contain the disease name, an underscore, and than either \"microarray\" or \"bulk\" depending on the source data. For example, if the disease is lupus, and the source data is bulk RNA-seq, the name of the sheet should be \"lupus_bulk\". This can be seen in the example sheet. If using bulk RNA-seq data, there should be a count matrix file in `/work/data/data_matrices/<tissue_name>/disease/` called `BulkRNAseqDataMatrix_<disease name>_<tissue name>`. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*** Specify input files for step 3 here ***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "# specify tissue names to perform a disease analysis on. The diseases to analyze should be\n",
    "# specified in `/work/data/config_sheets/disease/diease_data_inputs_<tissue name>`\n",
    "tissue_names = ['Naive']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Differential gene expression analysis\n",
    "for tissue_name in tissue_names:\n",
    "    disease_config_file = \"\".join([\"disease_data_inputs_\", tissue_name, \".xlsx\"])\n",
    "    cmd = ' '.join(['python3', 'disease_analysis.py',\n",
    "                  '-t', '\"{}\"'.format(tissue_name),\n",
    "                  '-c', '\"{}\"'.format(disease_config_file)])\n",
    "    !{cmd}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 6: Identification of drug targets and repurposable drugs\n",
    "This step maps drug targets in metabolic models,prforms knock out simulation, and compare simulation results with disease genes and identifies drug targets and repurposable drugs\n",
    "\n",
    "*** Specify input files for step 4 here ***\n",
    "\n",
    "1. Instruction: A processed Drug-Target file is included in the `/root/pipelines/data/`. (Optional step) For the updated versions the users can download `Repurposing_Hub_export.txt` from [Drug Repurposing Hub](https://clue.io/repurposing-app). From the downloaded file first remove all the activators, agonists, and withdrawn drugs and then upload to to `/root/pipelines/data/`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. To use automatically created tissue specific models. Note: It is recommended to use refined and validated models for further analysis. User can define cutomized models in next sub-step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'tissue_spec_model' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mNameError\u001B[0m                                 Traceback (most recent call last)",
      "Input \u001B[0;32mIn [79]\u001B[0m, in \u001B[0;36m<cell line: 2>\u001B[0;34m()\u001B[0m\n\u001B[1;32m      1\u001B[0m \u001B[38;5;66;03m# tissue specific models\u001B[39;00m\n\u001B[0;32m----> 2\u001B[0m \u001B[43mtissue_spec_model\u001B[49m\n",
      "\u001B[0;31mNameError\u001B[0m: name 'tissue_spec_model' is not defined"
     ]
    }
   ],
   "source": [
    "# tissue specific models\n",
    "tissue_spec_model "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3. To use customized model, please specify `tissue_spec_model` manually, e.g. uncomment tissue_spec_model in the following cell."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Manually specify Up and Down Regulated Genes for Disease. (Please upload manually created files `/pipelines/data/`. Use filenames as given belwo or change them accordingly.)\n",
    "# Disease_Down = 'Disease_DOWN.txt'\n",
    "# Disease_Up = 'Disease_UP.txt'\n",
    "# drug_raw_file = 'Repurposing_Hub_export.txt'\n",
    "\n",
    "# Manually specify tissue specific models fine-tuned by user. Change names of the files accordingly. Users can use single or multiple models here. Using multiple models, simulation time will increase.\n",
    "# tissue_spec_model = {'Th1':'Th1Model.mat',\n",
    "#                      'Th2':'Th2Model.mat',\n",
    "#                      'Th17':'Th17Model.mat',\n",
    "#                      'Naive':'NaiveModel.mat'}\n",
    "\n",
    "# Manually specify tissue specific model created by matlab cobratoolbox. For example run, we have provided four models of CD4+ T cells (niave, Th1, Th2, and Th17) please uncomment all or any specific model\n",
    "# tissue_spec_model = {'Th1':'Th1_SpecificModel_matlab.mat',\n",
    "#                      'Th2':'Th2_SpecificModel_matlab.mat',\n",
    "#                      'Th17':'Th17_SpecificModel_matlab.mat',\n",
    "#                      'Naive':'Naive_SpecificModel_matlab.mat'}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Knock out simulation for the analyzed tissues and diseases\n",
    "diseases = ['lupus', 'arthritis']\n",
    "for key,value in tissue_spec_model.items():\n",
    "    for dis in diseases:\n",
    "        # load the results of step 3 to dictionary 'disease_files'\n",
    "        step3_results_file = os.path.join(configs.datadir, 'results', key, \n",
    "                                          dis, 'step2_results_files.json')\n",
    "        with open(step3_results_file) as json_file:\n",
    "            disease_files = json.load(json_file)\n",
    "        #print(disease_files)\n",
    "        Disease_Down = disease_files['DN_Reg']\n",
    "        Disease_Up = disease_files['UP_Reg']\n",
    "        drug_raw_file = 'Repurposing_Hub_export.txt'\n",
    "        \n",
    "        out_dir = os.path.join(configs.datadir, \"results\", key, dis)\n",
    "        tissueSpecificModelfile  = os.path.join(configs.datadir, \"results\", key, value)\n",
    "        print(tissueSpecificModelfile)\n",
    "        tissue_gene_folder = os.path.join(configs.datadir, key)\n",
    "        os.makedirs(tissue_gene_folder, exist_ok=True)\n",
    "        inhibitors_file = '{}_inhibitors_Entrez.txt'.format(key)\n",
    "        cmd = ' '.join(['python3' , 'knock_out_simulation.py',\n",
    "                      '-t', tissueSpecificModelfile,\n",
    "                      '-i', inhibitors_file,\n",
    "                      '-u', Disease_Up,\n",
    "                      '-d', Disease_Down,\n",
    "                      '-f', out_dir,\n",
    "                      '-r', drug_raw_file])\n",
    "        !{cmd}\n",
    "\n",
    "        # copy generated output to output folder\n",
    "        cmd = ' '.join(['cp', '-a', os.path.join(configs.datadir, key), configs.outputdir])\n",
    "        !{cmd}\n",
    "        #break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}