{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Three functional blocks\n",
    "\n",
    "1. Transcriptomics, Proteomics, Active Gene IDs\n",
    "2. Create or load tissue specific model\n",
    "3. Disease Analysis: Differential Expressions\n",
    "4. Knock out simulation: Score Genes, Generate Drug List"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/zhzhao/Dropbox/Helikar/pipelines\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import os\n",
    "import pandas\n",
    "import numpy\n",
    "import json\n",
    "from subprocess import call\n",
    "from project import configs\n",
    "print(configs.rootdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1: Transcriptomics and Proteomics Generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "transcriptomics_config_file = 'GeneExpressionDataUsed.xlsx'\n",
    "proteomics_data_file = 'ni.3693-S5.xlsx'\n",
    "proteomics_supplementary_data_file = 'Supplementary Data 1.xlsx'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Input file is \" GeneExpressionDataUsed.xlsx\n",
      "---\n",
      "Start Collecting Data for:\n",
      "['GSE22886' 'GSE43005' 'GSE22045' 'GSE24634']\n",
      "['GSM565273' 'GSM565274' 'GSM565275' 'GSM565290' 'GSM565291' 'GSM565292'\n",
      " 'GSM1054773' 'GSM1054779' 'GSM1054781' 'GSM1054789' 'GSM548000'\n",
      " 'GSM548001' 'GSM607510' 'GSM607511' 'GSM607512']\n",
      "---\n",
      "\n",
      "Initialize project (GSE22886):\n",
      "Root: /Users/zhzhao/Dropbox/Helikar/pipelines\n",
      "Raw data: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM565273.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM565274.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM565275.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM565290.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM565291.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM565292.tar\n",
      "Retrieve Samples Completed.\n",
      "GSE22886 already in database, skip over.\n",
      "Initialize project (GSE43005):\n",
      "Root: /Users/zhzhao/Dropbox/Helikar/pipelines\n",
      "Raw data: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43005_RAW\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43005_RAW/GSM1054773.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43005_RAW/GSM1054779.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43005_RAW/GSM1054781.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43005_RAW/GSM1054789.tar\n",
      "Retrieve Samples Completed.\n",
      "GSE43005 already in database, skip over.\n",
      "Initialize project (GSE22045):\n",
      "Root: /Users/zhzhao/Dropbox/Helikar/pipelines\n",
      "Raw data: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22045_RAW\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22045_RAW/GSM548000.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22045_RAW/GSM548001.tar\n",
      "Retrieve Samples Completed.\n",
      "GSE22045 already in database, skip over.\n",
      "Initialize project (GSE24634):\n",
      "Root: /Users/zhzhao/Dropbox/Helikar/pipelines\n",
      "Raw data: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE24634_RAW\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE24634_RAW/GSM607510.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE24634_RAW/GSM607511.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE24634_RAW/GSM607512.tar\n",
      "Retrieve Samples Completed.\n",
      "GSE24634 already in database, skip over.\n",
      "/usr/local/lib/python3.7/site-packages/pandas/core/ops/__init__.py:1115: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  result = method(y)\n",
      "1733 single ENTREZ_GENE_IDs to merge\n",
      "id_list: 4620, set: 3301\n",
      "entrez_single_id_list: 22318, set: 22318\n",
      "entrez_id_list: 1691, set: 1652\n",
      "dups: 2115, set: 796\n",
      "1227 id merged\n",
      "Save to /Users/zhzhao/Dropbox/Helikar/pipelines/data/transcriptomics_Naive.csv\n",
      "---\n",
      "Start Collecting Data for:\n",
      "['GSE2770' 'GSE22886' 'GSE43769' 'GSE43005' 'GSE50175' 'GSE30664']\n",
      "['GSM60353' 'GSM60354' 'GSM60360' 'GSM60367' 'GSM60377' 'GSM60381'\n",
      " 'GSM60738' 'GSM60739' 'GSM60745' 'GSM60752' 'GSM60762' 'GSM60766'\n",
      " 'GSM60704' 'GSM60705' 'GSM60711' 'GSM60718' 'GSM60728' 'GSM60732'\n",
      " 'GSM565993' 'GSM565994' 'GSM565995' 'GSM565996' 'GSM565997' 'GSM565276'\n",
      " 'GSM565277' 'GSM565278' 'GSM565279' 'GSM565280' 'GSM1070758' 'GSM1070754'\n",
      " 'GSM1070762' 'GSM1054775' 'GSM1054783' 'GSM1054787' 'GSM1054791'\n",
      " 'GSM1215107' 'GSM1215109' 'GSM1215111' 'GSM1215113' 'GSM760599'\n",
      " 'GSM760601']\n",
      "---\n",
      "\n",
      "Initialize project (GSE2770):\n",
      "Root: /Users/zhzhao/Dropbox/Helikar/pipelines\n",
      "Raw data: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60353.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60354.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60360.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60367.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60377.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60381.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60738.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60739.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60745.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60752.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60762.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60766.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60704.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60705.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60711.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60718.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60728.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60732.tar\n",
      "Retrieve Samples Completed.\n",
      "GSE2770 already in database, skip over.\n",
      "Initialize project (GSE22886):\n",
      "Root: /Users/zhzhao/Dropbox/Helikar/pipelines\n",
      "Raw data: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM565993.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM565994.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM565995.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM565996.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM565997.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM565276.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM565277.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM565278.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM565279.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM565280.tar\n",
      "Retrieve Samples Completed.\n",
      "GSE22886 already in database, skip over.\n",
      "Initialize project (GSE43769):\n",
      "Root: /Users/zhzhao/Dropbox/Helikar/pipelines\n",
      "Raw data: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43769_RAW\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43769_RAW/GSM1070758.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43769_RAW/GSM1070754.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43769_RAW/GSM1070762.tar\n",
      "Retrieve Samples Completed.\n",
      "GSE43769 already in database, skip over.\n",
      "Initialize project (GSE43005):\n",
      "Root: /Users/zhzhao/Dropbox/Helikar/pipelines\n",
      "Raw data: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43005_RAW\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43005_RAW/GSM1054775.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43005_RAW/GSM1054783.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43005_RAW/GSM1054787.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43005_RAW/GSM1054791.tar\n",
      "Retrieve Samples Completed.\n",
      "GSE43005 already in database, skip over.\n",
      "Initialize project (GSE50175):\n",
      "Root: /Users/zhzhao/Dropbox/Helikar/pipelines\n",
      "Raw data: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE50175_RAW\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE50175_RAW/GSM1215107.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE50175_RAW/GSM1215109.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE50175_RAW/GSM1215111.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE50175_RAW/GSM1215113.tar\n",
      "Retrieve Samples Completed.\n",
      "GSE50175 already in database, skip over.\n",
      "Initialize project (GSE30664):\n",
      "Root: /Users/zhzhao/Dropbox/Helikar/pipelines\n",
      "Raw data: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE30664_RAW\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE30664_RAW/GSM760599.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE30664_RAW/GSM760601.tar\n",
      "Retrieve Samples Completed.\n",
      "GSE30664 already in database, skip over.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1924 single ENTREZ_GENE_IDs to merge\n",
      "id_list: 5080, set: 3422\n",
      "entrez_single_id_list: 22524, set: 22524\n",
      "entrez_id_list: 1821, set: 1773\n",
      "dups: 2568, set: 910\n",
      "1253 id merged\n",
      "Save to /Users/zhzhao/Dropbox/Helikar/pipelines/data/transcriptomics_Th1.csv\n",
      "---\n",
      "Start Collecting Data for:\n",
      "['GSE2770' 'GSE22886' 'GSE43769' 'GSE43005']\n",
      "['GSM60357' 'GSM60358' 'GSM60362' 'GSM60368' 'GSM60373' 'GSM60379'\n",
      " 'GSM60708' 'GSM60709' 'GSM60713' 'GSM60719' 'GSM60724' 'GSM60730'\n",
      " 'GSM60742' 'GSM60743' 'GSM60747' 'GSM60753' 'GSM60758' 'GSM60764'\n",
      " 'GSM565281' 'GSM565282' 'GSM565283' 'GSM565284' 'GSM565285' 'GSM565286'\n",
      " 'GSM565998' 'GSM565999' 'GSM566000' 'GSM566001' 'GSM566002' 'GSM566003'\n",
      " 'GSM1070756' 'GSM1070760' 'GSM1070764' 'GSM1054776' 'GSM1054784'\n",
      " 'GSM1054788' 'GSM1054792']\n",
      "---\n",
      "\n",
      "Initialize project (GSE2770):\n",
      "Root: /Users/zhzhao/Dropbox/Helikar/pipelines\n",
      "Raw data: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60357.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60358.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60362.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60368.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60373.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60379.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60708.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60709.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60713.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60719.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60724.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60730.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60742.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60743.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60747.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60753.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60758.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE2770_RAW/GSM60764.tar\n",
      "Retrieve Samples Completed.\n",
      "GSE2770 already in database, skip over.\n",
      "Initialize project (GSE22886):\n",
      "Root: /Users/zhzhao/Dropbox/Helikar/pipelines\n",
      "Raw data: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM565281.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM565282.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM565283.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM565284.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM565285.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM565286.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM565998.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM565999.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM566000.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM566001.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM566002.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE22886_RAW/GSM566003.tar\n",
      "Retrieve Samples Completed.\n",
      "GSE22886 already in database, skip over.\n",
      "Initialize project (GSE43769):\n",
      "Root: /Users/zhzhao/Dropbox/Helikar/pipelines\n",
      "Raw data: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43769_RAW\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43769_RAW/GSM1070756.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43769_RAW/GSM1070760.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43769_RAW/GSM1070764.tar\n",
      "Retrieve Samples Completed.\n",
      "GSE43769 already in database, skip over.\n",
      "Initialize project (GSE43005):\n",
      "Root: /Users/zhzhao/Dropbox/Helikar/pipelines\n",
      "Raw data: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43005_RAW\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43005_RAW/GSM1054776.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43005_RAW/GSM1054784.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43005_RAW/GSM1054788.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43005_RAW/GSM1054792.tar\n",
      "Retrieve Samples Completed.\n",
      "GSE43005 already in database, skip over.\n",
      "1825 single ENTREZ_GENE_IDs to merge\n",
      "id_list: 4952, set: 3405\n",
      "entrez_single_id_list: 22329, set: 22329\n",
      "entrez_id_list: 1778, set: 1735\n",
      "dups: 2428, set: 881\n",
      "1254 id merged\n",
      "Save to /Users/zhzhao/Dropbox/Helikar/pipelines/data/transcriptomics_Th2.csv\n",
      "---\n",
      "Start Collecting Data for:\n",
      "['GSE43005' 'GSE30664' 'GSE42569' 'GSE60679']\n",
      "['GSM1054777' 'GSM1054785' 'GSM1054795' 'GSM1054793' 'GSM760596'\n",
      " 'GSM760597' 'GSM1045312' 'GSM1045314' 'GSM1484728' 'GSM1484744'\n",
      " 'GSM1484752' 'GSM1484760' 'GSM1484729' 'GSM1484745' 'GSM1484753'\n",
      " 'GSM1484761' 'GSM1484730' 'GSM1484746' 'GSM1484754' 'GSM1484762'\n",
      " 'GSM1484731' 'GSM1484747' 'GSM1484755' 'GSM1484763' 'GSM1484736'\n",
      " 'GSM1484738' 'GSM1484740' 'GSM1484742']\n",
      "---\n",
      "\n",
      "Initialize project (GSE43005):\n",
      "Root: /Users/zhzhao/Dropbox/Helikar/pipelines\n",
      "Raw data: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43005_RAW\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43005_RAW/GSM1054777.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43005_RAW/GSM1054785.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43005_RAW/GSM1054795.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43005_RAW/GSM1054793.tar\n",
      "Retrieve Samples Completed.\n",
      "GSE43005 already in database, skip over.\n",
      "Initialize project (GSE30664):\n",
      "Root: /Users/zhzhao/Dropbox/Helikar/pipelines\n",
      "Raw data: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE30664_RAW\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE30664_RAW/GSM760596.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE30664_RAW/GSM760597.tar\n",
      "Retrieve Samples Completed.\n",
      "GSE30664 already in database, skip over.\n",
      "Initialize project (GSE42569):\n",
      "Root: /Users/zhzhao/Dropbox/Helikar/pipelines\n",
      "Raw data: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE42569_RAW\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE42569_RAW/GSM1045312.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE42569_RAW/GSM1045314.tar\n",
      "Retrieve Samples Completed.\n",
      "GSE42569 already in database, skip over.\n",
      "Initialize project (GSE60679):\n",
      "Root: /Users/zhzhao/Dropbox/Helikar/pipelines\n",
      "Raw data: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484728.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484744.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484752.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484760.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484729.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484745.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484753.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484761.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484730.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484746.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484754.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484762.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484731.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484747.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484755.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484763.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484736.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484738.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484740.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484742.tar\n",
      "Retrieve Samples Completed.\n",
      "GSE60679 already in database, skip over.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1831 single ENTREZ_GENE_IDs to merge\n",
      "id_list: 4748, set: 3318\n",
      "entrez_single_id_list: 22514, set: 22514\n",
      "entrez_id_list: 1734, set: 1690\n",
      "dups: 2256, set: 826\n",
      "1234 id merged\n",
      "Save to /Users/zhzhao/Dropbox/Helikar/pipelines/data/transcriptomics_Th17.csv\n",
      "---\n",
      "Start Collecting Data for:\n",
      "['GSE43005' 'GSE60679' 'GSE4527' 'GSE24634']\n",
      "['GSM1054778' 'GSM1054786' 'GSM1054796' 'GSM1054794' 'GSM1484732'\n",
      " 'GSM1484748' 'GSM1484756' 'GSM1484764' 'GSM1484733' 'GSM1484749'\n",
      " 'GSM1484757' 'GSM1484765' 'GSM1484734' 'GSM1484750' 'GSM1484758'\n",
      " 'GSM1484766' 'GSM1484735' 'GSM1484751' 'GSM1484759' 'GSM1484767'\n",
      " 'GSM1484737' 'GSM1484739' 'GSM1484741' 'GSM1484743' 'GSM101519'\n",
      " 'GSM101521' 'GSM607516' 'GSM607517' 'GSM607518' 'GSM607528' 'GSM607529'\n",
      " 'GSM607530' 'GSM607540' 'GSM607541' 'GSM607542' 'GSM607552' 'GSM607553'\n",
      " 'GSM607554']\n",
      "---\n",
      "\n",
      "Initialize project (GSE43005):\n",
      "Root: /Users/zhzhao/Dropbox/Helikar/pipelines\n",
      "Raw data: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43005_RAW\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43005_RAW/GSM1054778.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43005_RAW/GSM1054786.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43005_RAW/GSM1054796.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE43005_RAW/GSM1054794.tar\n",
      "Retrieve Samples Completed.\n",
      "GSE43005 already in database, skip over.\n",
      "Initialize project (GSE60679):\n",
      "Root: /Users/zhzhao/Dropbox/Helikar/pipelines\n",
      "Raw data: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484732.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484748.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484756.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484764.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484733.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484749.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484757.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484765.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484734.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484750.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484758.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484766.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484735.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484751.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484759.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484767.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484737.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484739.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484741.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE60679_RAW/GSM1484743.tar\n",
      "Retrieve Samples Completed.\n",
      "GSE60679 already in database, skip over.\n",
      "Initialize project (GSE4527):\n",
      "Root: /Users/zhzhao/Dropbox/Helikar/pipelines\n",
      "Raw data: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE4527_RAW\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE4527_RAW/GSM101519.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE4527_RAW/GSM101521.tar\n",
      "Retrieve Samples Completed.\n",
      "GSE4527 already in database, skip over.\n",
      "Initialize project (GSE24634):\n",
      "Root: /Users/zhzhao/Dropbox/Helikar/pipelines\n",
      "Raw data: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE24634_RAW\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE24634_RAW/GSM607516.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE24634_RAW/GSM607517.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE24634_RAW/GSM607518.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE24634_RAW/GSM607528.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE24634_RAW/GSM607529.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE24634_RAW/GSM607530.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE24634_RAW/GSM607540.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE24634_RAW/GSM607541.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE24634_RAW/GSM607542.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE24634_RAW/GSM607552.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE24634_RAW/GSM607553.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE24634_RAW/GSM607554.tar\n",
      "Retrieve Samples Completed.\n",
      "GSE24634 already in database, skip over.\n",
      "1183 single ENTREZ_GENE_IDs to merge\n",
      "id_list: 3101, set: 2226\n",
      "entrez_single_id_list: 19446, set: 19446\n",
      "entrez_id_list: 1110, set: 1087\n",
      "dups: 1417, set: 542\n",
      "822 id merged\n",
      "Save to /Users/zhzhao/Dropbox/Helikar/pipelines/data/transcriptomics_Treg.csv\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "cmd = ' '.join(['python3', 'transcriptomic_gen.py', \n",
    "      '-i', '\"{}\"'.format(transcriptomics_config_file)])\n",
    "!{cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data file is \" ni.3693-S5.xlsx\n",
      "Supplementary Data file is \" Supplementary Data 1.xlsx\n",
      "Test Data Saved to /Users/zhzhao/Dropbox/Helikar/pipelines/data/Proteomics_Naïve.csv\n",
      "Test Data Saved to /Users/zhzhao/Dropbox/Helikar/pipelines/data/Proteomics_Th1.csv\n",
      "Test Data Saved to /Users/zhzhao/Dropbox/Helikar/pipelines/data/Proteomics_Th2.csv\n",
      "Test Data Saved to /Users/zhzhao/Dropbox/Helikar/pipelines/data/Proteomics_Th17.csv\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "cmd = ' '.join(['python3', 'proteomics_gen.py', \n",
    "      '-d', '\"{}\"'.format(proteomics_data_file), \n",
    "      '-s', '\"{}\"'.format(proteomics_supplementary_data_file)])\n",
    "!{cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Transcriptomics file is \"GeneExpressionDataUsed.xlsx\"\n",
      "Proteomics file is \"Supplementary Data 1.xlsx\"\n",
      "Read from /Users/zhzhao/Dropbox/Helikar/pipelines/data/transcriptomics_Naive.csv\n",
      "Read from /Users/zhzhao/Dropbox/Helikar/pipelines/data/transcriptomics_Th1.csv\n",
      "Read from /Users/zhzhao/Dropbox/Helikar/pipelines/data/transcriptomics_Th2.csv\n",
      "Read from /Users/zhzhao/Dropbox/Helikar/pipelines/data/transcriptomics_Th17.csv\n",
      "Read from /Users/zhzhao/Dropbox/Helikar/pipelines/data/transcriptomics_Treg.csv\n",
      "Test Data Load From /Users/zhzhao/Dropbox/Helikar/pipelines/data/Proteomics_Naïve.csv\n",
      "Test Data Load From /Users/zhzhao/Dropbox/Helikar/pipelines/data/Proteomics_Th1.csv\n",
      "Test Data Load From /Users/zhzhao/Dropbox/Helikar/pipelines/data/Proteomics_Th2.csv\n",
      "Test Data Load From /Users/zhzhao/Dropbox/Helikar/pipelines/data/Proteomics_Th17.csv\n",
      "(21850, 2)\n",
      "(9150, 2)\n",
      "699 single ENTREZ_GENE_IDs to merge\n",
      "id_list: 3536, set: 3429\n",
      "entrez_single_id_list: 21338, set: 21337\n",
      "entrez_id_list: 1263, set: 1261\n",
      "dups: 212, set: 105\n",
      "1233 id merged\n",
      "Th1: save to /Users/zhzhao/Dropbox/Helikar/pipelines/data/merged_Th1.csv\n",
      "\n",
      "(21808, 2)\n",
      "(9150, 2)\n",
      "676 single ENTREZ_GENE_IDs to merge\n",
      "id_list: 3394, set: 3308\n",
      "entrez_single_id_list: 21302, set: 21301\n",
      "entrez_id_list: 1236, set: 1234\n",
      "dups: 170, set: 84\n",
      "1210 id merged\n",
      "Naive: save to /Users/zhzhao/Dropbox/Helikar/pipelines/data/merged_Naive.csv\n",
      "\n",
      "(21753, 2)\n",
      "(9150, 2)\n",
      "698 single ENTREZ_GENE_IDs to merge\n",
      "id_list: 3511, set: 3412\n",
      "entrez_single_id_list: 21243, set: 21242\n",
      "entrez_id_list: 1262, set: 1260\n",
      "dups: 196, set: 97\n",
      "1231 id merged\n",
      "Th2: save to /Users/zhzhao/Dropbox/Helikar/pipelines/data/merged_Th2.csv\n",
      "\n",
      "(21912, 2)\n",
      "(9150, 2)\n",
      "677 single ENTREZ_GENE_IDs to merge\n",
      "id_list: 3412, set: 3325\n",
      "entrez_single_id_list: 21399, set: 21398\n",
      "entrez_id_list: 1242, set: 1240\n",
      "dups: 172, set: 85\n",
      "1213 id merged\n",
      "Th17: save to /Users/zhzhao/Dropbox/Helikar/pipelines/data/merged_Th17.csv\n",
      "\n",
      "0 single ENTREZ_GENE_IDs to merge\n",
      "id_list: 2242, set: 2226\n",
      "entrez_single_id_list: 18263, set: 18263\n",
      "entrez_id_list: 821, set: 821\n",
      "dups: 32, set: 16\n",
      "813 id merged\n",
      "Treg: save to /Users/zhzhao/Dropbox/Helikar/pipelines/data/merged_Treg.csv\n",
      "\n",
      "{'Th1': '/Users/zhzhao/Dropbox/Helikar/pipelines/data/merged_Th1.csv', 'Naive': '/Users/zhzhao/Dropbox/Helikar/pipelines/data/merged_Naive.csv', 'Th2': '/Users/zhzhao/Dropbox/Helikar/pipelines/data/merged_Th2.csv', 'Th17': '/Users/zhzhao/Dropbox/Helikar/pipelines/data/merged_Th17.csv', 'Treg': '/Users/zhzhao/Dropbox/Helikar/pipelines/data/merged_Treg.csv'}\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "cmd = ' '.join(['python3', 'merge_xomics.py', \n",
    "      '-t', '\"{}\"'.format(transcriptomics_config_file), \n",
    "      '-p', '\"{}\"'.format(proteomics_supplementary_data_file)])\n",
    "!{cmd}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1.5 Create Tissue Specific Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Th1': '/Users/zhzhao/Dropbox/Helikar/pipelines/data/merged_Th1.csv', 'Naive': '/Users/zhzhao/Dropbox/Helikar/pipelines/data/merged_Naive.csv', 'Th2': '/Users/zhzhao/Dropbox/Helikar/pipelines/data/merged_Th2.csv', 'Th17': '/Users/zhzhao/Dropbox/Helikar/pipelines/data/merged_Th17.csv', 'Treg': '/Users/zhzhao/Dropbox/Helikar/pipelines/data/merged_Treg.csv'}\n"
     ]
    }
   ],
   "source": [
    "step1_results_file = os.path.join(configs.datadir, 'step1_results_files.json')\n",
    "with open(step1_results_file) as json_file:\n",
    "    tissue_gene_exp = json.load(json_file)\n",
    "print(tissue_gene_exp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# (input) filename of General Model \n",
    "GeneralModelFile = 'GeneralModel.mat'\n",
    "# (input) filename of Tissue Gene Expression\n",
    "genefile = 'merged_Th1.csv'\n",
    "# (output) filename of Tissue Specific Model\n",
    "tissuefile = 'Th1_SpecificModel.mat'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "General Model file is \" GeneralModel.mat\n",
      "Gene Expression file is \" merged_Th1.csv\n",
      "Output file is \" Th1_SpecificModel.mat\n",
      "/Users/zhzhao/Dropbox/Helikar/pipelines\n",
      "(20638, 2)\n",
      "(1233, 2)\n",
      "(3429, 2)\n",
      "(24067, 2)\n",
      "Map gene expression to reactions, 0 errors.\n",
      "1's: 7582\n",
      "2's: 0\n",
      "Teff\n",
      "Genes: 1108\n",
      "Metabolites: 5507\n",
      "Reactions: 7582\n",
      "1.0*biomass_reaction_Mphage - 1.0*biomass_reaction_Mphage_reverse_6cff5\n",
      "<Solution 0.000 at 0x18292b9e8>\n",
      "General Model file is \"GeneralModel.mat\"\n",
      "Gene Expression file is \"merged_Naive.csv\"\n",
      "Output file is \"Naive_SpecificModel.mat\"\n",
      "(20625, 2)\n",
      "(1210, 2)\n",
      "(3321, 2)\n",
      "(23946, 2)\n",
      "Map gene expression to reactions, 0 errors.\n",
      "1's: 7494\n",
      "2's: 0\n",
      "Teff\n",
      "Genes: 1069\n",
      "Metabolites: 5457\n",
      "Reactions: 7494\n",
      "1.0*biomass_reaction_Mphage - 1.0*biomass_reaction_Mphage_reverse_6cff5\n",
      "<Solution 0.000 at 0x17b43e6a0>\n",
      "General Model file is \"GeneralModel.mat\"\n",
      "Gene Expression file is \"merged_Th2.csv\"\n",
      "Output file is \"Th2_SpecificModel.mat\"\n",
      "(20544, 2)\n",
      "(1231, 2)\n",
      "(3428, 2)\n",
      "(23972, 2)\n",
      "Map gene expression to reactions, 0 errors.\n",
      "1's: 7059\n",
      "2's: 0\n",
      "Teff\n",
      "Genes: 1083\n",
      "Metabolites: 5468\n",
      "Reactions: 7059\n",
      "1.0*biomass_reaction_Mphage - 1.0*biomass_reaction_Mphage_reverse_6cff5\n",
      "<Solution 0.000 at 0x180f73630>\n",
      "General Model file is \"GeneralModel.mat\"\n",
      "Gene Expression file is \"merged_Th17.csv\"\n",
      "Output file is \"Th17_SpecificModel.mat\"\n",
      "(20721, 2)\n",
      "(1213, 2)\n",
      "(3339, 2)\n",
      "(24060, 2)\n",
      "Map gene expression to reactions, 0 errors.\n",
      "1's: 7541\n",
      "2's: 0\n",
      "Teff\n",
      "Genes: 1105\n",
      "Metabolites: 5527\n",
      "Reactions: 7541\n",
      "1.0*biomass_reaction_Mphage - 1.0*biomass_reaction_Mphage_reverse_6cff5\n",
      "<Solution 0.000 at 0x18adbdb70>\n",
      "General Model file is \"GeneralModel.mat\"\n",
      "Gene Expression file is \"merged_Treg.csv\"\n",
      "Output file is \"Treg_SpecificModel.mat\"\n",
      "(18263, 2)\n",
      "(813, 2)\n",
      "(2229, 2)\n",
      "(20492, 2)\n",
      "Map gene expression to reactions, 0 errors.\n",
      "1's: 7472\n",
      "2's: 36\n",
      "Teff\n",
      "Genes: 1103\n",
      "Metabolites: 5433\n",
      "Reactions: 7508\n",
      "1.0*biomass_reaction_Mphage - 1.0*biomass_reaction_Mphage_reverse_6cff5\n",
      "<Solution 0.000 at 0x17f2ebac8>\n",
      "{'Th1': 'Th1_SpecificModel.mat', 'Naive': 'Naive_SpecificModel.mat', 'Th2': 'Th2_SpecificModel.mat', 'Th17': 'Th17_SpecificModel.mat', 'Treg': 'Treg_SpecificModel.mat'}\n"
     ]
    }
   ],
   "source": [
    "tissue_spec_model = {}\n",
    "for key,value in tissue_gene_exp.items():\n",
    "    tissuefile = '{}_SpecificModel.mat'.format(key)\n",
    "    tissue_spec_model[key] = tissuefile\n",
    "    tissue_gene_file = value.split('/')[-1]\n",
    "    tissue_gene_folder = os.path.join(configs.datadir, key)\n",
    "    os.makedirs(tissue_gene_folder, exist_ok=False)\n",
    "    cmd = ' '.join(['python3', 'create_tissue_specific_model.py', \n",
    "                      '-m', '\"{}\"'.format(GeneralModelFile), \n",
    "                      '-g', '\"{}\"'.format(tissue_gene_file),\n",
    "                      '-o', '\"{}\"'.format(tissuefile)])\n",
    "    !{cmd}\n",
    "\n",
    "print(tissue_spec_model)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2: Disease Gene Analysis\n",
    "Differential Expression Analysis\n",
    "\n",
    "Only 1 disease to be analyzed, output files in data folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "disease_gene_file = 'Disease_Gene_Analyzed.xlsx'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Input file is \" Disease_Gene_Analyzed.xlsx\n",
      "Initialize project (GSE21942):\n",
      "Root: /Users/zhzhao/Dropbox/Helikar/pipelines\n",
      "Raw data: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE21942_RAW\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE21942_RAW/GSM1066005.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE21942_RAW/GSM1066006.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE21942_RAW/GSM1066007.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE21942_RAW/GSM1066008.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE21942_RAW/GSM1066009.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE21942_RAW/GSM1066010.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE21942_RAW/GSM1066011.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE21942_RAW/GSM1066012.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE21942_RAW/GSM1066013.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE21942_RAW/GSM1066014.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE21942_RAW/GSM1066015.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE21942_RAW/GSM1066016.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE21942_RAW/GSM1066017.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE21942_RAW/GSM1066018.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE21942_RAW/GSM1066019.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE21942_RAW/GSM1066020.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE21942_RAW/GSM1066021.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE21942_RAW/GSM1066022.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE21942_RAW/GSM1066023.tar\n",
      "Sample exist: /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE21942_RAW/GSM1066024.tar\n",
      "Retrieve Samples Completed.\n",
      "GPL570:affy, /Users/zhzhao/Dropbox/Helikar/pipelines/data/GSE21942_RAW/GPL570\n",
      "R[write to console]: \n",
      "\n",
      "Background correcting\n",
      "Normalizing\n",
      "Calculating Expression\n",
      "\u001b[32mINFO    [bioservices:BioDBNet]: \u001b[0m \u001b[34mInitialising BioDBNet service (REST)\u001b[0m\n",
      "retrieve 0:76\n",
      "disease_analysis.py:17: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n",
      "  RA_UP['Gene ID'].to_csv(outputFullPath, index=False)\n",
      "\u001b[32mINFO    [bioservices:BioDBNet]: \u001b[0m \u001b[34mInitialising BioDBNet service (REST)\u001b[0m\n",
      "retrieve 0:81\n",
      "                 Gene ID                                    Ensembl Gene ID\n",
      "Affy ID                                                                    \n",
      "210387_at     8970//8339                ENSG00000124635//OTTHUMG00000014446\n",
      "205898_at           1524                ENSG00000168329//OTTHUMG00000156249\n",
      "1564139_at        144571                ENSG00000245105//OTTHUMG00000168289\n",
      "212070_at           9289                OTTHUMG00000176742//ENSG00000205336\n",
      "215894_at           5729                ENSG00000168229//OTTHUMG00000140299\n",
      "...                  ...                                                ...\n",
      "204170_s_at         1164                OTTHUMG00000020180//ENSG00000123975\n",
      "230337_at           6654                OTTHUMG00000102109//ENSG00000115904\n",
      "206082_at          10866  ENSG00000206337//ENSG00000230389//OTTHUMG00000...\n",
      "1555486_a_at       79899                ENSG00000135362//OTTHUMG00000166334\n",
      "57715_at           51063                OTTHUMG00000018990//ENSG00000138172\n",
      "\n",
      "[63 rows x 2 columns]\n",
      "            Gene ID                      Ensembl Gene ID\n",
      "Affy ID                                                 \n",
      "204439_at     10964  ENSG00000137959//OTTHUMG00000009724\n",
      "202388_at      5997  OTTHUMG00000035600//ENSG00000116741\n",
      "202917_s_at    6279  ENSG00000143546//OTTHUMG00000013124\n",
      "204959_at      4332  ENSG00000163563//OTTHUMG00000022776\n",
      "202859_x_at    3576  ENSG00000169429//OTTHUMG00000151316\n",
      "...             ...                                  ...\n",
      "232304_at     57162                      ENSG00000197329\n",
      "203455_s_at    6303  ENSG00000130066//OTTHUMG00000021256\n",
      "1568665_at     7844  OTTHUMG00000153197//ENSG00000239305\n",
      "214370_at      6279                      ENSG00000143546\n",
      "202446_s_at    5359  OTTHUMG00000159427//ENSG00000188313\n",
      "\n",
      "[70 rows x 2 columns]\n",
      "\u001b[32mINFO    [bioservices:BioDBNet]: \u001b[0m \u001b[34mInitialising BioDBNet service (REST)\u001b[0m\n",
      "retrieve 0:500\n",
      "retrieve 500:1000\n",
      "retrieve 1000:1500\n",
      "retrieve 1500:2000\n",
      "retrieve 2000:2500\n",
      "retrieve 2500:3000\n",
      "retrieve 3000:3500\n",
      "retrieve 3500:4000\n",
      "retrieve 4000:4500\n",
      "retrieve 4500:5000\n",
      "retrieve 5000:5500\n",
      "retrieve 5500:6000\n",
      "retrieve 6000:6500\n",
      "retrieve 6500:7000\n",
      "retrieve 7000:7500\n",
      "retrieve 7500:8000\n",
      "retrieve 8000:8500\n",
      "retrieve 8500:9000\n",
      "retrieve 9000:9500\n",
      "retrieve 9500:10000\n",
      "retrieve 10000:10500\n",
      "retrieve 10500:11000\n",
      "retrieve 11000:11500\n",
      "retrieve 11500:12000\n",
      "retrieve 12000:12500\n",
      "retrieve 12500:13000\n",
      "retrieve 13000:13500\n",
      "retrieve 13500:14000\n",
      "retrieve 14000:14500\n",
      "retrieve 14500:15000\n",
      "retrieve 15000:15500\n",
      "retrieve 15500:16000\n",
      "retrieve 16000:16500\n",
      "retrieve 16500:17000\n",
      "retrieve 17000:17500\n",
      "retrieve 17500:18000\n",
      "retrieve 18000:18500\n",
      "retrieve 18500:19000\n",
      "retrieve 19000:19500\n",
      "retrieve 19500:20000\n",
      "retrieve 20000:20500\n",
      "retrieve 20500:21000\n",
      "retrieve 21000:21500\n",
      "retrieve 21500:22000\n",
      "retrieve 22000:22500\n",
      "retrieve 22500:23000\n",
      "retrieve 23000:23500\n",
      "retrieve 23500:24000\n",
      "retrieve 24000:24500\n",
      "retrieve 24500:25000\n",
      "retrieve 25000:25500\n",
      "retrieve 25500:26000\n",
      "retrieve 26000:26500\n",
      "retrieve 26500:27000\n",
      "retrieve 27000:27500\n",
      "retrieve 27500:28000\n",
      "retrieve 28000:28500\n",
      "retrieve 28500:29000\n",
      "retrieve 29000:29500\n",
      "retrieve 29500:30000\n",
      "retrieve 30000:30500\n",
      "retrieve 30500:31000\n",
      "retrieve 31000:31500\n",
      "retrieve 31500:32000\n",
      "retrieve 32000:32500\n",
      "retrieve 32500:33000\n",
      "retrieve 33000:33500\n",
      "retrieve 33500:34000\n",
      "retrieve 34000:34500\n",
      "retrieve 34500:35000\n",
      "retrieve 35000:35500\n",
      "retrieve 35500:36000\n",
      "retrieve 36000:36500\n",
      "retrieve 36500:37000\n",
      "retrieve 37000:37500\n",
      "retrieve 37500:38000\n",
      "retrieve 38000:38500\n",
      "retrieve 38500:39000\n",
      "retrieve 39000:39500\n",
      "retrieve 39500:40000\n",
      "retrieve 40000:40500\n",
      "retrieve 40500:41000\n",
      "retrieve 41000:41500\n",
      "retrieve 41500:42000\n",
      "retrieve 42000:42500\n",
      "retrieve 42500:43000\n",
      "retrieve 43000:43500\n",
      "retrieve 43500:44000\n",
      "retrieve 44000:44500\n",
      "retrieve 44500:45000\n",
      "retrieve 45000:45500\n",
      "retrieve 45500:46000\n",
      "retrieve 46000:46500\n",
      "retrieve 46500:47000\n",
      "retrieve 47000:47500\n",
      "retrieve 47500:48000\n",
      "retrieve 48000:48500\n",
      "retrieve 48500:49000\n",
      "retrieve 49000:49500\n",
      "retrieve 49500:50000\n",
      "retrieve 50000:50500\n",
      "retrieve 50500:51000\n",
      "retrieve 51000:51500\n",
      "retrieve 51500:52000\n",
      "retrieve 52000:52500\n",
      "retrieve 52500:53000\n",
      "retrieve 53000:53500\n",
      "retrieve 53500:54000\n",
      "retrieve 54000:54500\n",
      "retrieve 54500:54675\n",
      "Raw Data saved to\n",
      "/Users/zhzhao/Dropbox/Helikar/pipelines/data/Raw_Fit_GSE21942.csv\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "cmd = ' '.join(['python3', 'disease_analysis.py', \n",
    "              '-i', '\"{}\"'.format(disease_gene_file)])\n",
    "!{cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'GSE': 'GSE21942', 'UP_Reg': '/Users/zhzhao/Dropbox/Helikar/pipelines/data/RA_UP_GSE21942.txt', 'DN_Reg': '/Users/zhzhao/Dropbox/Helikar/pipelines/data/RA_DOWN_GSE21942.txt', 'RAW_Data': '/Users/zhzhao/Dropbox/Helikar/pipelines/data/Raw_Fit_GSE21942.csv'}\n"
     ]
    }
   ],
   "source": [
    "step2_results_file = os.path.join(configs.datadir, 'step2_results_files.json')\n",
    "with open(step2_results_file) as json_file:\n",
    "    disease_files = json.load(json_file)\n",
    "print(disease_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 3: Knock Out Simulation\n",
    "Knock out simulation and repurposing drug List generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# datafolder = 'Th1'\n",
    "# tissueSpecificModelfile = 'Th1_Cell_SpecificModel4manuscript.mat'\n",
    "# inhibitors_file = 'Th1_inhibitors_Entrez.txt'\n",
    "\n",
    "# RA_Down = 'RA_DOWN.txt'\n",
    "# RA_Up = 'RA_UP.txt'\n",
    "# drug_raw_file = 'Repurposing_Hub_export.txt'\n",
    "# # drug_pre_file = 'Repurposing_Hub_Preproc.csv'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cmd = ' '.join(['python3' , 'knock_out_simulation.py',\n",
    "#               '-t', tissueSpecificModelfile,\n",
    "#               '-i', inhibitors_file,\n",
    "#               '-u', RA_Up,\n",
    "#               '-d', RA_Down,\n",
    "#               '-f', datafolder,\n",
    "#               '-r', drug_raw_file])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Th1': 'Th1_SpecificModel.mat',\n",
       " 'Naive': 'Naive_SpecificModel.mat',\n",
       " 'Th2': 'Th2_SpecificModel.mat',\n",
       " 'Th17': 'Th17_SpecificModel.mat',\n",
       " 'Treg': 'Treg_SpecificModel.mat'}"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tissue_spec_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "RA_Down = disease_files['DN_Reg']\n",
    "RA_Up = disease_files['UP_Reg']\n",
    "\n",
    "drug_raw_file = 'Repurposing_Hub_export.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for key,value in tissue_spec_model.items():\n",
    "    tissueSpecificModelfile = value\n",
    "    tissue_gene_folder = os.path.join(configs.datadir, key)\n",
    "    os.makedirs(tissue_gene_folder, exist_ok=False)\n",
    "    inhibitors_file = '{}_inhibitors_Entrez.txt'.format(key)\n",
    "    cmd = ' '.join(['python3' , 'knock_out_simulation.py',\n",
    "                  '-t', tissueSpecificModelfile,\n",
    "                  '-i', inhibitors_file,\n",
    "                  '-u', RA_Up,\n",
    "                  '-d', RA_Down,\n",
    "                  '-f', key,\n",
    "                  '-r', drug_raw_file])\n",
    "    !{cmd}\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
