{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import time\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "# from bioservices import BioDBNet\n",
    "from transcriptomic_gen import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load Proteomics\n",
    "datafilename = 'ni.3693-S5.xlsx'\n",
    "data_sheet = list(range(1)) # first 5 sheets\n",
    "dataFullPath = os.path.join(projectdir, 'data', datafilename)\n",
    "fulldata = pd.read_excel(dataFullPath, sheet_name=data_sheet, header=0)\n",
    "proteomics_data = fulldata[0]\n",
    "\n",
    "# Preprocess data, drop na, duplicate ';' in Gene names, \n",
    "proteomics_data['Gene names'] = proteomics_data['Gene names'].astype(str)\n",
    "proteomics_data.dropna(subset=['Gene names'],inplace=True)\n",
    "pluralnames = proteomics_data[proteomics_data['Gene names'].str.contains(';')==True]\n",
    "for idx, row in pluralnames.iterrows():\n",
    "    # print('{}: {}'.format(idx, row['Gene names']))\n",
    "    names = row['Gene names'].split(';')\n",
    "    rows = []\n",
    "    for name in names:\n",
    "        rowcopy = row.copy()\n",
    "        rowcopy['Gene names'] = name\n",
    "        rows.append(rowcopy)\n",
    "    proteomics_data.drop(index=idx,inplace=True)\n",
    "    proteomics_data.append(rows, ignore_index=True)\n",
    "\n",
    "proteomics_data.rename(columns={'Gene names':'Gene Symbol'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "suppfilename = 'Supplementary Data 1.xlsx'\n",
    "suppFullPath = os.path.join(projectdir, 'data', suppfilename)\n",
    "supp_sheet = ['Proteomics']\n",
    "supplements = pd.read_excel(suppFullPath, sheet_name=supp_sheet, header=0)\n",
    "Proteomics = supplements['Proteomics']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read From Saved Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteomics_data = pd.read_csv(os.path.join(projectdir,'data','proteomics_data.csv'),index_col='ENTREZ_GENE_ID')\n",
    "sym2id = pd.read_csv(os.path.join(projectdir,'data','proteomics_entrez_map.csv'), index_col='Gene Symbol')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO    [bioservices:BioDBNet]:  Initialising BioDBNet service (REST)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "retrieve 0:500\n",
      "retrieve 500:1000\n",
      "retrieve 1000:1500\n",
      "retrieve 1500:2000\n",
      "retrieve 2000:2500\n",
      "retrieve 2500:3000\n",
      "retrieve 3000:3500\n",
      "retrieve 3500:4000\n",
      "retrieve 4000:4500\n",
      "retrieve 4500:5000\n",
      "retrieve 5000:5500\n",
      "retrieve 5500:6000\n",
      "retrieve 6000:6500\n",
      "retrieve 6500:7000\n",
      "retrieve 7000:7500\n",
      "retrieve 7500:8000\n",
      "retrieve 8000:8500\n",
      "retrieve 8500:9000\n",
      "retrieve 9000:9500\n",
      "retrieve 9500:10000\n",
      "retrieve 10000:10264\n"
     ]
    }
   ],
   "source": [
    "sym2id = fetch_entrez_gene_id(proteomics_data['Gene Symbol'].tolist(), input_db='Gene Symbol')\n",
    "sym2id.loc[sym2id['Gene ID']=='-', ['Gene ID']] = np.nan\n",
    "\n",
    "proteomics_data.set_index('Gene Symbol', inplace=True)\n",
    "proteomics_data['ENTREZ_GENE_ID'] = sym2id['Gene ID']\n",
    "proteomics_data.dropna(subset=['ENTREZ_GENE_ID'], inplace=True)\n",
    "proteomics_data.set_index('ENTREZ_GENE_ID', inplace=True)\n",
    "proteomics_data.drop(columns=['Majority protein IDs'], inplace=True)\n",
    "\n",
    "proteomics_data.to_csv(os.path.join(projectdir,'data','proteomics_data.csv'),index_label='ENTREZ_GENE_ID')\n",
    "sym2id.to_csv(os.path.join(projectdir,'data','proteomics_entrez_map.csv'), index_label='Gene Symbol')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteomics_data.index.name='ENTREZ_GENE_ID'\n",
    "proteomics_data.to_csv(os.path.join(projectdir,'data','proteomics_data.csv'),index_label='ENTREZ_GENE_ID')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Data by Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test Data Saved to /Users/zhzhao/Dropbox/Helikar/pipelines/data/Proteomics_Naïve.csv\n",
      "Test Data Saved to /Users/zhzhao/Dropbox/Helikar/pipelines/data/Proteomics_Th1.csv\n",
      "Test Data Saved to /Users/zhzhao/Dropbox/Helikar/pipelines/data/Proteomics_Th2.csv\n",
      "Test Data Saved to /Users/zhzhao/Dropbox/Helikar/pipelines/data/Proteomics_Th17.csv\n"
     ]
    }
   ],
   "source": [
    "tests = []\n",
    "files = []\n",
    "datas = []\n",
    "for test in list(Proteomics):\n",
    "    if test == 'Statistics':\n",
    "        break\n",
    "    cols = Proteomics[test].dropna().tolist()\n",
    "    testdata = proteomics_data.loc[:,cols]\n",
    "    # Logical Calculation\n",
    "    thresholds = testdata.quantile(0.75, axis=0)\n",
    "    testbool = pd.DataFrame(0,columns=list(testdata),index=testdata.index)\n",
    "    for col in list(testdata):\n",
    "        testbool.loc[testdata[col] > thresholds[col], [col]] = 1\n",
    "\n",
    "    testdata['pos'] = (testdata > 0).sum(axis=1)/testdata.count(axis=1)\n",
    "    testdata['0.5'] = 0\n",
    "    testdata.loc[(testdata['pos'] > 0.5),['0.5']] = 1\n",
    "    testdata['top 0.25'] = 0\n",
    "    testdata.loc[testbool.any(axis=1),'top 0.25'] = 1\n",
    "    \n",
    "    output_path = os.path.join(projectdir,'data','Proteomics_{}.csv'.format(test.strip()))\n",
    "    testdata.to_csv(output_path, index_label='ENTREZ_GENE_ID')\n",
    "    print('Test Data Saved to {}'.format(output_path))\n",
    "    tests.append(test.strip())\n",
    "    files.append(output_path)\n",
    "    datas.append(testdata)\n",
    "    \n",
    "protemoics_dict = dict(zip(tests,files))\n",
    "testdata_dict = dict(zip(tests,datas))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test Data Load From /Users/zhzhao/Dropbox/Helikar/pipelines/data/Proteomics_Naïve.csv\n",
      "Test Data Load From /Users/zhzhao/Dropbox/Helikar/pipelines/data/Proteomics_Th1.csv\n",
      "Test Data Load From /Users/zhzhao/Dropbox/Helikar/pipelines/data/Proteomics_Th2.csv\n",
      "Test Data Load From /Users/zhzhao/Dropbox/Helikar/pipelines/data/Proteomics_Th17.csv\n"
     ]
    }
   ],
   "source": [
    "# Read data from csv files\n",
    "tests = []\n",
    "datas = []\n",
    "for test in list(Proteomics):\n",
    "    if test == 'Statistics':\n",
    "        break\n",
    "    cols = Proteomics[test].dropna().tolist()\n",
    "    output_path = os.path.join(projectdir,'data','Proteomics_{}.csv'.format(test.strip()))\n",
    "    testdata = pd.read_csv(output_path, index_col='ENTREZ_GENE_ID')\n",
    "    print('Test Data Load From {}'.format(output_path))\n",
    "    tests.append(test.strip().replace('ï','i'))\n",
    "    datas.append(testdata)\n",
    "    \n",
    "proteomics_dict = dict(zip(tests,datas))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "testdata_bool = testdata > 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Naive':                 CopyNumber_T4.naive_01_activated  \\\n",
       " ENGREZ_GENE_ID                                     \n",
       " 5269                                           0   \n",
       " 1048                                           0   \n",
       " 8828                                          97   \n",
       " 3069                                      142088   \n",
       " 11309                                          0   \n",
       " ...                                          ...   \n",
       " 25912                                          0   \n",
       " 9712                                         786   \n",
       " 10284                                      76249   \n",
       " 6929                                        3364   \n",
       " 23085                                      16283   \n",
       " \n",
       "                 CopyNumber_T4.naive_02_activated  \\\n",
       " ENGREZ_GENE_ID                                     \n",
       " 5269                                           0   \n",
       " 1048                                           0   \n",
       " 8828                                           0   \n",
       " 3069                                      178332   \n",
       " 11309                                          0   \n",
       " ...                                          ...   \n",
       " 25912                                          0   \n",
       " 9712                                           0   \n",
       " 10284                                      48262   \n",
       " 6929                                           0   \n",
       " 23085                                      11563   \n",
       " \n",
       "                 CopyNumber_T4.naive_03_activated  \\\n",
       " ENGREZ_GENE_ID                                     \n",
       " 5269                                           0   \n",
       " 1048                                           0   \n",
       " 8828                                         416   \n",
       " 3069                                      134839   \n",
       " 11309                                          0   \n",
       " ...                                          ...   \n",
       " 25912                                          0   \n",
       " 9712                                           0   \n",
       " 10284                                     193409   \n",
       " 6929                                           0   \n",
       " 23085                                       5043   \n",
       " \n",
       "                 CopyNumber_T4.naive_04_activated  \\\n",
       " ENGREZ_GENE_ID                                     \n",
       " 5269                                           0   \n",
       " 1048                                           0   \n",
       " 8828                                         178   \n",
       " 3069                                      126964   \n",
       " 11309                                          0   \n",
       " ...                                          ...   \n",
       " 25912                                          0   \n",
       " 9712                                           0   \n",
       " 10284                                      65768   \n",
       " 6929                                        4778   \n",
       " 23085                                       3220   \n",
       " \n",
       "                 CopyNumber_T4.naive_01_steady-state  \\\n",
       " ENGREZ_GENE_ID                                        \n",
       " 5269                                              0   \n",
       " 1048                                              0   \n",
       " 8828                                              0   \n",
       " 3069                                          27078   \n",
       " 11309                                             0   \n",
       " ...                                             ...   \n",
       " 25912                                             0   \n",
       " 9712                                              0   \n",
       " 10284                                         89578   \n",
       " 6929                                           4654   \n",
       " 23085                                          1735   \n",
       " \n",
       "                 CopyNumber_T4.naive_02_steady-state  \\\n",
       " ENGREZ_GENE_ID                                        \n",
       " 5269                                              0   \n",
       " 1048                                              0   \n",
       " 8828                                              0   \n",
       " 3069                                          29479   \n",
       " 11309                                             0   \n",
       " ...                                             ...   \n",
       " 25912                                             0   \n",
       " 9712                                              0   \n",
       " 10284                                        105993   \n",
       " 6929                                              0   \n",
       " 23085                                          2362   \n",
       " \n",
       "                 CopyNumber_T4.naive_03_steady-state  \\\n",
       " ENGREZ_GENE_ID                                        \n",
       " 5269                                              0   \n",
       " 1048                                              0   \n",
       " 8828                                              0   \n",
       " 3069                                          28525   \n",
       " 11309                                             0   \n",
       " ...                                             ...   \n",
       " 25912                                             0   \n",
       " 9712                                              0   \n",
       " 10284                                         88649   \n",
       " 6929                                              0   \n",
       " 23085                                          5009   \n",
       " \n",
       "                 CopyNumber_T4.naive_04_steady-state    pos  0.5  top 0.25  \n",
       " ENGREZ_GENE_ID                                                             \n",
       " 5269                                              0  0.000    0         0  \n",
       " 1048                                              0  0.000    0         0  \n",
       " 8828                                              0  0.375    0         0  \n",
       " 3069                                          38335  1.000    1         1  \n",
       " 11309                                             0  0.000    0         0  \n",
       " ...                                             ...    ...  ...       ...  \n",
       " 25912                                             0  0.000    0         0  \n",
       " 9712                                              0  0.125    0         0  \n",
       " 10284                                        113089  1.000    1         1  \n",
       " 6929                                              0  0.375    0         0  \n",
       " 23085                                          4579  1.000    1         0  \n",
       " \n",
       " [9839 rows x 11 columns],\n",
       " 'Th1':                 CopyNumber_Th1_01_steady-state  \\\n",
       " ENGREZ_GENE_ID                                   \n",
       " 5269                                         0   \n",
       " 1048                                         0   \n",
       " 8828                                         0   \n",
       " 3069                                     29000   \n",
       " 11309                                        0   \n",
       " ...                                        ...   \n",
       " 25912                                        0   \n",
       " 9712                                       778   \n",
       " 10284                                   256556   \n",
       " 6929                                      2267   \n",
       " 23085                                     1622   \n",
       " \n",
       "                 CopyNumber_Th1_02_steady-state  \\\n",
       " ENGREZ_GENE_ID                                   \n",
       " 5269                                         0   \n",
       " 1048                                      2559   \n",
       " 8828                                         0   \n",
       " 3069                                     39292   \n",
       " 11309                                        0   \n",
       " ...                                        ...   \n",
       " 25912                                        0   \n",
       " 9712                                      1179   \n",
       " 10284                                   162096   \n",
       " 6929                                         0   \n",
       " 23085                                      685   \n",
       " \n",
       "                 CopyNumber_Th1_03_steady-state  \\\n",
       " ENGREZ_GENE_ID                                   \n",
       " 5269                                         0   \n",
       " 1048                                         0   \n",
       " 8828                                      1989   \n",
       " 3069                                     34518   \n",
       " 11309                                        0   \n",
       " ...                                        ...   \n",
       " 25912                                        0   \n",
       " 9712                                       629   \n",
       " 10284                                   214902   \n",
       " 6929                                      6737   \n",
       " 23085                                     1673   \n",
       " \n",
       "                 CopyNumber_Th1_04_steady-state   pos  0.5  top 0.25  \n",
       " ENGREZ_GENE_ID                                                       \n",
       " 5269                                         0  0.00    0         0  \n",
       " 1048                                      3360  0.50    0         0  \n",
       " 8828                                         0  0.25    0         0  \n",
       " 3069                                     24917  1.00    1         0  \n",
       " 11309                                        0  0.00    0         0  \n",
       " ...                                        ...   ...  ...       ...  \n",
       " 25912                                        0  0.00    0         0  \n",
       " 9712                                       751  1.00    1         0  \n",
       " 10284                                   105698  1.00    1         1  \n",
       " 6929                                         0  0.50    0         0  \n",
       " 23085                                     4901  1.00    1         0  \n",
       " \n",
       " [9839 rows x 7 columns],\n",
       " 'Th2':                 CopyNumber_Th2_01_steady-state  \\\n",
       " ENGREZ_GENE_ID                                   \n",
       " 5269                                         0   \n",
       " 1048                                         0   \n",
       " 8828                                         0   \n",
       " 3069                                     31993   \n",
       " 11309                                        0   \n",
       " ...                                        ...   \n",
       " 25912                                        0   \n",
       " 9712                                         0   \n",
       " 10284                                   145872   \n",
       " 6929                                         0   \n",
       " 23085                                     3494   \n",
       " \n",
       "                 CopyNumber_Th2_02_steady-state  \\\n",
       " ENGREZ_GENE_ID                                   \n",
       " 5269                                         0   \n",
       " 1048                                         0   \n",
       " 8828                                         0   \n",
       " 3069                                     34241   \n",
       " 11309                                        0   \n",
       " ...                                        ...   \n",
       " 25912                                     4990   \n",
       " 9712                                         0   \n",
       " 10284                                   137003   \n",
       " 6929                                      5217   \n",
       " 23085                                     6967   \n",
       " \n",
       "                 CopyNumber_Th2_03_steady-state  \\\n",
       " ENGREZ_GENE_ID                                   \n",
       " 5269                                         0   \n",
       " 1048                                         0   \n",
       " 8828                                         0   \n",
       " 3069                                     29452   \n",
       " 11309                                        0   \n",
       " ...                                        ...   \n",
       " 25912                                     2476   \n",
       " 9712                                         0   \n",
       " 10284                                   104060   \n",
       " 6929                                         0   \n",
       " 23085                                      368   \n",
       " \n",
       "                 CopyNumber_Th2_04_steady-state   pos  0.5  top 0.25  \n",
       " ENGREZ_GENE_ID                                                       \n",
       " 5269                                         0  0.00    0         0  \n",
       " 1048                                         0  0.00    0         0  \n",
       " 8828                                         0  0.00    0         0  \n",
       " 3069                                     55736  1.00    1         0  \n",
       " 11309                                        0  0.00    0         0  \n",
       " ...                                        ...   ...  ...       ...  \n",
       " 25912                                        0  0.50    0         0  \n",
       " 9712                                       791  0.25    0         0  \n",
       " 10284                                   176350  1.00    1         1  \n",
       " 6929                                         0  0.25    0         0  \n",
       " 23085                                     4612  1.00    1         0  \n",
       " \n",
       " [9839 rows x 7 columns],\n",
       " 'Th17':                 CopyNumber_Th17_01_steady-state  \\\n",
       " ENGREZ_GENE_ID                                    \n",
       " 5269                                          0   \n",
       " 1048                                          0   \n",
       " 8828                                          0   \n",
       " 3069                                      44039   \n",
       " 11309                                         0   \n",
       " ...                                         ...   \n",
       " 25912                                         0   \n",
       " 9712                                          0   \n",
       " 10284                                    124929   \n",
       " 6929                                          0   \n",
       " 23085                                      4446   \n",
       " \n",
       "                 CopyNumber_Th17_02_steady-state  \\\n",
       " ENGREZ_GENE_ID                                    \n",
       " 5269                                          0   \n",
       " 1048                                          0   \n",
       " 8828                                          0   \n",
       " 3069                                      36377   \n",
       " 11309                                         0   \n",
       " ...                                         ...   \n",
       " 25912                                         0   \n",
       " 9712                                       1254   \n",
       " 10284                                    167703   \n",
       " 6929                                          0   \n",
       " 23085                                      2041   \n",
       " \n",
       "                 CopyNumber_Th17_03_steady-state  \\\n",
       " ENGREZ_GENE_ID                                    \n",
       " 5269                                          0   \n",
       " 1048                                          0   \n",
       " 8828                                          0   \n",
       " 3069                                      44926   \n",
       " 11309                                         0   \n",
       " ...                                         ...   \n",
       " 25912                                         0   \n",
       " 9712                                        836   \n",
       " 10284                                    145642   \n",
       " 6929                                       1335   \n",
       " 23085                                      4897   \n",
       " \n",
       "                 CopyNumber_Th17_04_steady-state  pos  0.5  top 0.25  \n",
       " ENGREZ_GENE_ID                                                       \n",
       " 5269                                          0  0.0    0         0  \n",
       " 1048                                          0  0.0    0         0  \n",
       " 8828                                          0  0.0    0         0  \n",
       " 3069                                      43494  1.0    1         0  \n",
       " 11309                                         0  0.0    0         0  \n",
       " ...                                         ...  ...  ...       ...  \n",
       " 25912                                         0  0.0    0         0  \n",
       " 9712                                          0  0.5    0         0  \n",
       " 10284                                    155576  1.0    1         1  \n",
       " 6929                                       1893  0.5    0         0  \n",
       " 23085                                      7689  1.0    1         0  \n",
       " \n",
       " [9839 rows x 7 columns]}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteomics_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Naive', 'Th1', 'Th2', 'Th17']"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read from /Users/zhzhao/Dropbox/Helikar/pipelines/data/transcriptomics_Naive.csv\n",
      "Read from /Users/zhzhao/Dropbox/Helikar/pipelines/data/transcriptomics_Th1.csv\n",
      "Read from /Users/zhzhao/Dropbox/Helikar/pipelines/data/transcriptomics_Th2.csv\n",
      "Read from /Users/zhzhao/Dropbox/Helikar/pipelines/data/transcriptomics_Th17.csv\n",
      "Read from /Users/zhzhao/Dropbox/Helikar/pipelines/data/transcriptomics_Treg.csv\n"
     ]
    }
   ],
   "source": [
    "filename = 'GeneExpressionDataUsed.xlsx'\n",
    "# sheet_name = list(range(5)) # first 5 sheets\n",
    "\n",
    "# gse2770 = GSEproject('GSE2770',projectdir)\n",
    "\n",
    "inqueryFullPath = os.path.join(projectdir, 'data', filename)\n",
    "xl = pd.ExcelFile(inqueryFullPath)\n",
    "sheet_name = xl.sheet_names\n",
    "\n",
    "tests = []\n",
    "datas = []\n",
    "for sheet in sheet_name:\n",
    "    # print(list(inqueries[i]))\n",
    "    filename = 'transcriptomics_{}.csv'.format(sheet)\n",
    "    fullsavepath = os.path.join(projectdir,'data',filename)\n",
    "    data = pd.read_csv(fullsavepath,index_col='ENTREZ_GENE_ID')\n",
    "    print('Read from {}'.format(fullsavepath))\n",
    "    datas.append(data)\n",
    "    tests.append(sheet)\n",
    "transcriptomics_dict = dict(zip(tests,datas))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "## Merge Transcriptomics and Proteomics\n",
    "1. Intersection of Both 0.5\n",
    "2. Transcriptomics 0.9\n",
    "3. Proteomics top 25% quantile value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "699 single ENTREZ_GENE_IDs to merge\n",
      "id_list: 3539, set: 3429\n",
      "entrez_single_id_list: 22027, set: 21337\n",
      "entrez_id_list: 1263, set: 1261\n",
      "dups: 218, set: 108\n",
      "1233 id merged\n",
      "698 single ENTREZ_GENE_IDs to merge\n",
      "id_list: 3505, set: 3412\n",
      "entrez_single_id_list: 21932, set: 21242\n",
      "entrez_id_list: 1259, set: 1257\n",
      "dups: 184, set: 91\n",
      "1230 id merged\n",
      "676 single ENTREZ_GENE_IDs to merge\n",
      "id_list: 3394, set: 3308\n",
      "entrez_single_id_list: 21991, set: 21301\n",
      "entrez_id_list: 1235, set: 1233\n",
      "dups: 170, set: 84\n",
      "1210 id merged\n",
      "677 single ENTREZ_GENE_IDs to merge\n",
      "id_list: 3415, set: 3325\n",
      "entrez_single_id_list: 22088, set: 21398\n",
      "entrez_id_list: 1243, set: 1241\n",
      "dups: 173, set: 83\n",
      "1214 id merged\n"
     ]
    }
   ],
   "source": [
    "keys1 = proteomics_dict.keys()\n",
    "keys2 = transcriptomics_dict.keys()\n",
    "tests = set(keys1).intersection(set(keys2))\n",
    "for test in tests:\n",
    "    merge_data = pd.DataFrame()\n",
    "    \n",
    "    prote_data = proteomics_dict[test].loc[:,['0.5','top 0.25']]\n",
    "    prote_data.rename(columns={'0.5':'prote_0.5','top 0.25':'prote_top'},inplace=True)\n",
    "\n",
    "    trans_data = transcriptomics_dict[test].loc[:,['0.5','0.9']]\n",
    "    trans_data.rename(columns={'0.5':'trans_0.5','0.9':'trans_top'},inplace=True)\n",
    "    \n",
    "    print(trans_data.shape)\n",
    "    print(prote_data.shape)\n",
    "    \n",
    "    merge_data = prote_data.join(trans_data, how='outer')\n",
    "    merge_data = mergeLogicalTable(merge_data)\n",
    "    merge_data['Express'] = 0\n",
    "    merge_data.loc[merge_data[['prote_0.5','trans_0.5']].sum(axis=1)==2, 'Express'] = 1\n",
    "    merge_data.loc[merge_data[['prote_top','trans_top']].sum(axis=1)>0,'Express'] = 1\n",
    "    filepath = os.path.join(projectdir,'data','merged_{}.csv'.format(test))\n",
    "    merge_data.to_csv(filepath, index_label='ENTREZ_GENE_ID')\n",
    "    \n",
    "    print('{}: save to {}\\n'.format(test, filepath))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 single ENTREZ_GENE_IDs to merge\n",
      "id_list: 2241, set: 2226\n",
      "entrez_single_id_list: 18263, set: 18263\n",
      "entrez_id_list: 819, set: 819\n",
      "dups: 30, set: 15\n",
      "814 id merged\n",
      "Treg: save to /Users/zhzhao/Dropbox/Helikar/pipelines/data/merged_Treg.csv\n"
     ]
    }
   ],
   "source": [
    "# Those only with transcriptomics or proteomics\n",
    "tests2 = (set(keys1).union(set(keys2))).difference(tests)\n",
    "for test in tests2:\n",
    "    if test in keys1:\n",
    "        merge_data = proteomics_dict[test].loc[:,['0.5','top 0.25']]\n",
    "        merge_data.rename(columns={'0.5':'0.5','top 0.25':'top'},inplace=True)\n",
    "    elif test in keys2:\n",
    "        merge_data = transcriptomics_dict[test].loc[:,['0.5','0.9']]\n",
    "        merge_data.rename(columns={'0.5':'0.5','0.9':'top'},inplace=True)\n",
    "\n",
    "    merge_data = mergeLogicalTable(merge_data)\n",
    "    merge_data['Express'] = 0\n",
    "    merge_data.loc[merge_data[['0.5','top']].sum(axis=1)>0, 'Express'] = 1\n",
    "    filepath = os.path.join(projectdir,'data','merged_{}.csv'.format(test))\n",
    "    merge_data.to_csv(filepath, index_label='ENTREZ_GENE_ID')\n",
    "    \n",
    "    print('{}: save to {}\\n'.format(test, filepath))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = 'Naive'\n",
    "merge_data = pd.DataFrame()\n",
    "\n",
    "prote_data = proteomics_dict[test].loc[:,['0.5','top 0.25']]\n",
    "prote_data.rename(columns={'0.5':'prote_0.5','top 0.25':'prote_top'},inplace=True)\n",
    "\n",
    "trans_data = transcriptomics_dict[test].loc[:,['0.5','0.9']]\n",
    "trans_data.rename(columns={'0.5':'trans_0.5','0.9':'trans_top'},inplace=True)\n",
    "\n",
    "# merge_data = pd.concat([prote_data,trans_data], axis=1, inplace=True)\n",
    "merge_data = prote_data.join(trans_data, how='outer')\n",
    "merge_data = mergeLogicalTable(merge_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0.5</th>\n",
       "      <th>0.9</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENTREZ_GENE_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1000</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10000</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10001</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10002</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10003</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100033411</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10004</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100049587</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10005</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10006</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10007</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10008</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10009</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100093630</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100093698</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1001</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10010</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100101467</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10011</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100113407</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100124700</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100125288</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100126572</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100126791</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100126793</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100127206</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100127886</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9963</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9965</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9966</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9967</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9968</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9969</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9970</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9971</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9972</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9973</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9975</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9976</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9978</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9980</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9982</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9984</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9985</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9986</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9987</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9988</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9989</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9990</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9991</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9992</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9993</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9994</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9997</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19082 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                0.5  0.9\n",
       "ENTREZ_GENE_ID          \n",
       "1                 1    1\n",
       "10                0    0\n",
       "100               1    0\n",
       "1000              0    0\n",
       "10000             1    0\n",
       "...             ...  ...\n",
       "9991              1    1\n",
       "9992              0    0\n",
       "9993              1    1\n",
       "9994              1    0\n",
       "9997              1    0\n",
       "\n",
       "[19082 rows x 2 columns]"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcriptomics_dict[test].loc[:,['0.5','0.9']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prote_0.5</th>\n",
       "      <th>prote_top</th>\n",
       "      <th>trans_0.5</th>\n",
       "      <th>trans_top</th>\n",
       "      <th>Express</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENTREZ_GENE_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1000</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10000</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100009676</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10001</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10002</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10003</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100033411</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10004</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100048912</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100049587</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100049716</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10005</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10006</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10007</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10008</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10009</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100093698</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1001</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10010</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100101266</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100101467</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100101938</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10011</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100113403</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100113407</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100124700</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100125288</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9963</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9965</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9966</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9967</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9968</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9969</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9970</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9971</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9972</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9973</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9975</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9976</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9978</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9980</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9982</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9984</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9985</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9986</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9987</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9988</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9989</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9990</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9991</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9992</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9993</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9994</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9997</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21835 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                prote_0.5  prote_top  trans_0.5  trans_top  Express\n",
       "ENTREZ_GENE_ID                                                     \n",
       "1                     1.0        0.0        1.0        0.0        1\n",
       "10                    NaN        NaN        0.0        0.0        0\n",
       "100                   1.0        1.0        1.0        1.0        1\n",
       "1000                  0.0        0.0        0.0        0.0        0\n",
       "10000                 1.0        0.0        1.0        0.0        1\n",
       "...                   ...        ...        ...        ...      ...\n",
       "9991                  1.0        1.0        1.0        1.0        1\n",
       "9992                  NaN        NaN        0.0        0.0        0\n",
       "9993                  NaN        NaN        1.0        1.0        1\n",
       "9994                  1.0        0.0        1.0        0.0        1\n",
       "9997                  1.0        1.0        1.0        0.0        1\n",
       "\n",
       "[21835 rows x 5 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merge_data['Express'] = 0\n",
    "merge_data.loc[merge_data[['prote_0.5','trans_0.5']].sum(axis=1)==2, 'Express'] = 1\n",
    "merge_data.loc[merge_data[['prote_top','trans_top']].sum(axis=1)>0,'Express'] = 1\n",
    "merge_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(21807, 2)\n",
      "(9839, 2)\n"
     ]
    }
   ],
   "source": [
    "print(trans_data.shape)\n",
    "print(prote_data.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Merge Output\n",
    "def mergeLogicalTable(df_results):\n",
    "    '''\n",
    "    Merge the Rows of Logical Table belongs to the same ENTREZ_GENE_ID\n",
    "    :param df_results:\n",
    "    :return: pandas dataframe of merged table\n",
    "    '''\n",
    "    # step 1: get all plural ENTREZ_GENE_IDs in the input table, extract unique IDs\n",
    "    df_results.reset_index(drop=False, inplace=True)\n",
    "    df_results['ENTREZ_GENE_ID'] = df_results['ENTREZ_GENE_ID'].str.replace(\" /// \", \"//\")\n",
    "    id_list = []\n",
    "    entrez_single_id_list = df_results[~df_results['ENTREZ_GENE_ID'].str.contains(\"//\")]['ENTREZ_GENE_ID'].tolist()\n",
    "    entrez_id_list = df_results[df_results['ENTREZ_GENE_ID'].str.contains(\"//\")]['ENTREZ_GENE_ID'].tolist()\n",
    "    for entrez_id in entrez_id_list:\n",
    "        entrez_ids = entrez_id.split(\"//\")\n",
    "        id_list.extend(entrez_ids)\n",
    "        df_dups = pd.DataFrame([], columns=list(df_results), index=list(range(len(entrez_ids))))\n",
    "        dup_rows = []\n",
    "        for eid in entrez_ids:\n",
    "            rows = df_results.loc[df_results['ENTREZ_GENE_ID']==entrez_id].copy()\n",
    "            rows['ENTREZ_GENE_ID'] = eid\n",
    "            dup_rows.append(rows)\n",
    "        df_results = df_results.append(dup_rows,ignore_index=True)\n",
    "        df_results.drop(df_results[df_results['ENTREZ_GENE_ID']==entrez_id].index, inplace=True)\n",
    "\n",
    "    # step 2: print out information about merge\n",
    "    common_elements = list(set(entrez_single_id_list).intersection(set(id_list)))\n",
    "    # information of merge\n",
    "    print('{} single ENTREZ_GENE_IDs to merge'.format(len(common_elements)))\n",
    "    print('id_list: {}, set: {}'.format(len(id_list),len(set(id_list))))\n",
    "    print('entrez_single_id_list: {}, set: {}'.format(len(entrez_single_id_list),len(set(entrez_single_id_list))))\n",
    "    print('entrez_id_list: {}, set: {}'.format(len(entrez_id_list),len(set(entrez_id_list))))\n",
    "\n",
    "    dups = [x for x in id_list if id_list.count(x) > 1]\n",
    "    # dups = list(set(id_list))\n",
    "    print('dups: {}, set: {}'.format(len(dups),len(set(dups))))\n",
    "\n",
    "    full_entre_id_sets = []\n",
    "    cnt = 0\n",
    "    entrez_dups_list = []\n",
    "    idx_list = list(range(len(entrez_id_list)))\n",
    "    for idx1 in range(len(entrez_id_list)):\n",
    "        if not idx1 in idx_list:\n",
    "            continue\n",
    "        set1 = set(entrez_id_list[idx1].split(\"//\"))\n",
    "        idx_list.remove(idx1)\n",
    "        toremove = []\n",
    "        for idx2 in idx_list:\n",
    "            set2 = set(entrez_id_list[idx2].split(\"//\"))\n",
    "            intersect = set1.intersection(set2)\n",
    "            if bool(intersect):\n",
    "                set1 = set1.union(set2)\n",
    "                toremove.append(idx2)\n",
    "        for idx3 in toremove:\n",
    "            idx_list.remove(idx3)\n",
    "        sortlist = list(set1)\n",
    "        sortlist.sort(key=int)\n",
    "        new_entrez_id = \" /// \".join(sortlist)\n",
    "        full_entre_id_sets.append(new_entrez_id)\n",
    "    full_entre_id_sets = list(set(full_entre_id_sets))\n",
    "\n",
    "    for full_entrez_id in full_entre_id_sets:\n",
    "        singles = full_entrez_id.split(\" /// \")\n",
    "        entrez_dups_list.append(singles)\n",
    "        cnt += 1\n",
    "\n",
    "    print('{} id merged'.format(cnt))\n",
    "    entrez_dups_dict = dict(zip(full_entre_id_sets,entrez_dups_list))\n",
    "    # full_entre_id_sets = list(set(full_entre_id_sets))\n",
    "\n",
    "    # df_results.reset_index(inplace=True)\n",
    "    for merged_entrez_id, entrez_dups_list in entrez_dups_dict.items():\n",
    "        df_results['ENTREZ_GENE_ID'].replace(to_replace=entrez_dups_list,\n",
    "                                             value=merged_entrez_id,\n",
    "                                             inplace=True)\n",
    "    # df_results.drop_duplicates(subset=['ENTREZ_GENE_ID'], keep='first', inplace=True)\n",
    "    df_results.set_index('ENTREZ_GENE_ID', inplace=True)\n",
    "\n",
    "    df_output = df_results.fillna(-1).groupby(level=0).max()\n",
    "    df_output.replace(-1, np.nan, inplace=True)\n",
    "    return df_output\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prote_0.5</th>\n",
       "      <th>prote_top</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENTREZ_GENE_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>84939//3662</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3045//100187828</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2618//2617</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7006//100124696</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5859//2058</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84174//4141</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8360//8368//8367//8361//8364//8294//8366//554313//8362//8363//8370//121504//8359//8365</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3017//8343//8344//8346//8339//8347</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8351//8353//8354//8355//8357//8358//8968//8356//8350//8352</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387680//253725</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126961//333932//653604</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>445815//11217</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7795//51072</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    prote_0.5  prote_top\n",
       "ENTREZ_GENE_ID                                                          \n",
       "84939//3662                                                 1          0\n",
       "3045//100187828                                             0          0\n",
       "2618//2617                                                  1          1\n",
       "7006//100124696                                             0          0\n",
       "5859//2058                                                  1          1\n",
       "84174//4141                                                 1          1\n",
       "8360//8368//8367//8361//8364//8294//8366//55431...          1          1\n",
       "3017//8343//8344//8346//8339//8347                          1          1\n",
       "8351//8353//8354//8355//8357//8358//8968//8356/...          1          1\n",
       "387680//253725                                              1          0\n",
       "126961//333932//653604                                      1          1\n",
       "445815//11217                                               1          0\n",
       "7795//51072                                                 1          1"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prote_data[prote_data.index.str.contains('//')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "676 single ENTREZ_GENE_IDs to merge\n",
      "id_list: 3394, set: 3308\n",
      "entrez_single_id_list: 21991, set: 21301\n",
      "entrez_id_list: 1235, set: 1233\n",
      "dups: 170, set: 84\n",
      "1210 id merged\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prote_0.5</th>\n",
       "      <th>prote_top</th>\n",
       "      <th>trans_0.5</th>\n",
       "      <th>trans_top</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENTREZ_GENE_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1000</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10000</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100009676</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10001</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10002</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10003</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100033411</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10004</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100048912</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100049587</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100049716</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10005</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10006</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10007</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10008</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10009</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100093698</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1001</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10010</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100101266</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100101467</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100101938</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10011</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100113403</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100113407</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100124700</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100125288</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9963</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9965</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9966</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9967</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9968</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9969</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9970</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9971</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9972</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9973</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9975</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9976</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9978</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9980</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9982</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9984</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9985</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9986</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9987</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9988</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9989</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9990</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9991</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9992</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9993</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9994</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9997</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21835 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                prote_0.5  prote_top  trans_0.5  trans_top\n",
       "ENTREZ_GENE_ID                                            \n",
       "1                     1.0        0.0        1.0        0.0\n",
       "10                    NaN        NaN        0.0        0.0\n",
       "100                   1.0        1.0        1.0        1.0\n",
       "1000                  0.0        0.0        0.0        0.0\n",
       "10000                 1.0        0.0        1.0        0.0\n",
       "...                   ...        ...        ...        ...\n",
       "9991                  1.0        1.0        1.0        1.0\n",
       "9992                  NaN        NaN        0.0        0.0\n",
       "9993                  NaN        NaN        1.0        1.0\n",
       "9994                  1.0        0.0        1.0        0.0\n",
       "9997                  1.0        1.0        1.0        0.0\n",
       "\n",
       "[21835 rows x 4 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_boolean = mergeLogicalTable(merge_data)\n",
    "test_boolean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'Gene Symbol'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-82-026c1f0ac07d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mproteomics_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprojectdir\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'data'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'proteomics_data.csv'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mindex_col\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'ENTREZ_GENE_ID'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mproteomics_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgroupby\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Gene Symbol'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0magg\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'max'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mgroupby\u001b[0;34m(self, by, axis, level, as_index, sort, group_keys, squeeze, observed, **kwargs)\u001b[0m\n\u001b[1;32m   7894\u001b[0m             \u001b[0msqueeze\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msqueeze\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   7895\u001b[0m             \u001b[0mobserved\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mobserved\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 7896\u001b[0;31m             \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   7897\u001b[0m         )\n\u001b[1;32m   7898\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.7/site-packages/pandas/core/groupby/groupby.py\u001b[0m in \u001b[0;36mgroupby\u001b[0;34m(obj, by, **kwds)\u001b[0m\n\u001b[1;32m   2476\u001b[0m         \u001b[0;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"invalid type: {}\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2477\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2478\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mklass\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mby\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/usr/local/lib/python3.7/site-packages/pandas/core/groupby/groupby.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, obj, keys, axis, level, grouper, exclusions, selection, as_index, sort, group_keys, squeeze, observed, **kwargs)\u001b[0m\n\u001b[1;32m    389\u001b[0m                 \u001b[0msort\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msort\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    390\u001b[0m                 \u001b[0mobserved\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mobserved\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 391\u001b[0;31m                 \u001b[0mmutated\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmutated\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    392\u001b[0m             )\n\u001b[1;32m    393\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.7/site-packages/pandas/core/groupby/grouper.py\u001b[0m in \u001b[0;36m_get_grouper\u001b[0;34m(obj, key, axis, level, sort, observed, mutated, validate)\u001b[0m\n\u001b[1;32m    619\u001b[0m                 \u001b[0min_axis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlevel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgpr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgpr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    620\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 621\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgpr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    622\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgpr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mGrouper\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mgpr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkey\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    623\u001b[0m             \u001b[0;31m# Add key to exclusions\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'Gene Symbol'"
     ]
    }
   ],
   "source": [
    "proteomics_data = pd.read_csv(os.path.join(projectdir,'data','proteomics_data.csv'),index_col='ENTREZ_GENE_ID')\n",
    "df = proteomics_data.groupby('Gene Symbol').agg('max')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
